{% extends 'layouts/base.html' %}
{% block title %}{{ config.MAIN_TITLE|safe }}{% endblock %}
{% block meta_description %}{% endblock %}
{% block head %}
    <link rel="stylesheet"
          href="//cdnjs.cloudflare.com/ajax/libs/animate.css/3.7.0/animate.min.css"/>

    <link rel="stylesheet"
          href="//cdn.jsdelivr.net/npm/bootstrap-select@1.13.7/dist/css/bootstrap-select.min.css"/>

    <link rel="stylesheet"
          href="//cdn.datatables.net/1.11.5/css/jquery.dataTables.min.css"/>

    <link rel="stylesheet"
          href="//cdn.datatables.net/1.11.5/css/dataTables.bootstrap4.min.css"/>

    <link rel="stylesheet"
          href="//unpkg.com/@genome-spy/core@0.19.0/dist/style.css"/>

    <link rel="stylesheet"
          href="{{ url_for('static', filename='vendor/please-wait-0.0.5/please-wait.css') }}"/>

    <link rel="stylesheet"
          href="{{ url_for('static', filename='vendor/dynatables-0.3.1/jquery.dynatable.css')}}"/>

    <link rel="stylesheet"
          href="{{ url_for('static', filename='vendor/bootstrap-multiselect-0.9.15/dist/css/bootstrap-multiselect.css') }}"/>

    <link rel="stylesheet"
          href="{{ url_for('static', filename='styles/main.css') }}"/>

    <!-- Global site tag (gtag.js) - Google Analytics -->
    <script async src="https://www.googletagmanager.com/gtag/js?id=G-D1707LSMF4"></script>
    <script>
      window.dataLayer = window.dataLayer || [];
      function gtag(){dataLayer.push(arguments);}
      gtag('js', new Date());
      gtag('config', 'G-D1707LSMF4');
    </script>

    <style>
    <!--
        body {
            /*font-family: 'Roboto', sans-serif !important;*/
            /*font-family: IBM Plex Sans, sans-serif !important;*/
            font-family: system-ui, "Segoe UI", Roboto, Helvetica, Arial, sans-serif !important;
        }

        .genome-spy .tooltip {
            opacity: 100;
        }

        .snpMenuIcon a:hover {
            background: #3374C2;
        }

.modal.modal-fullscreen .modal-dialog {
  width: 100vw;
  height: 100vh;
  margin: 0;
  padding: 0;
  max-width: none;
}

.modal.modal-fullscreen .modal-content {
  height: auto;
  height: 100vh;
  border-radius: 0;
  border: none;
}

.modal.modal-fullscreen .modal-body {
  overflow-y: auto;
}

    //-->
    </style>
{% endblock %}
{% block body %}
<body>
<div class="inner">

    <nav id="dataSetSelectMenu" class="navbar navbar-expand-sm bg-dark navbar-dark fixed-top">
        <a class="navbar-brand" href="#">{{ config.NAV_TITLE|safe }}</a>
        <!-- will be filled in programatically //-->

    </nav>

    <div class="container-fluid">

        <main class="col-12" role="main">
            <div class="row">

                <div class="col-md-6 col-lg-5 col-xl-4">

                    <div class="row">
                        <div class="col">

                            <!-- Gene/Protein/Pheno selection //-->
                            <div class="card h-100">
                                <div class="card-header">
                                    <ul class="nav nav-tabs" role="tablist">
                                        <li class="nav-item">
                                            <a id="navPheno"
                                               class="nav-link"
                                               data-toggle="tab"
                                               href="#tabSearchPheno"><i class="fab fa-searchengin"></i> Data Search</a>
                                        </li>
                                        <li class="nav-item">
                                            <a id="navGene"
                                               class="nav-link"
                                               data-toggle="tab"
                                               href="#tabSearchGene"><i class="fas fa-search"></i> {{ config.TAB_SEARCH_MRNA_PROTEIN_TEXT|safe }}</a>
                                        </li>
                                        <li class="nav-item">
                                            <a id="navLodPeaks"
                                               class="nav-link"
                                               data-toggle="tab"
                                               href="#tabLodPeaks"><i class="fas fa-th"></i> {{ config.TAB_LOD_PEAKS_TEXT|safe }}</a>
                                        </li>
                                    </ul>
                                </div>

                                <div class="card-body">
                                    <div class="tab-content">
                                        <div class="tab-pane h-100" role="tabpanel" id="tabSearchPheno">

                                            <div class="row">
                                                <div class="col" id="searchPheno">
                                                </div>
                                            </div>

                                            <div class="row row-spacer"></div>

                                            <div class="row">
                                                <div class="col" id="searchPhenoCategoryText">
                                                </div>
                                            </div>

                                            <div class="row">
                                                <div class="col" id="searchPhenoCategory">
                                                </div>
                                            </div>

                                            <div class="row">
                                                <div class="col">
                                                    <div class="py-0" id="searchPhenoResultsInfo">
                                                    </div>
                                                </div>
                                            </div>

                                            <div class="row">
                                                <div class="col results-wrapper">
                                                    <div id="searchPhenoResultsDiv"></div>
                                                </div>
                                            </div>
                                        </div>

                                        <div class="tab-pane h-100" role="tabpanel" id="tabSearchGene">
                                            <div class="row">
                                                <div class="col">
                                                    <div class="input-group">
                                                        <input type="text" id="searchTerm" name="searchTerm" value="{{ search_term }}" class="form-control" placeholder="{{ config.SEARCH_BOX_TEXT|safe }}">
                                                        <span class="input-group-append">
                                                            <button id="btnGo" class="btn btn-primary"><i class="fas fa-search"></i></button>
                                                        </span>
                                                    </div>
                                                </div>
                                            </div>

                                            <div class="row">
                                                <div class="col">
                                                    <div class="py-2" id="searchResultsTableInfo"></div>
                                                </div>
                                            </div>

                                            <div class="row">
                                                <div class="col">
                                                    <div id="searchResultsDiv"></div>
                                                </div>
                                            </div>
                                        </div>

                                        <div class="tab-pane h-100" role="tabpanel" id="tabLodPeaks">

                                            <div class="row">
                                                <div class="col">
                                                    <div id="divInteractiveCovariatesPeaks"></div>
                                                </div>
                                            </div>
                                            <div class="row" id="divLODPeaksMenu"></div>
                                            <div class="row">
                                                <div class="col">
                                                    <div id="divLODPeaks"></div>
                                                </div>
                                            </div>

                                        </div>
                                    </div>
                                </div>
                            </div>

                            <!-- end Gene/Protein/Pheno selection //-->

                        </div>
                    </div>

                    <div class="row row-spacer"></div>

                    <div class="row invisible" id="divProfileCorrelation">
                        <div class="col">

                            <div class="card h-100">
                                <div class="card-header" id="profilePlot">
                                    <ul id="infoTabs" class="nav nav-tabs" role="tablist">
                                        <li class="nav-item">
                                            <a class="nav-link active" id="navProfile" data-toggle="tab" href="#tabProfile"><i class="fas fa-calculator"></i> {{ config.TAB_PROFILE_PLOT_TEXT|safe }}</a>
                                        </li>
                                        <li class="nav-item">
                                            <a class="nav-link" id="navCorrelation" data-toggle="tab" href="#tabCorrelation"><i class="fa-brands fa-uncharted"></i> {{ config.TAB_CORRELATION_TEXT|safe }}</a>
                                        </li>
                                    </ul>
                                </div>
                                <div class="card-body">

                                    <div class="tab-content">
                                        <div role="tabpanel" class="tab-pane active" id="tabProfile">

                                            <div class="row" id="divProfilePlotOptions">

                                            </div>

                                            <div class="row" id="fact-svg-panel">
                                                <div class="col">
                                                    <div id="profilePlotChart"></div>
                                                </div>
                                            </div>
                                        </div>
                                        <div role="tabpanel" class="tab-pane" id="tabCorrelation">
                                            <div class="row">
                                                <div class="col font-weight-bold">
                                                    <span class="align-bottom">{{ config.CORRELATION_SELECT_DATASET_TEXT|safe }}</span>
                                                </div>
                                            </div>

                                            <div class="row">
                                                <div class="col" id="correlationDatasetSelection">
                                                </div>
                                            </div>

                                            <div class="row row-spacer"></div>

                                            <div class="row">
                                                <div class="col font-weight-bold" id="correlationCovariateSelectionText">
                                                    <span class="align-bottom">{{ config.CORRELATION_SELECT_COVARIATE_TEXT|safe }}</span>
                                                </div>
                                            </div>

                                            <div class="row">
                                                <div class="col" id="correlationCovariateSelection">
                                                </div>
                                            </div>

                                            <div class="row row-spacer"></div>

                                            <div class="row">
                                                <div class="col" id="correlationDataTable">
                                                </div>
                                            </div>

                                            <div class="row row-spacer"></div>

                                            <div class="row">
                                                <div class="col" id="correlationImputation">
                                                </div>
                                            </div>

                                            <div class="row row-spacer"></div>

                                            <div class="row">
                                                <div class="col">
                                                    <button id="btnDownloadCorrelationData" type="button" class="btn btn-sm btn-flat btn-primary">
                                                        <i class="fas fa-download"></i> {{ config.CORRELATION_BUTTON_DOWNLOAD_TEXT|safe }}</button>
                                                </div>
                                            </div>

                                            <div class="row row-spacer"></div>

                                            <div class="row">
                                                <div class="col">
                                                    {{ config.CORRELATION_SELECT_ID_TEXT|safe }}
                                                </div>
                                            </div>
                                            <div class="row">
                                                <div class="col" id="correlationDataPlot">
                                                    <div id="plotCorrelation"></div>
                                                </div>
                                            </div>

                                            <div id="divCorrelationCovars">

                                                <div class="row row-spacer"></div>

                                                <div class="row">
                                                    <div class="col font-weight-bold">
                                                        {{ config.CORRELATION_SELECT_SERIES_TEXT|safe }}
                                                    </div>
                                                </div>

                                                <div class="row">
                                                    <div class="col" id="divFactorSeriesCorrelation">
                                                    </div>
                                                </div>
                                            </div>
                                        </div>
                                    </div>
                                </div>
                            </div>

                            <!-- end Correlation //-->


                        </div>
                    </div>
                </div>

                <!-- Second column //-->

                <div class="col-md-6 col-lg-7 col-xl-8">

                    <!-- LOD Scan section //-->

                    <!-- only display at first, remove after a succesful selection //-->
                    <div class="row" id="divWelcomeMesage">
                        <div class="col">
                            <div class="jumbotron jumbotron-fluid" style="background: none">
                                <div class="container">
                                    <h1 class="display-3">{{ config.JUMBO_TITLE|safe }}</h1>
                                    <p class="lead">{{ config.JUMBO_TEXT|safe }}</p>
                                </div>
                            </div>
                        </div>
                    </div>

                    <div class="row">
                        <div class="col" id="divGenome">
                        </div>
                    </div>

                    <div class="row invisible" id="divItemInfo">
                        <div class="col">
                            <div class="card">
                                <div class="card-header">
                                    <a class="collapsed" data-toggle="collapse" href="#collapseExample" aria-expanded="false" aria-controls="collapseExample">
                                    <i class="fa" aria-hidden="true"></i>
                                    <span id="div_current_item_information_header"></span>
                                    </a>
                                </div>
                                <div class="card-body collapse" id="collapseExample">
                                    <div id="div_current_item_information_body"></div>
                                </div>
                            </div>
                        </div>
                    </div>

                    <div class="row invisible" id="divLOD">
                        <div class="col">
                            <div class="card h-100">
                                <div class="card-header" id="lod">
                                    <i class="fas fa-chart-area"></i> {{ config.LOD_CARD_TITLE |safe }}
                                </div>
                                <div class="card-body">

                                    <div class="row">
                                        <div class="col">
                                            <div id="divInteractiveCovariatesLOD"></div>
                                        </div>
                                    </div>

                                    <div class="row">
                                        <div class="col">
                                            <div id="lodPlotChart"></div>
                                        </div>
                                    </div>

                                    <div class="row">
                                        <div class="col">
                                            <div id="lodPlotChartCovariateFull"></div>
                                        </div>
                                    </div>

                                    <div class="row">
                                        <div class="col">
                                            <div id="lodPlotChartCovariateDiff"></div>
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>

                    <!-- end LOD Scan section //-->

                    <div class="row row-spacer"></div>

                    <div class="row invisible" id="divSecondRow">



                        <!-- Secondary plots //-->

                        <div class="col">
                            <div class="card" id="secondaryPlots">
                                <div class="card-header">
                                    <ul id="secondaryPlotTabs" class="nav nav-tabs" role="tablist">
                                        <li class="nav-item">
                                            <a class="nav-link active" id="navPlotEffect" data-toggle="tab" href="#tabPlotEffect"><i class="fa-solid fa-chart-line"></i> {{ config.TAB_EFFECT_PLOT_TEXT|safe }}</a>
                                        </li>
                                        <li class="nav-item">
                                            <a class="nav-link" id="navPlotMediation" data-toggle="tab" href="#tabPlotMediation"><i class="far fa-calendar-alt fa-rotate-180"></i> {{ config.TAB_MEDIATION_PLOT_TEXT|safe }}</a>
                                        </li>
                                        <li class="nav-item">
                                            <a class="nav-link" id="navPlotSNPs" data-toggle="tab" href="#tabPlotSNPs"><i class="fa-solid fa-chart-gantt"></i> {{ config.TAB_SNP_ASSOCIATION_TEXT|safe }}</a>
                                        </li>
                                    </ul>
                                </div>
                                <div class="card-body">
                                    <div class="tab-content">
                                        <div role="tabpanel" class="tab-pane active" id="tabPlotEffect">

                                        {% if 'EFFECT_PLOT_OPTIONS' in config %}
                                            {% set EFFECT_PLOT_OPTIONS = config.EFFECT_PLOT_OPTIONS %}
                                        {% else %}
                                            {% set EFFECT_PLOT_OPTIONS = 'NB' %}
                                        {% endif %}

                                        <input type="hidden" id="effect_plot_options" value="{{EFFECT_PLOT_OPTIONS}}"/>

                                        {% if EFFECT_PLOT_OPTIONS == 'NB' %}
                                            <div class="row">
                                                <div class="col">
                                                    <div class="custom-control custom-switch">
                                                        <input type="checkbox" class="custom-control-input" id="performBLUP">
                                                        <label class="custom-control-label" for="performBLUP">{{ config.SWITCH_BLUP_TEXT|safe }}</label>
                                                    </div>
                                                </div>
                                            </div>
                                        {% elif EFFECT_PLOT_OPTIONS == 'N' %}
                                            <div class="row">
                                                <div class="col">
                                                    <div class="custom-control custom-switch">
                                                        <input type="checkbox" class="custom-control-input" disabled id="performBLUP">
                                                        <label class="custom-control-label" for="performBLUP">{{ config.SWITCH_BLUP_TEXT|safe }}</label>
                                                    </div>
                                                </div>
                                            </div>
                                        {% elif EFFECT_PLOT_OPTIONS == 'B' %}
                                            <div class="row">
                                                <div class="col">
                                                    <div class="custom-control custom-switch">
                                                        <input type="checkbox" class="custom-control-input" disabled checked id="performBLUP">
                                                        <label class="custom-control-label" for="performBLUP">{{ config.SWITCH_BLUP_TEXT|safe }}</label>
                                                    </div>
                                                </div>
                                            </div>
                                        {% else %}
                                            <div class="row">
                                                <div class="col">
                                                    <div class="custom-control custom-switch">
                                                        <input type="checkbox" class="custom-control-input" id="performBLUP">
                                                        <label class="custom-control-label" for="performBLUP">{{ config.SWITCH_BLUP_TEXT|safe }}</label>
                                                    </div>
                                                </div>
                                            </div>
                                        {% endif %}


                                            <div class="row no-gutters">
                                                <div class="col" id="allEffectPlots">
                                                </div>
                                            </div>
                                        </div>
                                        <div role="tabpanel" class="tab-pane" id="tabPlotMediation">
                                            <div class="row">
                                                <div class="col" id="divMediationSelect">
                                                </div>
                                            </div>
                                            <div class="row">
                                                <div class="col">
                                                    <div id="plotMediation"></div>
                                                </div>
                                            </div>
                                        </div>
                                        <div role="tabpanel" class="tab-pane" id="tabPlotSNPs">
                                            <div class="row" id="snpWindowMenuShift"></div>
                                            <div class="row" id="snpWindowMenu"></div>
                                            <div class="row">
                                                <div class="col">
                                                    <div id="plotSNPAssociation"></div>
                                                </div>
                                            </div>
                                        </div>
                                    </div>
                                </div>
                            </div>
                        </div>

                        <!-- end Secondary plots //-->

                    </div>
                </div>
                <!-- end Second column //-->

            </div>

        </main>
    </div>

<!-- make some nice circles //-->
<div id="ap" class="invisible">
<svg class="lds-curve-bars" width="200px"  height="200px"  xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 100 100" preserveAspectRatio="xMidYMid" style="background: rgba(0, 0, 0, 0) none repeat scroll 0% 0%;"><g transform="translate(50,50)">
</circle><circle cx="0" cy="0" r="18" fill="none" stroke="#1aafd0" stroke-width="1" stroke-dasharray="52 52">
<animateTransform attributeName="transform" type="rotate" values="0 0 0;360 0 0" times="0;1" dur="1.4s" calcMode="spline" keySplines="0.2 0 0.8 1" begin="-0.2" repeatCount="indefinite"></animateTransform>
</circle><circle cx="0" cy="0" r="22" fill="none" stroke="#6a67ce" stroke-width="1" stroke-dasharray="78 78">
<animateTransform attributeName="transform" type="rotate" values="0 0 0;360 0 0" times="0;1" dur=".8s" calcMode="spline" keySplines="0.2 0 0.8 1" begin="-0.4" repeatCount="indefinite"></animateTransform>
</circle><circle cx="0" cy="0" r="26" fill="none" stroke="#ffb900" stroke-width="1" stroke-dasharray="104 104">
<animateTransform attributeName="transform" type="rotate" values="0 0 0;360 0 0" times="0;1" dur="1s" calcMode="spline" keySplines="0.2 0 0.8 1" begin="-0.6" repeatCount="indefinite"></animateTransform>
</circle><circle cx="0" cy="0" r="30" fill="none" stroke="#fc636b" stroke-width="1" stroke-dasharray="130 130">
<animateTransform attributeName="transform" type="rotate" values="0 0 0;360 0 0" times="0;1" dur="1.2s" calcMode="spline" keySplines="0.2 0 0.8 1" begin="-0.8" repeatCount="indefinite"></animateTransform>
</circle></g></svg>
</div>

<!-- Main Footer -->
<nav class="navbar fixed-bottom navbar-dark bg-dark">
    <ul class="navbar-nav mr-auto">
      <li class="nav-item">
        <span class="text-muted">
            <strong>{{ config.CREATED_BY_TEXT|safe }} <a href="{{ config.LAB_URL }}">{{ config.LAB_NAME }}</a>.</strong>
        </span>
      </li>
    </ul>
    <ul class="navbar-nav navbar-expand-sm ml-auto">
        {% if admin %}
        <li class="nav-item pr-5">
            <strong><a class="text-light" href="{{ url_for('admin.index') }}"><i class="fas fa-cogs"></i> Admin</a></strong>
        </li>
        {% endif %}
    <!--
        <li class="nav-item pr-5">
            <strong><a class="text-light" href="{{ url_for('page.dl') }}"><i class="fas fa-file-download"></i> Download Data</a></strong>
        </li>
     //-->
        <li class="nav-item">
            <a class="font-italic text-primary nav-link" id="viewerInfoInformation" data-toggle="modal" data-target="#viewerInfoModal"><i class="fa-solid fa-circle-info"></i> VERSION {{ app_version }}</a>
        </li>
    </ul>
</nav>
<!-- end Main Footer -->


</div>



<!-- START DOWNLOAD DATA MODAL ========================================== //-->
<div class="modal fade modal-fullscreen" id="downloadModal" tabindex="-1" aria-labelledby="downloadLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header rounded-0 bg-dark text-light">
                <h5 class="modal-title" id="downloadLabel">Download Data</h5>
                <a data-dismiss="modal" aria-label="Close"><i class="fa-solid fa-circle-xmark"></i></a>
            </div>

            <div class="modal-body bg-light">
                <div class="container">
                    <div class="row">
                        <div class="col">
                            <h2>{{ config.MAIN_TITLE|safe }}</h2>
                            <hr/>
                        </div>
                    </div>

                    <div class="row">
                        <div class="col">
                            <p class="lead">To download data from {{ config.MAIN_TITLE|safe }}, please click on the filename from the table below.</p>
                        </div>
                    </div>

                    <div class="row">
                        <div class="col">
                            <div id="envInfoDiv"></div>
                        </div>
                    </div>
                </div>
            </div>
            <!--
            <div class="modal-footer">
            </div>
            //-->
        </div>
    </div>
</div>
<!-- END DOWNLOAD DATA MODAL ============================================ //-->

<!-- START VIEWER INFO MODAL ============================================ //-->
<div class="modal fade modal-fullscreen" id="viewerInfoModal" tabindex="-1" aria-labelledby="viewerInfoLabel" aria-hidden="true">
    <div class="modal-dialog">
        <div class="modal-content">
            <div class="modal-header rounded-0 bg-dark text-light">
                <h5 class="modal-title" id="viewerInfoLabel">Viewer Information</h5>
                <a data-dismiss="modal" aria-label="Close"><i class="fa-solid fa-circle-xmark"></i></a>
            </div>

            <div class="modal-body bg-light">
                <div class="container">
                    <div class="row">
                        <div class="col">
                            <h2>{{ config.MAIN_TITLE|safe }}</h2>
                            <hr/>
                        </div>
                    </div>

                    <div class="row">
                        <div class="col">
                            <p class="lead">The QTL Viewer is makes use of a variety of technologies and libraries.</p>
                            QTL Viewer Github Repository: <a target="_new" href="https://github.com/churchill-lab/qtl2web">https://github.com/churchill-lab/qtl2web</a>
                        </div>
                    </div>

                    <br/><br/>

                    <div class="row">
                        <div class="col">
                            <h5 class="modal-title" id="viewerInfoLabel">HTML/Javascript</h5>
                            <hr/>
                            {% for tech in TECHNOLOGIES %}
                                {% if "html" in tech.category %}
                                <a target="_new" href="{{ tech.url }}">{{ tech.name }}</a> {{ tech.version }}<br/>
                                {% endif %}
                            {% endfor %}
                        </div>
                    </div>

                    <br/><br/>

                    <div class="row">
                        <div class="col">
                            <h5 class="modal-title" id="viewerInfoLabel">Python</h5>
                            <hr/>
                            {% for tech in TECHNOLOGIES %}
                                {% if "python" in tech.category %}
                                <a target="_new" href="{{ tech.url }}">{{ tech.name }}</a> {{ tech.version }}<br/>
                                {% endif %}
                            {% endfor %}
                        </div>
                    </div>

                    <br/><br/>

                    <div class="row">
                        <div class="col">
                            <h5 class="modal-title" id="viewerInfoLabel">R</h5>
                            <hr/>
                            {% for tech in TECHNOLOGIES %}
                                {% if "R" in tech.category %}
                                <a target="_new" href="{{ tech.url }}">{{ tech.name }}</a> {{ tech.version }}<br/>
                                {% endif %}
                            {% endfor %}
                        </div>
                    </div>

                    <br/><br/>

                    <div class="row">
                        <div class="col">
                            <h5 class="modal-title" id="viewerInfoLabel">Docker</h5>
                            <hr/>
                            {% for tech in TECHNOLOGIES %}
                                {% if "docker" in tech.category %}
                                <a target="_new" href="{{ tech.url }}">{{ tech.name }}</a> {{ tech.version }}<br/>
                                {% endif %}
                            {% endfor %}
                        </div>
                    </div>


                </div>
            </div>
            <!--
            <div class="modal-footer">
            </div>
            //-->
        </div>
    </div>
</div>
<!-- END VIEWER INFO MODAL ============================================== //-->

</body>
{% endblock %}

{% block javascript %}
    <!-- please wait -->
    <script src="{{ url_for('static', filename='vendor/please-wait-0.0.5/please-wait.js') }}"></script>

    <!-- Loading app //-->
    <script>

    HTMLCanvasElement.prototype.getContext = (function(oldGetContextFn) {
        return function(type, attrs) {
            attrs = attrs || {};
            if (type === 'webgl2') {
                attrs.preserveDrawingBuffer = true;
            }
            return oldGetContextFn.call(this, type, attrs);
        };
    }(HTMLCanvasElement.prototype.getContext));


    function downloadGenomeSpy(parentId, imageName, bgRGB = null){
        let downloadLink = document.createElement('a');
        downloadLink.setAttribute('download', imageName);

        let canvas = document.querySelector(`${parentId} canvas`);

        if (bgRGB != null) {
            var canvasBG = document.createElement('canvas');
            canvasBG.id = 'bgCanvas';
            canvasBG.width = canvas.width;
            canvasBG.height = canvas.height;

            var ctx = canvasBG.getContext('2d');
            ctx.fillStyle = `rgba(${bgRGB}, 1.0)`;
            ctx.fillRect(0, 0, canvas.width, canvas.height);
            ctx.drawImage(canvas, 0, 0);

            canvasBG.toBlob(function(blob) {
                let url = URL.createObjectURL(blob);
                downloadLink.setAttribute('href', url);
                downloadLink.click();
            });
        } else {
            canvas.toBlob(function(blob) {
                let url = URL.createObjectURL(blob);
                downloadLink.setAttribute('href', url);
                downloadLink.click();
            });
        }
    }

    window.loading_screen = window.pleaseWait({
        logo: '',
        backgroundColor: '#000000',
        loadingHtml: '<div id="loader-wrapper"><div id="loader"></div><div class="loader-section section-left"></div><div class="loader-section section-right"></div></div>'
    });
    </script>

    <script src="//cdn.jsdelivr.net/npm/bootstrap-select@1.13.7/dist/js/bootstrap-select.min.js"></script>

    <!-- ENSIMPL API -->
    <script src="//churchilllab.jax.org/ensimpl/js/ensimpl.js"></script>

    <!-- Highcharts/Highstock //-->
    {% if debug %}
    <script src="//code.highcharts.com/11.4/highcharts.src.js"></script>
    <!--
    <script src="//code.highcharts.com/11.4/modules/debugger.js"></script>
    //-->
    {% else %}
    <script src="//code.highcharts.com/11.4/highcharts.js"></script>
    {% endif %}
    <script src="//code.highcharts.com/11.4/highcharts-more.js"></script>
    <script src="//code.highcharts.com/11.4/modules/boost.js"></script>
    <script src="//code.highcharts.com/11.4/modules/exporting.js"></script>
    <script src="//code.highcharts.com/11.4/modules/offline-exporting.js"></script>

    <!-- DataTables //-->
    {% if debug %}
    <script src="//cdn.datatables.net/1.11.5/js/jquery.dataTables.js"></script>
    <script src="//cdn.datatables.net/1.11.5/js/dataTables.bootstrap4.js"></script>
    {% else %}
    <script src="//cdn.datatables.net/1.11.5/js/jquery.dataTables.min.js"></script>
    <script src="//cdn.datatables.net/1.11.5/js/dataTables.bootstrap4.min.js"></script>
    {% endif %}

    <!-- D3 //-->
    <script src="//d3js.org/d3.v3.min.js"></script>
    <script src="//d3js.org/d3-queue.v3.min.js"></script>
    <script src="//d3js.org/d3-request.v1.min.js"></script>

    <!-- Genome Spy //-->
    <script src="//unpkg.com/@genome-spy/core@0.19.0/dist/index.js"></script>

    <!-- Font Awesome, user: matt.vincent@jax.org //-->
    <script src="//kit.fontawesome.com/56d0cfd040.js" crossorigin="anonymous"></script>

    <!-- Extra //-->
    <script src="//cdn.jsdelivr.net/npm/sweetalert2@11/dist/sweetalert2.all.min.js"></script>

    <script src="{{ url_for('static', filename='vendor/dynatables-0.3.1/jquery.dynatable.js') }}"></script>
    <script src="{{ url_for('static', filename='vendor/bootstrap-multiselect-0.9.15/dist/js/bootstrap-multiselect.js') }}"></script>
    <script src="{{ url_for('static', filename='js/utilities.js') }}"></script>

    <script>

    {% if debug %}
    var logDebug = console.log.bind(window.console);
    {% else %}
    var logDebug = function(){};
    {% endif %}
    var logInfo = console.info.bind(window.console);
    var logError = console.error.bind(window.console);

    /**
     * Create a global getSVG method that takes an array of charts as an argument. The SVG is returned as an argument in the callback.
     */
    Highcharts.getSVG = function (charts, options, callback) {
        console.log('getSVG options=', options);
        var svgArr = [],
            top = 0,
            width = 0,
            addSVG = function (svgres) {
                // Grab width/height from exported chart
                let svgWidth = svgres.match(
                    /^<svg[^>]*width\s*=\s*\"?(\d+)\"?[^>]*>/
                );
                svgWidth = +svgWidth[1];

                let svgHeight = svgres.match(
                    /^<svg[^>]*height\s*=\s*\"?(\d+)\"?[^>]*>/
                );
                svgHeight = +svgHeight[1];

                // Offset the position of this chart in the final SVG
                let svg = svgres.replace('<svg', '<g transform="translate(0,' + top + ')" ');
                svg = svg.replace('</svg>', '</g>');
                top += svgHeight;
                width = Math.max(width, svgWidth);
                svgArr.push(svg);
            },
            exportChart = function (i) {
                if (i === charts.length) {
                    return callback('<svg height="' + top + '" width="' + width +
                      '" version="1.1" xmlns="http://www.w3.org/2000/svg">' + svgArr.join('') + '</svg>');
                }
                charts[i].getSVGForLocalExport(options, {}, function () {
                    logDebug("Failed to get SVG");
                }, function (svg) {
                    addSVG(svg);
                    return exportChart(i + 1); // Export next only when this SVG is received
                });
            };
        exportChart(0);
    };

    /**
     * Create a global exportCharts method that takes an array of charts as an argument,
     * and exporting options as the second argument
     */
    Highcharts.exportCharts = function (charts, options) {
        logDebug('options a', options);
        options = Highcharts.merge(Highcharts.getOptions().exporting, options);
        logDebug('options b', options);

        // Get SVG asynchronously and then download the resulting SVG
        Highcharts.getSVG(charts, options, function (svg) {

            Highcharts.downloadSVGLocal(svg, options, function (a,b,c) {
                logDebug('downloadSVGLocal',options);
                logDebug("Failed to export on client side");
                logDebug(a,b,c);
            });
        });
    };

    /**
     * Add a button to the Highcharts export menu.
     * @param {string} mytext - title for button
     * @param {function} myfunction - function to call
     * @return {Object} buttons for highchart export
     */
    function appendExportButton(mytext, myfunction){
        // get default Highchart Export buttons
        let defaultButtons = Highcharts.getOptions().exporting.buttons;
        let myButtons = $.extend(true, {}, defaultButtons);

        myButtons.contextButton.menuItems.push({
            text: mytext,
            onclick: myfunction
        });

        return {
            buttons: myButtons
        };
    }

    // enhance the original "$.ajax" with a retry mechanism
    $.ajax = (($oldAjax) => {
        // on fail, retry by creating a new Ajax deferred
        function check(a, b, c){
            logDebug('CHECK', a, b, c);
            let shouldRetry = (b !== 'success' && b !== 'parsererror');
            if( shouldRetry && --this.retries > 0 ) {
                setTimeout(() => {
                    $.ajax(this);
                }, this.retryInterval || 100);
            }
        }

        return settings => $oldAjax(settings).always(check);
    })($.ajax);


    /** USAGE:
     * now we can use the "retries" property if we need to retry on fail
    $.ajax({
        type          : 'GET',
        url           : 'http://www.whatever123.gov',
        timeout       : 2000,
        retries       : 3,     //       <-------- Optional
        retryInterval : 2000   //       <-------- Optional
    })
    // "fail" will only be called once, and not for each retry
    .fail(console.warn);
     */


    /**
     *  Extended disable function
     */
    jQuery.fn.extend({
        disable: function(state) {
            return this.each(function() {
                let $this = $(this);
                if($this.is('input, button, textarea, select')) {
                    this.disabled = state;
                } else {
                    $this.toggleClass('disabled', state);
                }
            });
        }
    });

    /**
     * Download data.
     *
     * From d3.js documentation:
     *
     * When a task completes, it must call the provided callback.
     *
     * The first argument to the callback should be null if the task is
     * successful, or the error if the task failed.
     *
     * The optional second argument to the callback is the return value of the task.
     * (To return multiple values from a single callback, wrap the results in an object or array.)
     *
     * @param {string} URL - the URL to download data
     * @param {string} description - descriptive text
     * @param {function} callback - the callback function
     */
    function downloadData(URL, description, callback) {
        logDebug('Downloading: ', URL);

        $.ajax({
            url: URL,
            method: 'GET',
        }).done(function(data, textStatus, jqXHR) {
            logInfo(`Download of ${description}`);
            logDebug(`Download of ${description}:`, data);

            data['_url'] = URL;
            data['_description'] = description;

            callback(null, data);
        }).fail(function(jqXHR, textStatus, errorThrown) {
            logError(description, textStatus);
            logError(description, textStatus, errorThrown);
            callback(errorThrown, null);
        });
    }

    function sortObj(obj) {
        return Object.keys(obj).sort().reduce(function (result, key) {
            result[key] = obj[key];
            return result;
            },
            {}
        );
    }

    function buildEnvInfo(envInfo) {
        let htmlBody = '<table id="envInfoTable" class="table table-striped table-bordered table-sm table-sm-text">';
        htmlBody += '<thead><th>Filename</th><th>Size</th><th>Elements</th></thead>';
        $.each(envInfo, function(idx, value) {
            logDebug('value=', value);
            let fileSize = niceBytes(value.fileSize);
            htmlBody += `
                <tr>
                    <td data-order="${value.fileName}">
                        <a href="#" fileName="${value.fileName}">${value.fileName}</a>
                    </td>
                    <td>
                        ${fileSize}
                    </td>
                    <td>
            `;

            if(typeof value.elements === 'string') {
                htmlBody += value.elements;
            } else {
                htmlBody += value.elements.join('<br>');
            }

            htmlBody += '</td></tr>';
        });

        htmlBody += '</table>';
        $('#envInfoDiv').html(htmlBody);

        $('#envInfoTable').DataTable({
            "info":   false,
            "filter": false,
            "paging": false
        });

        $('#envInfoTable a').click(function(evt) {
            evt.preventDefault();
            let that = $(this);
            let f = that.attr('fileName');
            window.location = "{{ url_for('page.dl') }}" + `?fileName=${f}`;
            return false;
        });
    }

    QTL = function() {
        var g = {};

        function init() {
            /*
             * Global data
             */
            g.DATASETS = null;         // hash of all datasets
            g.ENSEMBL_RELEASE = null;  // Ensembl release for Ensimpl API
            g.SPECIES_ID = 'Mm';       // TODO: do not hard code species
            g.ENSIMPL = null;          // Ensimple API
            g.CHROMOSOMES = null;      // all chromosome information

            g.PLOT_COLORS = [
                '#e6194b',
                '#4363d8',
                '#3cb44b',
                '#ffe119',
                '#f58231',
                '#911eb4',
                '#46f0f0',
                '#f032e6',
                '#bcf60c',
                '#fabebe',
                '#008080',
                '#e6beff',
                '#9a6324',
                '#fffac8',
                '#800000',
                '#aaffc3',
                '#808000',
                '#ffd8b1',
                '#000075',
                '#808080'
            ];
            /*
                "#3d7cb9",
                "#c50004",
                "#38be2f",
                "#8085e9",
                "#91e8e1",
                "#f7a35c",
                "#f15c80",
                "#deca00",
                "#2b908f",
                "#434348"
            */

            /*
             * current selected information
             */
            g.dataSetID = null;        // selected dataset id
            g.geneID = null;           // selected gene id
            g.proteinID = null;        // selected protein id
            g.uniprotID = null;        // selected uniprot id
            g.phenotypeID = null;      // selected phenotype id
            g.ptmID = null;            // selected ptm id
            g.peptideID = null;        // selected peptide id
            g.gene = null;             // selected gene information

            /*
             * Interactive stored data
             */
            g.LODAdditiveMax = null;
            g.LODThreshold = null;           // threshold value of LOD scores
            g.LODChartData = null;
            g.interactiveCovariates = null;  // list of covariates for current data set
            g.phenoDynaTable = null;
            g.expressionData = null;
            g.searchTable = null;
            g.orderCount = null;
            g.secondaryPlotMarkerID = null;
            g.secondaryPlotChromosome = null;
            g.secondaryPlotLocation = null;
            g.secondaryPlotCovar = null;

            /*
             * Chart variables
             */
            g.chartPeaks = null;
            g.chartLOD = null;
            g.chartLODCovariateFull = null;
            g.chartLODCovariateDiff = null;
            g.chartMediation = null;
            g.chartEffect = null;
            g.chartSNPAssoc = null;

            // for tasks
            g.runningTask = false;

            g.PROTOCOL = window.location.protocol;

            d3.queue()
                .defer(downloadEnvInfo)
                .defer(downloadDataSets)
                .await(function(error, envInfo, data) {
                    logDebug('envInfo=', envInfo);
                    logDebug('dataSetsData=', data);

                    if (error) {
                        logError(error);
                        {% if not debug -%}
                        // ?debug=1 will prevent forwarding on error
                        window.location = '{{ url_for('page.error') }}';
                        {%- endif %}
                        return;
                    }

                    if ((data.result === undefined) || (data.result.datasets.length === 0)) {
                        logError('data=', data);
                        {% if not debug -%}
                        // ?debug=1 will prevent forwarding on error
                        window.location = '{{ url_for('page.error') }}';
                        {%- endif %}
                        return;
                    }

                    g.DATASETS = {};

                    buildEnvInfo(envInfo.result);

                    let ensembl_release = null;
                    if ('ensembl.version' in data.result) {
                        ensembl_release  = data.result['ensembl.version'];
                    }
                    if ('ensembl_version' in data.result) {
                        ensembl_release  = data.result['ensembl_version'];
                    }
                    // the new way is RELEASE, replacing VERSION
                    if ('ensembl.release' in data.result) {
                        ensembl_release  = data.result['ensembl.release'];
                    }
                    if ('ensembl_release' in data.result) {
                        ensembl_release  = data.result['ensembl_release'];
                    }

                    // to display dropdown menu by alphabetical name
                    let datasets_dropdown = {};

                    $.each(data.result.datasets, function(dsi, ds) {
                        ds.gene_ids = {};
                        ds.protein_ids = {};
                        ds.phos_ids = {};
                        ds.ptms = {};
                        ds.peptides = {};
                        ds.phenotypes = {};
                        ds.uniprot_ids = {};

                        if ('ensembl_version' in ds) {
                            ds.ensembl_release  = ds['ensembl_version'];
                        } else if ('ensembl_release' in ds) {
                            ds.ensembl_release  = ds['ensembl_release'];
                        } else {
                            ds.ensembl_release = ensembl_release;
                        }

                        // TODO: Fix hardcoding
                        ds.ensembl_species = 'Mm';

                        // change the case so we don't have to throughout
                        ds.datatype = ds.datatype.toLowerCase();
                        logDebug('ds.datatype=', ds.datatype);

                        if (ds.datatype === 'mrna') {
                            // convert array to object (hash)
                            // ds.gene_ids = dictionary with just a 1
                            $.each(ds.annotations, function(index, element) {
                                ds.gene_ids[element['gene_id']] = 1;
                            });
                        } else if (ds.datatype === 'protein') {
                            // ds.gene_ids = dictionary with id and array of protein_ids
                            // ds.protein_ids = dictionary with id mapped to gene id
                            $.each(ds.annotations, function(index, element) {
                                if (!(element['gene_id'] in ds.gene_ids)) {
                                    ds.gene_ids[element['gene_id']] = {
                                        'id': element['gene_id'],
                                        'protein_ids':[]
                                    };
                                }
                                ds.protein_ids[element['protein_id']] = element['gene_id'];
                                ds.gene_ids[element['gene_id']].protein_ids.push(element['protein_id']);
                            });
                        } else if (ds.datatype === 'protein_uniprot') {
                            $.each(ds.annotations, function(index, element) {
                                ds.uniprot_ids[element['uniprot_id']] = element;
                            });
                        } else if (ds.datatype === 'phos') {
                            // ds.gene_ids = dictionary with id and array of protein ids
                            // ds.protein_ids = dictionary with id and array of phos_ids and gene id
                            // ds.phos_ids = dictionary with id and gene_id and protein_id
                            $.each(ds.annotations, function(index, element) {
                                if (!(element['gene_id'] in ds.gene_ids)) {
                                    ds.gene_ids[element['gene_id']] = {
                                        'id': element['gene_id'],
                                        'protein_ids':[]
                                    };
                                }

                                if (!(element['protein_id'] in ds.protein_ids)) {
                                    ds.protein_ids[element['protein_id']] = {
                                        'id': element['protein_id'],
                                        'gene_id': element['gene_id'],
                                        'phos_ids':[]
                                    };

                                    ds.gene_ids[element['gene_id']].protein_ids.push(element['protein_id']);
                                }

                                ds.phos_ids[element['phos_id']] = {
                                    'gene_id': element['gene_id'],
                                    'protein_id': element['protein_id']
                                }

                                ds.protein_ids[element['protein_id']].phos_ids.push(element['phos_id']);

                            });
                        } else if (ds.datatype === 'peptide') {
                            logDebug(ds.id);
                            //logDebug(ds.annotations);

                            $.each(ds.annotations, function(index, element) {
                                ds.peptides[element['peptide_id']] = element;
                            });
                        } else if (ds.datatype === 'ptm') {
                            logDebug(ds.id);
                            //logDebug(ds.annotations);

                            $.each(ds.annotations, function(index, element) {
                                ds.ptms[element['ptm_id']] = element;
                            });
                        } else if ((ds.datatype === 'pheno') || (ds.datatype === 'phenotype')) {
                            ds.datatype = 'pheno';
                            logDebug(ds.id);
                            //logDebug(ds.annotations);

                            $.each(ds.annotations, function(index, element) {
                                ds.phenotypes[element['data_name']] = element;
                            });
                        } else {
                            // TODO: handle error
                            logError('ERROR in dataset data:', ds);
                        }

                        // don't need the extra annotations laying around
                        delete ds.annotations;

                        datasets_dropdown[ds['display_name']] = ds.id;

                        g.DATASETS[ds.id] = ds;
                    });

                    datasets_dropdown = sortObj(datasets_dropdown);

                    let requestedDataSetID = '{{ datasetid }}';

                    if (g.DATASETS[requestedDataSetID] !== undefined) {
                        g.dataSetID = requestedDataSetID;
                    } else {
                        g.dataSetID = datasets_dropdown[Object.keys(datasets_dropdown)[0]];
                    }

                    // the above if is working, but we need to think through all of
                    // the other ramifications of setting this dataset

                    configureCovarInfo(true);

                    logDebug('data.result.datasets.length=',data.result.datasets.length);

                    if (data.result.datasets.length === 1) {
                        // just show a menu item
                        let html = `<ul class="navbar-nav navbar-dark mr-auto">
                                        <li class="nav-item active" style="color:white">
                                            ${g.DATASETS[g.dataSetID]['display_name']}
                                        </li>
                                        <li class="nav-item">
                                            <a class="font-italic text-primary nav-link" id="downloadInformation" data-toggle="modal" data-target="#downloadModal"><i class="fa-solid fa-download"></i> Download Data</a>
                                        </li>
                                    </ul>`;

                        $('#dataSetSelectMenu').append(html);
                        $('#correlationDatasetSelection').html(`${g.DATASETS[g.dataSetID]['display_name']}`);
                        $('#correlationDatasetSelection').append(`<input type="hidden" id="correlationDatasetSelect" value="${g.DATASETS[g.dataSetID].id}"/>`);

                    } else if (data.result.datasets.length > 1) {
                        // show multiple menu items
                        let html = `<button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarTogglerQTL" aria-controls="navbarTogglerQTL" aria-expanded="false" aria-label="Toggle navigation">
                                        <span class="navbar-toggler-icon"></span>
                                    </button>
                                    <div class="collapse navbar-dark navbar-collapse" id="navbarTogglerQTL">
                                        <span class="navbar-text text-white pr-2">
                                            {{ config.NAV_DATASET|safe }} <i class="fas fa-long-arrow-alt-right"></i>
                                        </span>
                                        <form class="form-inline">
                                            <select data-style="btn-sm btn-light" class="selectpicker" id="datasetSelect">
                                            </select>
                                        </form>
                                        <ul class="navbar-nav mr-auto">
                                            <li class="nav-item">
                                                <a class="font-italic text-primary nav-link" id="downloadInformation" data-toggle="modal" data-target="#downloadModal"><i class="fa-solid fa-download"></i> Download Data</a>
                                            </li>
                                        </ul>
                                     </div>
                                    `;

                        $('#dataSetSelectMenu').append(html);

                        // display datasets by sorted name
                        for (let d in datasets_dropdown) {
                            let _id = datasets_dropdown[d];
                            if (g.dataSetID === _id) {
                                $('#datasetSelect').append(`<option value="${_id}" selected>${d}</option>`);
                            } else {
                                $('#datasetSelect').append(`<option value="${_id}">${d}</option>`);
                            }
                        }

                        $('#datasetSelect').selectpicker();

                        $('#datasetSelect').on('changed.bs.select', function (e, clickedIndex, isSelected, previousValue) {
                            let selected = $(this).find(':selected').val();
                            switchDataSet(selected);
                        });

                        // show multiple menu items
                        let htmlNew = `<select id="correlationDatasetSelect" data-width="100%" data-style="btn-secondary btn-sm" class="selectpicker">`;

                        for (let d in datasets_dropdown) {
                            let _id = datasets_dropdown[d];
                            if (g.dataSetID === _id) {
                                htmlNew += `<option value="${_id}" selected>${d}</option>`;
                            } else {
                                htmlNew += `<option value="${_id}">${d}</option>`;
                            }
                        }
                        htmlNew += '</select>';

                        $('#correlationDatasetSelection').html(htmlNew);

                        $('#correlationDatasetSelect').selectpicker();

                        $('#correlationDatasetSelect').on('changed.bs.select', function (e, clickedIndex, isSelected, previousValue) {
                            let selected = $(this).find(':selected').val();
                            let interactiveCovariate = $('#interactiveCovarCorrelation').find(':selected').val();

                            if ((selected == undefined) ||
                                (selected == null) ||
                                (selected === 'none')) {
                                selected = g.dataSetID;
                            }

                            switchCorrelationDataSet(selected, interactiveCovariate);
                        });
                    }

                    g.ENSIMPL = new ensimpl({
                        greedy: true,
                        search_limit: 100,
                        search_release: g.DATASETS[g.dataSetID].ensembl_release,
                        search_species: g.DATASETS[g.dataSetID].ensembl_species,
                    });

                    d3.queue()
                        .defer(downloadChromosomes, g.DATASETS[g.dataSetID].ensembl_release, g.DATASETS[g.dataSetID].ensembl_species, g.DATASETS[g.dataSetID].id)
                        .defer(downloadLODPeaks, g.dataSetID)
                        .await(function(errorPeaks, chromosomes, lodPeaks) {
                            g.DATASETS[g.dataSetID].chromosomes = {
                                'idx': [],
                                'chr': {},
                                'contigs':[]
                            };


                            g.DATASETS[g.dataSetID].ensembl = {
                                'species': chromosomes.meta.species,
                                'assembly': chromosomes.meta.assembly,
                                'assembly_patch': chromosomes.meta.assembly_patch,
                                'release': chromosomes.meta.release,
                                'url': chromosomes.meta.url
                            };

                            let configChromosomes = {{ config.CHROMOSOMES|safe }};
                            let start = 1;
                            let end = 0;

                            $.each(chromosomes.chromosomes, function(index, element) {
                                if (configChromosomes.indexOf(element.chromosome) !== -1) {
                                    let chrom = element;

                                    start = end + 1;
                                    end += chrom.length;

                                    chrom.start = start;
                                    chrom.end = end;
                                    chrom.mid = chrom.start + Math.ceil((chrom.end - chrom.start) / 2);

                                    g.DATASETS[g.dataSetID].chromosomes.idx.push(chrom);
                                    g.DATASETS[g.dataSetID].chromosomes.chr[element.chromosome] = chrom;
                                    g.DATASETS[g.dataSetID].chromosomes.contigs.push({
                                        "name": chrom.chromosome,
                                        "size": chrom.length,
                                    });
                                }
                            });

                            g.DATASETS[g.dataSetID].chromosomes.totalLength = end;
                            g.DATASETS[g.dataSetID].lodpeaks = lodPeaks.result;

                            setDataSet(g.dataSetID, true);

                            logDebug('Hiding loading screen');
                            window.loading_screen.finish();
                        });

                    // download all the other load peaks from all the datasets except the current one
                    // we do this in the background to gain some performance
                    let q = d3.queue();
                    logDebug('Downloading the peaks and chromosomes...');

                    $.each(g.DATASETS, function(key, value) {
                        if (key !== g.dataSetID) {
                            q.defer(downloadLODPeaks, key);
                            q.defer(downloadChromosomes, g.DATASETS[key].ensembl_release, g.DATASETS[key].ensembl_species, key);
                        }
                    });

                    q.awaitAll(function(errorAll, results) {
                        // TODO: handle error
                        $.each(results, function(key, value) {
                            if ('chromosomes' in value) {
                                let _id = value['_description'].split(' ')[1];
                                g.DATASETS[_id].chromosomes = {
                                    'idx': [],
                                    'chr': {},
                                    'contigs': []
                                };

                                g.DATASETS[_id].ensembl = {
                                    'species': value.meta.species,
                                    'assembly': value.meta.assembly,
                                    'assembly_patch': value.meta.assembly_patch,
                                    'release': value.meta.release,
                                    'url': value.meta.url
                                };

                                let configChromosomes = {{ config.CHROMOSOMES|safe }};
                                let start = 1;
                                let end = 0;

                                $.each(value.chromosomes, function(index, element) {
                                    if (configChromosomes.indexOf(element.chromosome) !== -1) {
                                        let chrom = element;

                                        start = end + 1;
                                        end += chrom.length;

                                        chrom.start = start;
                                        chrom.end = end;
                                        chrom.mid = chrom.start + Math.ceil((chrom.end - chrom.start) / 2);

                                        g.DATASETS[_id].chromosomes.idx.push(chrom);
                                        g.DATASETS[_id].chromosomes.chr[element.chromosome] = chrom;
                                        g.DATASETS[_id].chromosomes.contigs.push({
                                            "name": chrom.chromosome,
                                            "size": chrom.length,
                                        });
                                    }
                                });

                                g.DATASETS[_id].chromosomes.totalLength = end;
                                logDebug('g.DATASETS[_id].chromosomes=', g.DATASETS[_id].chromosomes);

                            } else {
                                g.DATASETS[value.parameters.dataset].lodpeaks = value.result;
                            }
                        });

                        logDebug('end awaitall');
                        // this is where the final stuff is
                        console.log('{{ search_term }}');
                        if (($('#navPheno').is(':hidden')) && ($('#searchTerm').val() !== '')) {
                            console.log($('#searchTerm').val());
                            $('#btnGo').click();
                        }


                        if (($('#navGene').is(':hidden')) && ($('#searchTermPheno').val() !== '')) {
                            console.log($('#searchTermPheno').val());
                            $("#searchTermPheno").trigger('keyup');


                            if ($('#phenotypesTable tbody a').length === 1) {
                                $('#phenotypesTable tbody a')[0].click();
                            }

                        }
                    });

                    logDebug('end await');
                });


                logDebug('end init');
            }

        /**
         * Display the gene information.
         * @param {Object} geneData - gene information
         */
        function displayGeneData(geneData) {
            // TODO: link to database based upon ensembl version?

            for(let key in geneData) {
                geneData = geneData[key];
            }

            let startFormat = geneData.start.toLocaleString();
            let endFormat = geneData.end.toLocaleString();
            let strand = geneData.strand;
            let htmlHeader = geneData.id;
            htmlHeader += ` <em>${geneData.symbol}</em>`;

            let htmlBody = '<table class="table table-sm">';
            htmlBody += `<tr><td class="border-0"><strong>ID</strong></td><td class="border-0">${geneData.id}`;
            htmlBody += '<br>';
            htmlBody += `<small><a target="_newwin" href="http://www.ensembl.org/id/${geneData.id}">Ensembl <i class="fas fa-external-link-alt" aria-hidden="true"></i></small></a>`;

            $.each(geneData.external_ids, function(idx, elem) {
                let url = '';
                let urlDesc = '';

                if (elem.db === 'EntrezGene') {
                    url = `https://www.ncbi.nlm.nih.gov/gene/?term=${elem.db_id}`;
                    urlDesc = `NCBI Gene: ${elem.db_id}`;
                } else if (elem.db === 'MGI') {
                    url = `http://www.informatics.jax.org/marker/${elem.db_id}`;
                    urlDesc = `${elem.db_id}`;
                } else if (elem.db === 'Uniprot_gn') {
                    url = `https://www.uniprot.org/uniprot/${elem.db_id}`;
                    urlDesc = `Uniprot: ${elem.db_id}`;
                }

                if (url !== '') {
                    htmlBody += ` <small><a target="_newwin" href="${url}">${urlDesc} <i class="fas fa-external-link-alt" aria-hidden="true"></i></small></a>`;
                }
            });

            htmlBody += '</td></tr>';
            htmlBody += `<tr><td><strong>Symbol</strong></td><td>${geneData.symbol}</td></tr>`;
            htmlBody += `<tr><td><strong>Location</strong></td><td>${geneData.chromosome}:${startFormat}-${endFormat} (${strand})</td></tr>`;

            if (geneData.synonyms !== undefined) {
                let synonyms = geneData.synonyms.join(', ');
                htmlBody += `<tr><td><strong>Synonyms</strong></td><td>${synonyms}</td></tr>`;
            }

            if (geneData.name) {
                htmlBody += `<tr><td><strong>Description</strong></td><td>${geneData.name}</td></tr>`;
            }

           /*
            let ensembl = g.DATASETS[g.dataSetID].ensembl;
            htmlBody += `<tr><td><strong>Source</strong></td><td><a target="_newwin" href="${ensembl.url}">Ensembl ${ensembl.release}<i class="fas fa-external-link-alt" aria-hidden="true"></i>, ${ensembl.assembl_patch}</td></tr>`;
            htmlBody += '</table>';
            */

            $('#div_current_item_information_header').html(htmlHeader);
            $('#div_current_item_information_body').html(htmlBody);
        }

        /**
         * Display the pheno information.
         * @param {Object} geneData - gene information
         */
        function displayPhenoData(phenoID) {
            let pheno = g.DATASETS[g.dataSetID].phenotypes[phenoID];
            $('#div_current_item_information_header').html(phenoID);
            $('#div_current_item_information_body').html(pheno.desc);
        }

        /**
         * Display the pheno information.
         * @param {Object} geneData - gene information
         */
        function displayPTMPeptideData(phenoID) {
            //let pheno = g.DATASETS[g.dataSetID].phenotypes[phenoID];
            //$('#div_current_item_information_header').html('');
            //$('#div_current_item_information_body').html('');
        }

        function exportLODPeaks(id, peaks) {
            let covar = $('#interactiveCovarPeaks').val();
            let csvContent = '';

            if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                csvContent = '"gene_id","symbol","gene_chrom","gene_mid","marker_id","marker_chrom","marker_position","lod"';
                if (peaks[0].length === 16) {
                    csvContent += ',"A","B","C","D","E","F","G","H"';
                }
                csvContent += '\n';
                // [marker_id, chrom, position, gene_id, symbol, gene_chrom, gene_mid, lod
                //     0         1       2        3       4         5          6        7
                // ["1_100007442","1",100.0074,"ENSMUSG00000028028","Alpk1","3",127.7255,6.1147]
                $.each(peaks, function(k, v) {
                    let line = `"${v[3]}","${v[4]}","${v[5]}",${v[6]},"${v[0]}","${v[1]}",${v[2]},${v[7]}`;
                    if (peaks[0].length === 16) {
                        line += `,${v[8]},${v[9]},${v[10]},${v[11]},${v[12]},${v[13]},${v[14]},${v[15]}`;
                    }
                    line += '\n';

                    csvContent += line;
                });
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                csvContent = '"protein_id","gene_id","symbol","gene_chrom","gene_mid","marker_id","marker_chrom","marker_position","lod"';
                if (peaks[0].length === 17) {
                    csvContent += ',"A","B","C","D","E","F","G","H"';
                }
                csvContent += '\n';
                // [marker_id, chrom, position, protein_id, gene_id, symbol, gene_chrom, gene_mid, lod
                //     0         1       2          3          4        5         6        7        8
                // ["1_106129060","1",106.12906,"ENSMUSP00000108356","ENSMUSG00000009907","Vps4b","1",106.7804,8.2928352859],
                $.each(peaks, function(k, v) {
                    let line = `"${v[3]}","${v[4]}","${v[5]}","${v[6]}",${v[7]},"${v[0]}","${v[1]}",${v[2]},${v[8]}`;
                    if (peaks[0].length === 17) {
                        line += `,${v[9]},${v[10]},${v[11]},${v[12]},${v[13]},${v[14]},${v[15]},${v[16]}`;
                    }
                    line += '\n';
                    csvContent += line;
                });
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                csvContent = '"uniprot_id","protein_id","gene_id","symbol","gene_chrom","gene_mid","marker_id","marker_chrom","marker_position","lod"';
                if (peaks[0].length === 18) {
                    csvContent += ',"A","B","C","D","E","F","G","H"';
                }
                csvContent += '\n';
                // [marker_id, chrom, position, uniprot_id, protein_id, gene_id, symbol, gene_chrom, gene_mid, lod
                //     0         1       2          3          4        5         6        7        8        9
                // ["1_106129060","1",106.12906,"UNIPROT_ID","ENSMUSP00000108356","ENSMUSG00000009907","Vps4b","1",106.7804,8.2928352859],
                $.each(peaks, function(k, v) {
                    let line = `"${v[3]}","${v[4]}","${v[5]}","${v[6]}","${v[7]}",${v[8]},"${v[0]}","${v[1]}",${v[2]},${v[9]}`;
                    if (peaks[0].length === 18) {
                        line += `,${v[10]},${v[11]},${v[12]},${v[13]},${v[14]},${v[15]},${v[16]},${v[17]}`;
                    }
                    line += '\n';
                    csvContent += line;
                });
            } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                csvContent = '"phos_id","protein_id","gene_id","symbol","gene_chrom","gene_mid","marker_id","marker_chrom","marker_position","lod"';
                if (peaks[0].length === 18) {
                    csvContent += ',"A","B","C","D","E","F","G","H"';
                }
                csvContent += '\n';
                // [marker_id, chrom, position, phos_id, protein_id, gene_id, symbol, gene_chrom, gene_mid, lod
                //     0         1       2          3          4        5         6        7        8        9
                // ["1_106129060","1",106.12906,"PHOS_ID","ENSMUSP00000108356","ENSMUSG00000009907","Vps4b","1",106.7804,8.2928352859],
                $.each(peaks, function(k, v) {
                    let line = `"${v[3]}","${v[4]}","${v[5]}","${v[6]}","${v[7]}",${v[8]},"${v[0]}","${v[1]}",${v[2]},${v[9]}`;
                    if (peaks[0].length === 18) {
                        line += `,${v[10]},${v[11]},${v[12]},${v[13]},${v[14]},${v[15]},${v[16]},${v[17]}`;
                    }
                    line += '\n';
                    csvContent += line;
                });
            } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                csvContent = '"marker_id","marker_chrom","marker_position","phenotype","lod"';
                if (peaks[0].length === 15) {
                    csvContent += ',"A","B","C","D","E","F","G","H"';
                }
                csvContent += '\n';
                // ["5_137006393","5",137.0064,"MEcyan","MEcyan","MEcyan",9.1161],[
                //       0         1     2        3        4        5        6
                $.each(peaks, function(k, v) {
                    let line = `"${v[0]}","${v[1]}",${v[2]},"${v[4]}",${v[6]}`;
                    if (peaks[0].length === 16) {
                        line += `,${v[7]},${v[8]},${v[9]},${v[10]},${v[11]},${v[12]},${v[13]},${v[14]}`;
                    }
                    line += '\n';
                    csvContent += line;
                });
            }

            downloadCSV(csvContent, `${id}_LODPEAKS.csv`, 'text/csv;encoding:utf-8');
        }

        /**
         * Plot LOD Peaks as a scatter plot.
         */
        function plotLODPeaks() {
            logDebug('plotLODPeaks()');
            let covar = $('#interactiveCovarPeaks').val();
            logDebug('covar=', covar);

            let data = [];
            let tracks = null;
            let handlers = null;
            let minLOD = Infinity;
            let maxLOD = -Infinity;
            let html = `<div class="row">
                            <div class="col">
                                <div id="plotLodPeaks" style="width: 100%"></div>
                            </div>
                        </div>`;

            g.LODThreshold = null;

            let trackPeakLOD = null;

            let chrs = g.DATASETS[g.dataSetID].chromosomes.chr;

            logDebug('chrs=', chrs);

            $("#plotLodPeaks").html('');
            $('#plotLodPeaks').height('400');

            if ((g.DATASETS[g.dataSetID].datatype === 'mrna') ||
                (g.DATASETS[g.dataSetID].datatype === 'protein') ||
                (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') ||
                (g.DATASETS[g.dataSetID].datatype === 'phos') ||
                (g.DATASETS[g.dataSetID].datatype === 'peptide') ||
                (g.DATASETS[g.dataSetID].datatype === 'ptm')) {
                $('#plotLodPeaks').width('400');

                //                 0   1      2   3   4        5        6       7
                //mrna:            lod,marker,chr,pos,gene_chr,gene_pos,gene_id,symbol
                //protein:         lod,marker,chr,pos,gene_chr,gene_pos,gene_id,symbol,protein_id
                //protein_uniprot: lod,marker,chr,pos,gene_chr,gene_pos,gene_id,symbol,protein_id,uniprot_id
                //phos:            lod,marker,chr,pos,gene_chr,gene_pos,gene_id,symbol,protein_id,phos_id
                //peptide:         lod,marker,chr,pos,gene_chr,gene_pos,gene_id,symbol,protein_id,peptide_id,uniprot_id
                //ptm:             lod,marker,chr,pos,gene_chr,gene_pos,gene_id,symbol,protein_id,peptide_id,ptm_id,uniprot_id

                let numElements = 8;

                let lodThreshold =  +$('#lodSliderValue').val();
                logDebug('lodThreshold=', lodThreshold);

                $.each(g.DATASETS[g.dataSetID].lodpeaks[covar], function(index, element) {
                    if ((element[2] in chrs) && (element[4] in chrs)) {
                        if ((lodThreshold != null) && (element[0] >= lodThreshold)) {
                            let lod = {
                                lod: element[0],
                                markerID: element[1],
                                markerChr: chrs[element[2]].chromosome,
                                markerPOrig: element[3],
                                markerPos: element[3],
                                geneChr: chrs[element[4]].chromosome,
                                genePos: element[5],
                                genePosOrig: element[5],
                                geneID: element[6],
                                geneSymbol: element[7],
                                covar: covar,
                            };

                            if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                                lod['proteinID'] = element[8];
                                numElements = 9;
                            } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                                lod['proteinID'] = element[8];
                                lod['uniprotID'] = element[9];
                                numElements = 10;
                            } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                                lod['proteinID'] = element[8];
                                lod['phosID'] = element[9];
                                numElements = 10;
                            } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                                lod['proteinID'] = element[8];
                                lod['peptideID'] = element[9];
                                lod['uniprotID'] = element[10];
                                numElements = 11;
                            } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                                lod['proteinID'] = element[8];
                                lod['peptideID'] = element[9];
                                lod['ptmID'] = element[10];
                                lod['uniprotID'] = element[11];
                                numElements = 12;
                            }

                            if (element.length > numElements) {
                                lod['A'] = element[numElements];
                                lod['B'] = element[numElements+1];
                                lod['C'] = element[numElements+2];
                                lod['D'] = element[numElements+3];
                                lod['E'] = element[numElements+4];
                                lod['F'] = element[numElements+5];
                                lod['G'] = element[numElements+6];
                                lod['H'] = element[numElements+7];
                            }

                            data.push(lod);
                        }

                        minLOD = Math.min(minLOD, element[0]);
                        maxLOD = Math.max(maxLOD, element[0]);
                    }
                });

                minLOD = Math.max(0, Math.floor(minLOD) - 1);
                maxLOD = Math.ceil(maxLOD) + 1;

                logDebug('minLOD=', minLOD, 'maxLOD=', maxLOD);
                logDebug('LOD Peaks Plotting:', data);

                trackPeakLOD = {
                    "name": "lod",
                    "layer": [
                        {
                            "title": "Xbound",
                            "data": {
                                "values": g.DATASETS[g.dataSetID].chromosomes.contigs
                            },
                            "mark": {
                                "type": "rule",
                                "tooltip": {
                                    "handler": "nullHandler"
                                },
                            },
                            "encoding": {
                                "x": {
                                    "chrom": "name",
                                    "pos": "size",
                                    "type": "locus"
                                },
                                "color": {
                                    "value": "lightgray"
                                }
                            },
                        },
                        {
                            "title": "Ybound",
                            "data": {
                                "values": g.DATASETS[g.dataSetID].chromosomes.contigs
                            },
                            "mark": {
                                "type": "rule",
                                "tooltip": {
                                    "handler": "nullHandler"
                                },
                            },
                            "encoding": {
                                "y": {
                                    "chrom": "name",
                                    "pos": "size",
                                    "type": "locus"
                                },
                                "color": {
                                    "value": "lightgray"
                                }
                            },
                        },
                        {
                            "data": {
                                "values": data
                            },
                            "title": "LODS",
                            "mark": {
                                "type": "point",
                                "tooltip": {
                                    "handler": "lodPeakGeneTooltipHandler"
                                },
                                "geometricZoomBound": 4,
                                "inwardStroke": true,
                                "stroke": "#233b5e",
                                "strokeWidth": 0.7,
                                "fillOpacity": 0.6,
                                "strokeOpacity": 0.8
                            },
                            "encoding": {
                                "x": {
                                    "chrom": "markerChr",
                                    "pos": "markerPos",
                                    "type": "locus",
                                    "title": "Marker"
                                },
                                "y": {
                                    "chrom": "geneChr",
                                    "pos": "genePos",
                                    "type": "locus",
                                    "title": "Gene"

                                },
                                "color": {
                                    "value": "#7090c0"
                                },
                                "size": {
                                    "value": 100
                                },
                                "opacity": {
                                    "value": 1
                                }
                            }
                        }
                    ]
                };

                tracks = [trackPeakLOD];
                handlers = {
                    nullHandler: nullHandler,
                    lodPeakGeneTooltipHandler: lodPeakGeneTooltipHandler
                };

            } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                // ["5_137006393","5",137.0064,"MEcyan","MEcyan","MEcyan",9.1161],[

                let seriesData = new Set();
                let seriesDataAll = new Set();
                let data = [];
                let rects = [];

                $.each(g.DATASETS[g.dataSetID].chromosomes.contigs, function(k, v) {
                    if ((k % 2) === 1) {
                        rects.push({
                            name: v.name,
                            size: v.size,
                            start: 1,
                            end: v.size
                        });
                    }
                });

                let checkedVals = $("#divLODPeaks").find(":input:checked").map(function () {
                    return this.value;
                }).get();
                console.log(checkedVals);

                $.each(g.DATASETS[g.dataSetID].lodpeaks[covar], function(index, element) {
                    // element[0] = marker_id
                    let markerChr = chrs[element[1]].chromosome;
                    let markerPos = element[2];

                    // lookup the phenotype
                    //logDebug('element=', element);
                    //logDebug('g.datasetID=', g.dataSetID);
                    //logDebug(g.DATASETS[g.dataSetID].phenotypes);

                    let phenotype = g.DATASETS[g.dataSetID].phenotypes[element[3]];

                    //logDebug(phenotype);
                    seriesDataAll.add(phenotype.category);
                    if (checkedVals.includes(phenotype.category)) {
                        seriesData.add(phenotype.category);

                        data.push({
                            lod: element[6],
                            phenoDataName: element[3],
                            phenoShortName: element[4],
                            phenoDescription: element[5],
                            phenoCategory: phenotype.category,
                            markerPosOrig: element[2],
                            markerChr: markerChr,
                            markerPos: markerPos,
                            markerID: element[0],
                            covar: covar,
                            phenotype: true
                        });

                        minLOD = Math.min(minLOD, element[6]);
                        maxLOD = Math.max(maxLOD, element[6]);
                    }
                });

                minLOD = Math.max(0, Math.floor(minLOD) - 1);
                maxLOD = Math.ceil(maxLOD) + 1;

                let axisTicks = calculateTicks(minLOD, maxLOD, 5);

                let axisLines = [];

                $.each(axisTicks, function(k,v) {
                    axisLines.push({
                        axisLOD: v
                    })
                });

                let phenotypeColorDomain = Array.from(seriesData);
                let allCategories = Array.from(seriesDataAll);
                let phenotypeColorRange = [];

                for (let a = 0; a < allCategories.length; a++) {
                    for (let b = 0; b < phenotypeColorDomain.length; b++) {
                        if (allCategories[a] === phenotypeColorDomain[b]) {
                            phenotypeColorRange.push(g.PLOT_COLORS[a]);
                        }
                    }
                }

                logDebug('phenotype lod=', data);
                logDebug('minLOD=', minLOD);
                logDebug('maxLOD=', maxLOD);
                logDebug('phenotypeColorDomain=', phenotypeColorDomain);
                logDebug('phenotypeColorRange=', phenotypeColorRange);

                trackPeakLOD = {
                    "name": "lod",
                    "layer": [
                        {
                            "title": "Xbound",
                            "data": {
                                "values": rects
                            },
                            "mark": {
                                "type": "rect",
                                "tooltip": {
                                    "handler": "nullHandler"
                                },
                            },
                            "encoding": {
                                "x": {
                                    "chrom": "name",
                                    "pos": "start",
                                    "type": "locus",
                                    "title": null
                                },
                                "x2": {
                                    "chrom": "name",
                                    "pos": "end",
                                    "type": "locus",
                                    "title": null
                                },
                                "color": {
                                    "value": "#eeeeee"
                                }
                            },
                        },
                        {
                            "title": "Ybound",
                            "data": {
                                "values": axisLines,
                            },
                            "mark": {
                                "type": "rule",
                                "tooltip": {
                                    "handler": "nullHandler"
                                },
                            },
                            "encoding": {
                                "y": {
                                    "field": "axisLOD",
                                    "type": "quantitative",
                                    "title": null
                                },
                                "color": {
                                    "value": "#e5e5e5"
                                }
                            },
                        },
                        {
                            "data": {
                                "values": data
                            },
                            "title": "LODS",
                            "mark": {
                                "type": "point",
                                "tooltip": {
                                    "handler": "lodPeakPhenotypeTooltipHandler"
                                },
                                "geometricZoomBound": 4,
                                "inwardStroke": true,
                                "stroke": "#222222",
                                "strokeWidth": 0.7,
                                "fillOpacity": 0.9,
                                "strokeOpacity": 0.8
                            },
                            "encoding": {
                                "x": {
                                    "chrom": "markerChr",
                                    "pos": "markerPos",
                                    "type": "locus",
                                    "title": "Marker Location"
                                },
                                "y": {
                                    "field": "lod",
                                    "type": "quantitative",
                                    "title": "LOD",
                                    "scale": { "domain": [minLOD, maxLOD] },
                                    "axis": {
                                      "values": axisTicks
                                    },
                                },
                                "color": {
                                    "type": "nominal",
                                    "field": "phenoCategory",
                                    "scale": {
                                        "domain": phenotypeColorDomain,
                                        "range": phenotypeColorRange,
                                    }
                                },
                                "size": {
                                    "value": 100
                                },
                                "opacity": {
                                    "value": 1
                                },
                            }
                        }
                    ]
                };

                tracks = [trackPeakLOD];
                handlers = {
                    nullHandler: nullHandler,
                    lodPeakPhenotypeTooltipHandler: lodPeakPhenotypeTooltipHandler
                };
            }

            const spec = {
                "genome": {
                    "name": g.DATASETS[g.dataSetID].ensembl_species + " " + g.DATASETS[g.dataSetID].ensembl_release,
                    "contigs": g.DATASETS[g.dataSetID].chromosomes.contigs
                },

                "vconcat": tracks
            };

            logDebug('SPEC=', spec);

            $('#divLODPeaksMenu').html('');

            $('#divLODPeaksMenu').append(`<div class="col text-center font-weight-bold">LOD Peaks
                <div class="dropdown dropleft float-right" id="lodPeaksMenuDropdown">
                  <a class="text-secondary" href="#" role="button" id="peaksDropdownMenuLink" data-toggle="dropdown" aria-expanded="false">
                    <i class="fa-solid fa-bars"></i>
                  </a>

                  <div class="dropdown-menu" aria-labelledby="peaksDropdownMenuLink">
                    <a class="dropdown-item" id="peaksDownloadPNG" style="font-size:11px !important">Download PNG image</a>
                    <a class="dropdown-item" id="peaksDownloadCSV" style="font-size:11px !important">Download CSV</a>
                  </div>
                </div>
            </div>`);

            let n = g.DATASETS[g.dataSetID].display_name;
            $("#peaksDownloadPNG").click(function (event) {
                downloadGenomeSpy('#plotLodPeaks', `${n}_PEAKS.png`, '255,255,255');
            });

            $("#peaksDownloadCSV").click(function (event) {
                exportLODPeaks(n, g.DATASETS[g.dataSetID].lodpeaks[covar]);
            });


            g.chartPeaks = genomeSpyEmbed.embed(
                document.querySelector("#plotLodPeaks"),
                spec,
                {
                    tooltipHandlers: handlers
                }
            ).then(function(api) {
                api.addEventListener("click", function(event) {
                    if (event.datum != null) {
                        let datum = event.datum;

                        if ('geneID' in datum) {
                            logDebug('lodPeak clicked on', datum);
                            logDebug(datum.geneID, datum.proteinID, g.dataSetID, covar);
                            selectGeneProtein(datum.geneID, datum?.proteinID, datum?.phosID, datum?.uniprotID, datum?.ptmID, datum?.peptideID, g.dataSetID, covar);
                        } else if ('phenoDataName' in datum) {
                            logDebug('lodPeak clicked on', datum);
                            logDebug(datum.phenoDataName, g.dataSetID, covar);
                            selectPhenotype(datum.phenoDataName, g.dataSetID, covar);
                        } else {
                            logDebug('DO NOTHING');
                        }
                    }
                });
            });
        }

        function generateLODPeaksHTML() {
            logDebug('generateLODPeaksHTML()');

            if (isEmpty(g.DATASETS[g.dataSetID].lodpeaks)) {
                logDebug('lodspeaks is null, returning');
                return;
            }

            let covar = $('#interactiveCovarPeaks').val();
            logDebug('covar=', covar);


            let minLOD = Infinity;
            let maxLOD = -Infinity;
            let html = `<div class="row">
                            <div class="col">
                                <div id="plotLodPeaks" style="width: 100%"></div>
                            </div>
                        </div>`;

            g.LODThreshold = null;

            if ((g.DATASETS[g.dataSetID].datatype === 'mrna') ||
                (g.DATASETS[g.dataSetID].datatype === 'protein') ||
                (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') ||
                (g.DATASETS[g.dataSetID].datatype === 'phos') ||
                (g.DATASETS[g.dataSetID].datatype === 'ptm') ||
                (g.DATASETS[g.dataSetID].datatype === 'peptide')) {

                let lods = g.DATASETS[g.dataSetID].lodpeaks[covar];

                // a.reduce((a,b)=>a.size>b.size?a:b).size;
                minLOD = lods.reduce((a, b) => a[0] < b[0] ? a : b)[0];
                maxLOD = lods.reduce((a, b) => a[0] > b[0] ? a : b)[0];

                minLOD = Math.max(0, Math.floor(minLOD) - 1);
                maxLOD = Math.ceil(maxLOD) + 1;

                logDebug('minLOD=', minLOD, 'maxLOD=', maxLOD);

                html = html + `
                    <div class="row">
                        <div class="col-sm-auto my-auto margin y auto font-weight-bold">
                            <span class="align-bottom">LOD Threshold: </span>
                        </div>
                        <div class="col-auto">
                            <div class="input-group">
                                <input id="lodSliderValue" type="number" min="${minLOD}" max="${maxLOD}" step=".5" value="${minLOD}" class="form-control" size="6">
                                <span class="input-group-append">
                                    <button id="btnGoLOD" class="btn btn-primary"><i class="fas fa-caret-right"></i></button>
                                </span>
                            </div>
                        </div>
                    </div>
                    <div class="row justify-content-center">
                        <div class="col text-right">
                            <span>${minLOD}</span>
                        </div>
                        <div class="col-8">
                            <input type="range" class="custom-range" value="${minLOD}" min="${minLOD}" max="${maxLOD}" id="lodRange">
                        </div>
                        <div class="col text-left">
                            <span>${maxLOD}</span>
                        </div>
                    </div>
                `;


                $('#divLODPeaks').html(html);

                $('#lodRange').on('change', function (evt) {
                    $('#lodSliderValue').val(evt.target.value);
                    plotLODPeaks(g.DATASETS[g.dataSetID].lodpeaks[covar], evt.target.value);
                });

                $('#btnGoLOD').click(function (evt) {
                    $('#lodRange').val(+$('#lodSliderValue').val());
                    plotLODPeaks(g.DATASETS[g.dataSetID].lodpeaks[covar], +$('#lodSliderValue').val());
                });


                $('#lodSliderValue').bind('keyup input', function (evt) {
                    $('#lodRange').val(+$('#lodSliderValue').val());
                    plotLODPeaks(g.DATASETS[g.dataSetID].lodpeaks[covar], +$('#lodSliderValue').val());
                });
            } else {
                //let checkedVals = $("#divLODPeaks").find(":input:checked").map(function () {
                //    return this.value;
                //}).get();
                //console.log('checkedVals=', checkedVals);

                let seriesData = new Set();

                $.each(g.DATASETS[g.dataSetID].lodpeaks[covar], function(index, element) {
                    let phenotype = g.DATASETS[g.dataSetID].phenotypes[element[3]];

                    //if (checkedVals.includes(phenotype.category)) {
                        seriesData.add(phenotype.category);
                    //}
                });

                let phenotypeColorDomain = Array.from(seriesData);
                let phenotypeColorRange = g.PLOT_COLORS.slice(0, phenotypeColorDomain.length);

                //logDebug('phenotypeColorDomain=', phenotypeColorDomain);
                //logDebug('phenotypeColorRange=', phenotypeColorRange);

                html = html + `
                    <div class="row justify-content-center">
                        <div class="col text-right">
                            <table border="0" cellspacing="0" cellpadding="1" align="center">
                                <tbody>
                `;

            for (c in phenotypeColorDomain) {
                //console.log('c=', c);
                //console.log('phenotypeColorDomain[c]=', phenotypeColorDomain[c]);
                //console.log('phenotypeColorRange[c]=', phenotypeColorRange[c]);

                if (c % 2 === 0) {
                    html += `<tr style="color: black;" align="left">`;
                }

                html += `<td>
                            <div class="form-check">
                                <input class="checkbox-round-${phenotypeColorRange[c].slice(1)}"
                                       type="checkbox"
                                       value="${phenotypeColorDomain[c]}"
                                       id="${phenotypeColorDomain[c]}"
                                       checked>
                                <label class="small form-check-label" for="${phenotypeColorDomain[c]}">
                                    ${phenotypeColorDomain[c]}
                                </label>
                            </div>
                        </td>`;

                if (c % 2 === 1) {
                    html += `</tr>`;
                }
            }

            html += `    </tbody>
    </table>`;

            $('#divLODPeaks').html(html);

            $("#divLODPeaks").find(":input").on('change', function () {
                let checkedVals = $("#divLODPeaks").find(":input:checked").map(function () {
                    return this.value;
                }).get();
                //console.log(checkedVals);

                plotLODPeaks(g.DATASETS[g.dataSetID].lodpeaks[covar]);
            });

            }

            plotLODPeaks(g.DATASETS[g.dataSetID].lodpeaks[covar], maxLOD);
        }

        /**
         *
         */
        function startTask() {
            g.runningTask = true;
            Swal.fire({
                allowEnterKey: false,
                allowEscapeKey: false,
                allowOutsideClick: false,
                buttonsStyling: false,
                //cancelButtonClass: 'btn btn-danger',
                html: $('#ap').html(),
                showConfirmButton: false,
                showCancelButton: false,
                showCloseButton: true,
                title: 'Processing...',
                width: '20rem'
            }).then((result) => {
                logDebug('then called');
                // result.dismiss can be 'cancel', 'overlay',
                // 'close', and 'timer'
                if (result.dismiss === 'close') {
                    stopTask();
                }
            });
        }

        function stopTask() {
            g.runningTask = false;
            Swal.close();
        }

        /**
         * Download chomo
         * @param release
         * @param species
         * @param _id
         * @param callback
         */
        function downloadChromosomes(release, species, _id, callback) {
            let chromosomeURL = `{{ url_for('api.api_get', url='', _external=False) }}${g.PROTOCOL}//churchilllab.jax.org/ensimpl/api/chromosomes`;
            let amp = '';

            if (release !== null) {
                chromosomeURL += `?release=${release}`;
                amp = '&';
            }

            if (species !== null) {
                chromosomeURL += `${amp}species=${species}`;
            }

            downloadData(chromosomeURL, `Chromosomes ${_id}`, callback);
        }

        function downloadLODPeaks(dataSetID, callback) {
            let lodPeaksURL = `{{ url_for('api.api_get', url=config.API_R_BASE + '/lodpeaks', _external=False) }}?dataset=${dataSetID}`;
            downloadData(lodPeaksURL, 'LODPeaks', callback);
        }

        function downloadEnvInfo(callback) {
            let envInfoURL = '{{ url_for('api.api_get', url=config.API_R_BASE + '/envinfo', _external=False) }}';
            downloadData(envInfoURL, 'envInfo', callback);
        }

        function downloadDataSets(callback) {
            let dataSetsURL = '{{ url_for('api.api_get', url=config.API_R_BASE + '/datasets', _external=False) }}';
            downloadData(dataSetsURL, 'dataSets', callback);
        }

        function exportSNPAssocData(id, snps, genes) {
            // TODO: export genes?
            let csvContent = `"id","chr","position","alleles","csq","lod"\n`;

            $.each(snps, function(k, v) {
                csvContent += `"${v.id}","${v.chr}",${v.pos},"${v.alleles}","${v.csq}",${v.lod}\n`;
            });

            downloadCSV(csvContent, `${id}_SNPASSOC.csv`, 'text/csv;encoding:utf-8');
        }

        function randomInterval(min, max) { // min and max included
            return Math.floor(Math.random() * (max - min + 1) + min)
        }

        function createGeneRankings(geneRankings, geneInfo) {
            let rankings = {};
            $.each(geneRankings, function(idx, elem) {
                rankings[elem['gene_id']] = elem['ranking'];
            });

            logDebug('rankings=', rankings);

            $.each(geneInfo, function(idx, elem) {
                elem['score'] = rankings[elem['id']] || randomInterval(10, 80);
                elem['exons'] = elem['exons'].join(',');
            });

            return geneInfo;
        }

        async function correlationTooltipHandler(datum, mark, props) {
            const itemTemplates = [];

            $.each(datum.map, function(k, v) {
                itemTemplates.push(genomeSpyEmbed.html`
                    <strong>${v}</strong> ${datum[k]}<br/>
                `);
            });

            if (datum.imputed) {
                itemTemplates.push(genomeSpyEmbed.html`
                    <i>Imputed</i>
                `);
            }

            return genomeSpyEmbed.html`
                <div class="title">
                    <strong>${datum.sample_id}</strong>
                </div>
                <p class="summary">
                    ${itemTemplates}
                </p>
            `;
        }

        async function mediationTooltipHandler(datum, mark, props) {
            logDebug(datum);
            let protein = '';
            let phos = '';
            let uniprot = '';
            let ptm = '';
            let peptide = '';

            if (('ptm_id' in datum) &&
                (datum['ptm_id'] != null) &&
                (datum['ptm_id'] !== '')) {
                ptm = genomeSpyEmbed.html`
                    <strong>PTM ID</strong> ${datum.ptm_id}
                    <br/>`;
            }

            if (('peptide_id' in datum) &&
                (datum['peptide_id'] != null) &&
                (datum['peptide_id'] !== '')) {
                peptide = genomeSpyEmbed.html`
                    <strong>Peptide ID</strong> ${datum.peptide_id}
                    <br/>`;
            }

            if (('uniprot_id' in datum) &&
                (datum['uniprot_id'] != null) &&
                (datum['uniprot_id'] !== '')) {
                uniprot = genomeSpyEmbed.html`
                    <strong>Uniprot ID</strong> ${datum.uniprot_id}
                    <br/>`;
            }

            if (('phos_id' in datum) &&
                (datum['phos_id'] != null) &&
                (datum['phos_id'] !== '')) {
                phos = genomeSpyEmbed.html`
                    <strong>Phos ID</strong> ${datum.phos_id}
                    <br/>`;
            }

            if (('protein_id' in datum) &&
                (datum['protein_id'] != null) &&
                (datum['protein_id'] !== '')) {
                protein = genomeSpyEmbed.html`
                    <strong>Protein ID</strong> ${datum.protein_id}
                    <br/>`;
            }

            return genomeSpyEmbed.html`
                <div class="title">
                    <strong>${datum.LOD}</strong>
                </div>
                <p class="summary">
                    ${phos}
                    ${uniprot}
                    ${ptm}
                    ${peptide}
                    ${protein}
                    <strong>Gene ID</strong> ${datum.gene_id}
                    <br/>
                    <strong>Symbol</strong> ${datum.symbol}
                    <br/>
                    <strong>Location</strong> ${datum.chrom}:${datum.position.toLocaleString()}
                    <br/>
                </p>
            `;
        }

        async function lodPeakGeneTooltipHandler(datum, mark, props) {
            logDebug(datum, mark, props);
            let AH = '';
            let protein = '';
            let phos = '';
            let ptm = '';
            let peptide = '';
            let uniprot = '';

            if (('uniprotID' in datum) &&
                (datum['uniprotID'] != null) &&
                (datum['uniprotID'] !== '')) {
                uniprot = genomeSpyEmbed.html`
                    <strong>Uniprot ID</strong> ${datum.uniprotID}
                    <br/>`;
            }

            if (('proteinID' in datum) &&
                (datum['proteinID'] != null) &&
                (datum['proteinID'] !== '')) {
                protein = genomeSpyEmbed.html`
                    <strong>Protein ID</strong> ${datum.proteinID}
                    <br/>`;
            }

            if (('phosID' in datum) &&
                (datum['phosID'] != null) &&
                (datum['phosID'] !== '')) {
                phos = genomeSpyEmbed.html`
                    <strong>Phos ID</strong> ${datum.phosID}
                    <br/>`;
            }

            if (('ptmID' in datum) &&
                (datum['ptmID'] != null) &&
                (datum['ptmID'] !== '')) {
                ptm = genomeSpyEmbed.html`
                    <strong>PTM ID</strong> ${datum.ptmID}
                    <br/>`;
            }

            if (('peptideID' in datum) &&
                (datum['peptideID'] != null) &&
                (datum['peptideID'] !== '')) {
                peptide = genomeSpyEmbed.html`
                    <strong>Peptide ID</strong> ${datum.peptideID}
                    <br/>`;
            }

            if ('A' in datum) {
                AH = genomeSpyEmbed.html`
                    <br/>
                    <strong>Allele Effects</strong>
                    <br/>  A: ${datum.A}
                    <br/>  B: ${datum.B}
                    <br/>  C: ${datum.C}
                    <br/>  D: ${datum.D}
                    <br/>  E: ${datum.E}
                    <br/>  F: ${datum.F}
                    <br/>  G: ${datum.G}
                    <br/>  H: ${datum.H}
                `;
            }

            return genomeSpyEmbed.html`
                <div class="title">
                    <strong>${datum.lod}</strong>
                </div>
                <p class="summary">
                    ${phos}
                    ${ptm}
                    ${peptide}
                    ${uniprot}
                    ${protein}
                    <strong>Gene ID</strong> ${datum.geneID}
                    <br/>
                    <strong>Symbol</strong> ${datum.geneSymbol}
                    <br/>
                    <strong>Gene Position</strong> ${datum.geneChr}:${datum.genePos.toLocaleString()}
                    <br/>
                    <strong>Marker Position</strong> ${datum.markerChr}:${datum.markerPos.toLocaleString()}
                    ${AH}
                </p>`;
        }

        async function lodPeakPhenotypeTooltipHandler(datum, mark, props) {
            logDebug(datum);
            return genomeSpyEmbed.html`
                <div class="title">
                    <strong>${datum.lod}</strong>
                </div>
                <p class="summary">
                    <strong>ID</strong> ${datum.phenoDataName}
                    <br/>
                    <strong>Phenotype</strong> ${datum.phenoShortName}
                    <br/>
                    <strong>Category</strong> ${datum.phenoCategory}
                    <br/>
                    <strong>Marker Position</strong> ${datum.markerChr}:${datum.markerPos.toLocaleString()}
                </p>
            `;
        }


        function parseCSQ(strCSQ) {
            if (strCSQ.length <= 1) {
                return "";
            }

            return strCSQ;

            /*

            let consequences = new Set();
            let csq_info_split = strCSQ.split(';');

            for (let x = 0; x < csq_info_split.length; x++) {
                let elems = csq_info_split[x].split('|')[4].split('&');
                for (let y = 0; y < elems.length; y++) {
                    consequences.add(elems[y]);
                }
            }

            return Array.from(consequences).join(', ');
            */
        }

        async function snpHandler(datum, mark, props) {
            //logDebug(datum, mark, props);
            let sdp = sdpn_to_sdpD(datum.sdpn);
            let csq = parseCSQ(datum.csq);

            if ('lodc' in datum) {
                return genomeSpyEmbed.html`
                    <div class="title">
                        <strong>${datum.snp}</strong>
                    </div>
                    <p class="summary">
                        <strong>Position</strong> ${datum.chr}:${datum.pos.toLocaleString()}
                        <br/>
                        <strong>REF</strong> ${datum.ref}
                        <br/>
                        <strong>ALT</strong> ${datum.alt}
                        <br/>
                        <strong>SDP</strong> ${sdp}
                        <br/>
                        <strong>CSQ</strong> ${csq}
                        <br/>
                        <strong>LOD</strong> ${datum.lod}
                        <br/>
                        <strong>LOD ${datum.cov}</strong> ${datum.lodc}
                    </p>`;
            } else {
                return genomeSpyEmbed.html`
                    <div class="title">
                        <strong>${datum.snp}</strong>
                    </div>
                    <p class="summary">
                        <strong>Position</strong> ${datum.chr}:${datum.pos.toLocaleString()}
                        <br/>
                        <strong>REF</strong> ${datum.ref}
                        <br/>
                        <strong>ALT</strong> ${datum.alt}
                        <br/>
                        <strong>SDP</strong> ${sdp}
                        <br/>
                        <strong>CSQ</strong> ${csq}
                        <br/>
                        <strong>LOD</strong> ${datum.lod}
                    </p>`;
            }
        }

        async function nullHandler(datum, mark, props) {
            //logDebug(datum, mark, props);
            return null;//genomeSpyEmbed.html``
        }

        // TODO: Replace with an LRU-cache
        const symbolSummaryCache = new Map();

        /**
         * @type {import("./tooltipHandler").TooltipHandler}
         */
        async function ensimplTooltipHandler(datum, mark, params) {
            const ensembl_id = datum.id;

            let summary =
                symbolSummaryCache.get(ensembl_id) ??
                (await debouncedFetchEnsimplSummary(ensembl_id));

            if (summary) {
                let e = g.DATASETS[g.dataSetID].ensembl_version;

                symbolSummaryCache.set(ensembl_id, summary);
                logDebug(summary);
                return genomeSpyEmbed.html`
                    <div class="title">
                        <strong>${summary.id}</strong>
                    </div>
                    <p class="summary">
                        <strong>Symbol</strong> ${summary.symbol}
                        <br/>
                        <strong>Name</strong> ${summary.name}
                        <br/>
                        <strong>Location</strong> ${summary.chromosome}:${summary.start}-${summary.end}
                    </p>
                    <p class="source">
                        Source: Ensimpl ${e}
                    </p>
                `;
            } else {
                return null;
            }
        }
        function debounce(func, wait, rejectOnDebounce = true) {
            /** @type {number} */
            let timeout;

            /** @type {(reason?: any) => void} */
            let rejectPrevious = _ => undefined;

            /**
             * @param {T} args
             * @return {Promise<R>} */
            const debouncer = function debouncer(...args) {
                return new Promise((resolve, reject) => {
                    const later = () => {
                        clearTimeout(timeout);
                        rejectPrevious = _ => undefined;

                        resolve(func(...args));
                    };

                    if (rejectOnDebounce) {
                        rejectPrevious("debounced");
                    }
                    clearTimeout(timeout);

                    rejectPrevious = reject;
                    timeout = setTimeout(later, wait);
                });
            };

            return debouncer;
        }
        /**
         * @param {string} symbol
         */
        async function fetchEnsimplSummary(ensemblID) {
            logDebug("Searching: " + ensemblID);

            let ensembl = g.DATASETS[g.dataSetID].ensembl_version;

            //const opts = { mode: "cors" };
            const opts = { };

            const searchResult = await fetch(
                `https://churchilllab.jax.org/ensimpl/api/gene/${ensemblID}?species=Mm&release=${ensembl}`,
                opts
            ).then(res => res.json());

            // TODO: Handle failed searchs
            if ('message' in searchResult) {
                return null;
            }

            return searchResult['gene'][ensemblID];
        }

        const debounced = debounce(fetchEnsimplSummary, 500);

        /**
         *
         * @param {string} symbol
         */
        function debouncedFetchEnsimplSummary(ensemblID) {
            return debounced(ensemblID);
        }

        function plotSNPAssoc(snpData, geneRanking, geneInfo, id, chromosome, location, zoomWindow, domainWindow, covar) {
            logDebug('plotSNPAssoc', chromosome, location, covar);
            logDebug('geneRanking=', geneRanking);
            logDebug('geneInfo=', geneInfo);

            // combine the geneInfo and geneRanking
            let geneInformation = createGeneRankings(geneRanking, geneInfo);

            logDebug('geneInformation=', geneInformation);

            let minZoomPos = Math.max(location - zoomWindow, 0);
            let maxZoomPos = Math.min(location + zoomWindow, g.DATASETS[g.dataSetID].chromosomes.chr[chromosome].length);
            let minDomainPos = Math.max(location - domainWindow, 0);
            let maxDomainPos = Math.min(location + domainWindow, g.DATASETS[g.dataSetID].chromosomes.chr[chromosome].length);
            let plotTitle = '';
            let currentID = '';

            if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                plotTitle = `${g.geneID} (${g.gene.gene[g.geneID].symbol})`;
                currentID = g.geneID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                plotTitle = `${g.proteinID} (${g.gene.gene[g.geneID].symbol})`;
                currentID = g.proteinID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                plotTitle = `${g.uniprotID} ${g.proteinID} (${g.gene.gene[g.geneID].symbol})`;
                currentID = g.uniprotID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                plotTitle = g.phosID;
                currentID = g.phosID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                plotTitle = g.ptmID;
                currentID = g.ptmID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                plotTitle = g.peptideID;
                currentID = g.peptideID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                plotTitle = g.phenotypeID;
                currentID = g.phenotypeID;
            } else {
                // TODO: handle error
                logError('MAJOR PROBLEM');
            }

            snpData.sort(function(a, b) {
                return a.pos - b.pos;
            });

            let snps = [];
            let snpsCovar = [];
            let snpsCovarTitle = '';

            if (snpData) {
                if ((covar !== null) && (covar !== 'additive')) {
                    // grab the covar display name
                    $.each(g.DATASETS[g.dataSetID]['covar_info'], function(idx, elem) {
                        if (covar === elem['sample_column']) {
                            snpsCovarTitle = 'LOD (' + elem['display_name'] + ')';
                        }
                    });
                }
            }

            /*
            0: SNP:  "rs12345678"
            1: CHR:  "3"
            2: POS:  118281232
            3: REF:  "C"
            4: ALT:  "T,G"
            5: SDPN: 21853
            6: CSQ:  "T||||intergenic_variant||||||||"
            7: LOD:  0.24

            IF covar:
            8: LODC: 0.0029
            */
            $.each(snpData, function(index, element) {
                let snp = {
                    snp:  element[0],
                    chr:  element[1],
                    pos:  element[2],
                    ref:  element[3],
                    alt:  element[4],
                    sdpn: element[5],
                    csq:  element[6],
                    lod:  element[7] * 1.0,
                    cov:  covar
                };

                if ((covar !== null) && (covar !== 'additive')) {
                    snp['lodc'] = element[8] * 1.0;
                }

                snps.push(snp);
            });
            let minLOD = snps.reduce((a, b) => a.lod < b.lod ? a : b).lod;
            let maxLOD = snps.reduce((a, b) => a.lod > b.lod ? a : b).lod;

            logDebug('minLOD=', minLOD);
            logDebug('maxLOD=', maxLOD);

            minLOD = Math.max(0, Math.floor(minLOD) - 1);
            maxLOD = Math.ceil(maxLOD) + 1;

            logDebug('minLOD=', minLOD);
            logDebug('maxLOD=', maxLOD);
            logDebug('minZoomPos', minZoomPos);
            logDebug('maxZoomPos', maxZoomPos);
            logDebug('minDomainPos', minDomainPos);
            logDebug('maxDomainPos', maxDomainPos);
            logDebug('#snps=', snps.length);
            logDebug('snps=', snps);
            logDebug('#snpsCovar=', snpsCovar.length);
            logDebug('snpsCovar=', snpsCovar);

            /**
             * BUILD GENOME SPY
             */

            let tracksAll = [];

            let layers = [];
            layers.push({
                "title": "SNPs",
                "mark": {
                    "type": "point",
                    "tooltip": {
                        "handler": "snpHandler"
                    },
                    "geometricZoomBound": 4,
                    "inwardStroke": true,
                    "stroke": "#962e1e",
                    "strokeWidth": 0.7,
                    "fillOpacity": 0.6,
                    "strokeOpacity": 0.8
                },

                "encoding": {
                    "x": {
                        "chrom": "chr",
                        "pos": "pos",
                        "type": "locus",
                        "scale": {
                            "domain": [
                                {
                                    "chrom": chromosome,
                                    "pos": minDomainPos
                                },
                                {
                                    "chrom": chromosome,
                                    "pos": maxDomainPos
                                }
                            ],

                            "zoom": {
                                "extent": [
                                    {
                                        "chrom": chromosome,
                                        "pos": minZoomPos
                                    },
                                    {
                                        "chrom": chromosome,
                                        "pos": maxZoomPos
                                    }
                                ]
                            }
                        }
                    },
                    "y": {
                        "field": "lod",
                        "type": "quantitative",
                        "title": "LOD",
                        "scale": { "domain": [minLOD, maxLOD] }
                    },
                    "color": {
                        "value": "#f35137"
                    },
                    "size": {
                        "value": 30
                    },
                    "opacity": {
                        "value": 0.5
                    },
                }
            }
            );

            if ((covar !== null) && (covar !== 'additive')) {
                layers.push({
                    "title": "SNPs Covar",
                    "mark": {
                        "type": "point",
                        "tooltip": {
                            "handler": "snpHandler"
                        },
                        "geometricZoomBound": 4,
                        "inwardStroke": true,
                        "stroke": "#113c65",
                        "strokeWidth": 0.7,
                        "fillOpacity": 0.6,
                        "strokeOpacity": 0.8

                    },

                    "encoding": {
                        "x": {
                            "chrom": "chr",
                            "pos": "pos",
                            "type": "locus",
                            "scale": {
                                "domain": [
                                    {
                                        "chrom": chromosome,
                                        "pos": minDomainPos
                                    },
                                    {
                                        "chrom": chromosome,
                                        "pos": maxDomainPos
                                    }
                                ],

                                "zoom": {
                                    "extent": [
                                        {
                                            "chrom": chromosome,
                                            "pos": minZoomPos
                                        },
                                        {
                                            "chrom": chromosome,
                                            "pos": maxZoomPos
                                        }
                                    ]
                                }
                            }
                        },
                        "y": {
                            "field": "lodc",
                            "type": "quantitative",
                            "title": "LOD",
                            "scale": { "domain": [minLOD, maxLOD] }
                        },
                        "color": {
                            "value": "#7cb5ec"
                        },
                        "size": {
                            "value": 30
                        },
                        "opacity": {
                            "value": 0.5
                        },
                    }
                }
                );

            }


            logDebug('layers=', layers);



            let trackSNPs = {
                "name": "snps",
                "data": {
                    "values": snps
                },
                "layer": layers

            };
            //"value": "#7cb5ec"
            tracksAll.push(trackSNPs);

            let trackGenes = {
                "name": "geneAnnotations",

                "description": [
                    "Ensimpl genes scored by their data values",
                ],

                //"height": { "px": 100 },
                "height": 70,

                "data": {
                    "values": geneInformation,
                },

                "transform": [{
                        "type": "linearizeGenomicCoordinate",
                        "chrom": "chr",
                        "pos": "start",
                        "as": "_start"
                    },
                    {
                        "type": "formula",
                        "expr": "ceil(random() * 100)",
                        "as": "score"
                    },
                    {
                        "type": "formula",
                        "expr": "datum._start + datum.length",
                        "as": "_end"
                    },
                    {
                        "type": "formula",
                        "expr": "datum._start + datum.length / 2",
                        "as": "_centroid"
                    },
                    {
                        "type": "collect",
                        "sort": {
                            "field": ["_start"]
                        }
                    },
                    {
                        "type": "pileup",
                        "start": "_start",
                        "end": "_end",
                        "as": "_lane",
                        "preference": "strand",
                        "preferredOrder": ["-", "+"]
                    },
                    {
                        "type": "filter",
                        "expr": "datum._lane < 5"
                    }
                ],

                "encoding": {
                    "y": {
                        "field": "_lane",
                        "type": "ordinal",
                        "scale": {
                            "type": "index",
                            "align": 0,
                            "paddingInner": 0.4,
                            "paddingOuter": 0.2,
                            "domain": [0, 5],
                            "reverse": true,
                            "zoom": false
                        },
                        "axis": null
                    }
                },
                "layer": [{
                        "name": "transcripts",

                        "opacity": {
                            "unitsPerPixel": [100000, 40000],
                            "values": [0, 1]
                        },

                        "encoding": {
                            "color": {
                                "value": "#909090"
                            }
                        },

                        "layer": [{
                                "name": "exons",

                                "transform": [{
                                        "type": "project",
                                        "fields": ["_lane", "_start", "exons"]
                                    },
                                    {
                                        "type": "flattenCompressedExons",
                                        "start": "_start"
                                    }
                                ],

                                "mark": {
                                    "type": "rect",
                                    "minOpacity": 0.2,
                                    "minWidth": 0.5,
                                    "buildIndex": true,
                                    "tooltip": null
                                },

                                "encoding": {
                                    "x": {
                                        "field": "exonStart",
                                        "type": "locus"
                                    },
                                    "x2": {
                                        "field": "exonEnd"
                                    }
                                }
                            },
                            {
                                "name": "bodies",

                                "mark": {
                                    "type": "rule",
                                    "minLength": 0.5,
                                    "size": 1,
                                    "buildIndex": true,
                                    "tooltip": null
                                },
                                "encoding": {
                                    "x": {
                                        "field": "_start",
                                        "type": "locus",
                                        "axis": null
                                    },
                                    "x2": {
                                        "field": "_end"
                                    },
                                    "search": {
                                        "field": "symbol",
                                        "type": "nominal"
                                    }
                                }
                            }
                        ]
                    },
                    {
                        "name": "symbols",

                        "transform": [{
                                "type": "measureText",
                                "fontSize": 11,
                                "field": "symbol",
                                "as": "_textWidth"
                            },
                            {
                                "type": "filterScoredLabels",
                                "lane": "_lane",
                                "score": "score",
                                "width": "_textWidth",
                                "pos": "_centroid",
                                "padding": 5
                            }
                        ],

                        "layer": [{
                            "name": "labels",
                            "mark": {
                                "type": "text",
                                "dynamicData": true,
                                "size": 11,
                                "yOffset": 7,
                                "tooltip": {
                                    "handler": "ensimplTooltipHandler"
                                }
                            },
                            "encoding": {
                                "x": {
                                    "field": "_centroid",
                                    "type": "locus"
                                },
                                "text": {
                                    "field": "symbol",
                                    "type": "nominal"
                                }
                            }
                        },
                        {
                            "name": "arrows",
                            "opacity": {
                                "unitsPerPixel": [100000, 40000],
                                "values": [0, 1]
                            },
                            "mark": {
                                "type": "point",
                                "dynamicData": true,
                                "yOffset": 7,
                                "size": 50,
                                "tooltip": null
                            },
                            "encoding": {
                                "x": {
                                    "field": "_centroid",
                                    "type": "locus"
                                },
                                "dx": {
                                    "expr": "(datum._textWidth / 2 + 5) * (datum.strand == '-' ? -1 : 1)",
                                    "type": "quantitative",
                                    "scale": null
                                },
                                "color": {
                                    "value": "black"
                                },
                                "shape": {
                                    "field": "strand",
                                    "type": "nominal",
                                    "scale": {
                                        "domain": ["-", "+"],
                                        "range": ["triangle-left", "triangle-right"]
                                    }
                                }
                            }
                        }]
                    }
                ]
            };

            tracksAll.push(trackGenes);

            const spec = {
                "description":[
                    "Put a description....",
                    "Here if you would like to!"
                ],

                "genome": {
                    "name": g.DATASETS[g.dataSetID].ensembl_species + " " + g.DATASETS[g.dataSetID].ensembl_release,
                    "contigs": g.DATASETS[g.dataSetID].chromosomes.contigs
                },

                "resolve": {
                    "scale": {
                        "x": "shared",
                    }
                },

                "vconcat": tracksAll
            };

            $('#snpWindowMenuShift').html(`
                <div class="col">
                    <button id="snpShiftLeft" class="btn btn-secondary"><i class="fas fa-arrow-left"></i> Shift Plot Left</button>
                </div>
                <div class="col">
                    <button id="snpShiftRight" class="btn btn-secondary float-right">Shift Plot Right <i class="fas fa-arrow-right"></i></button>
                </div>`);

            $('#snpShiftLeft').button().on('click', function(event) {
                logDebug(location);
                logDebug(id, chromosome, location - 2000000, covar);
                generateSNPAssocPlot(id, chromosome, location - 2000000, covar);
            });

            $('#snpShiftRight').button().on('click', function(event) {
                logDebug(location);
                logDebug(id, chromosome, location + 2000000, covar);
                generateSNPAssocPlot(id, chromosome, location + 2000000, covar);
            });

            $('#snpWindowMenu').append(`
                <div class="col text-center font-weight-bold">${plotTitle}
                <div class="dropdown dropleft float-right" id="snpMenuDropdown">
                  <a class="text-secondary" href="#" role="button" id="dropdownMenuLink" data-toggle="dropdown" aria-expanded="false">
                    <i class="fa-solid fa-bars"></i>
                  </a>

                  <div class="dropdown-menu" aria-labelledby="dropdownMenuLink">
                    <a class="dropdown-item" id="snpDownloadPNG" style="font-size:11px !important">Download PNG image</a>
                    <a class="dropdown-item" id="snpDownloadCSV" style="font-size:11px !important">Download CSV</a>
                  </div>
                </div>


                <!--
                TODO: change to the following icon when we switch out of highcharts
                <i class="fa-solid fa-file-arrow-down float-right"></i>
                //-->

                </div>`);

            $("#snpDownloadPNG").click(function (event) {
                downloadGenomeSpy('#plotSNPAssociation', `${id}.png`, '255,255,255');
            });

            $("#snpDownloadCSV").click(function (event) {
                exportSNPAssocData(id, snps, geneInfo);
            });

            $('#plotSNPAssociation').height('500');

            g.chartSNPAssoc = genomeSpyEmbed.embed(
                document.querySelector("#plotSNPAssociation"),
                spec,
                {
                    // Register the custom handler
                    tooltipHandlers: {
                        snpHandler: snpHandler,
                        ensimplTooltipHandler: ensimplTooltipHandler
                    }
                }
            );
        }

        /**
         * Start downloading snp
         */
        function generateSNPAssocPlot(id, chromosome, location, covar) {
            // TODO: windowSize needs to be figured out
            let locationMb = location;

            // what data is retrieved
            // [locationMb - zoomWindow, locationMb + zoomWindow]
            let zoomWindow = 3000000;

            // what data is initially displayed
            // [locationMb - domainWindow, locationMb + domainWindow]
            let domainWindow = 500000;

            let minPos = Math.max(locationMb - zoomWindow, 0);
            let maxPos = locationMb + zoomWindow;

            logDebug('location=', location);
            logDebug('locationMb=', locationMb);
            logDebug('zoomWindow=', zoomWindow);
            logDebug('domainWindow=', domainWindow);
            logDebug('minPos=', minPos);
            logDebug('maxPos=', maxPos);

            let snpAssocURL = `{{ config.API_R_BASE }}/snpassocmapping?dataset=${g.dataSetID}&id=${id}&chrom=${chromosome}&location=${locationMb}&window_size=${zoomWindow}`;
            let genesInfoURL = `${g.PROTOCOL}//churchilllab.jax.org/ensimpl/api/exon_info?release=${g.DATASETS[g.dataSetID].ensembl_release}&species=${g.DATASETS[g.dataSetID].ensembl_species}&chrom=${chromosome}`;

            if (covar !== 'additive') {
                snpAssocURL += `&intcovar=${covar}`;
            }

            {% if config.API_R_NUM_CORES %}
            snpAssocURL += '&cores={{ config.API_R_NUM_CORES }}';
            {% endif %}

            let submitData = {
                urls:[{
                    url_id: 'snpAssoc',
                    url: snpAssocURL
                }, {
                    url_id: 'genesInfoURL',
                    url: genesInfoURL
                }]
            };

            logDebug('g.DATASETS[QTL.g.dataSetID].datatype.toLowerCase()=', g.DATASETS[QTL.g.dataSetID].datatype.toLowerCase());

            if (g.DATASETS[QTL.g.dataSetID].datatype !== 'pheno') {
                let genesRankingURL = `{{ config.API_R_BASE }}/rankings?dataset=${g.dataSetID}&chrom=${chromosome}`;
                submitData['urls'].push({
                    url_id: 'genesRankingURL',
                    url: genesRankingURL
                });
            }

            logDebug('submitData=', submitData);

            startTask();

            // reset the chart and clear it
            g.chartSNPAssoc = null;

            $('#snpWindowMenuShift').html('');
            $('#snpWindowMenu').html('');
            $('#plotSNPAssociation').html('');

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=False) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                retries: 3,
                retryInterval: 1000,
                success: function(data, status, request) {
                    logDebug('data=', data);
                    updateSNPAssocData(data.group_id, id, chromosome, locationMb, zoomWindow, domainWindow, covar);
                },
                error: function(jqXHR, textStatus, errorThrown) {
                    showErrorMessage(errorThrown, textStatus);
                }
            });
        }

        /**
         * Update the status of the downloading effect data.
         *
         * @param {string} groupID - the group identifier of the task
         */
        function updateSNPAssocData(groupID, id, chromosome, location, zoomWindow, domainWindow, covar) {
            // send GET request to status URL
            logDebug('updateSNPAssocData: ', groupID);

            if (g.runningTask) {
                let statusURL = '{{ url_for('api.api_status', group_id='', _external=False) }}' + groupID;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    retries: 3,
                    retryInterval: 1000,
                    success: function (data, status, request) {
                        logDebug('DATA=======', data);
                        if (data.status === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data.number_tasks_errors !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.error}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data.number_tasks_errors === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if (value.status_code !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.response.error}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the SNP Association Plot.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result, there will be 2 datasets to get
                                    let rankings = null;
                                    if ('genesRankingURL' in data.response_data) {
                                        rankings = data.response_data.genesRankingURL.response.result;
                                    }

                                    plotSNPAssoc(data.response_data.snpAssoc.response.result,
                                                 rankings,
                                                 data.response_data.genesInfoURL.response.genes,
                                                 id, chromosome, location, zoomWindow, domainWindow, covar);
                                    stopTask();
                                }
                            }
                        } else {
                            // rerun in 1 seconds
                            logDebug('Not done, keep checking...');
                            setTimeout(function () {
                                updateSNPAssocData(groupID, id, chromosome, location, zoomWindow, domainWindow, covar);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                logDebug('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=False) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    g.chartSNPAssoc = null;
                    $('#snpWindowMenuShift').html('');
                    $('#snpWindowMenu').html('');
                    $('#plotSNPAssociation').html('');
                });
            }
        }

        function exportEffectData(id, effectData, lodData, covar, category) {
            let csvContent = `"id","chromosome","position"{%- for s in config.PLOT_EFFECT_STRAINS -%}
                ,"{{ s['name'] }}"
            {%- endfor -%},"lod"\n`;

            $.each(effectData, function(k, v) {
                let line = `"${v[0]}","${v[1]}",${v[2]},${v[3]},${v[4]},${v[5]},${v[6]},${v[7]},${v[8]},${v[9]},${v[10]},`;
                line += lodData[k].y;
                csvContent += `${line}\n`;
            });

            if (category !== null) {
                downloadCSV(csvContent, `${id}_EFFECT_${covar}_${category}.csv`, 'text/csv;encoding:utf-8');
            } else {
                downloadCSV(csvContent, `${id}_EFFECT.csv`, 'text/csv;encoding:utf-8');
            }
        }

        /**
         * Plot the effect data.
         * @param {Object} effectData - the effect data
         */
        function plotEffectData(effectData, lodData, chromosome, covar) {
            logDebug('plotEffectData() ', effectData, lodData, covar);

            if (covar === 'additive') {
                plotEffectChart(effectData[covar], null, covar, null, chromosome);
            } else {
                $.each(effectData, function(key, value) {
                    // key = category value, i.e for sex would be M,F
                    plotEffectChart(value, lodData[key], covar, key, chromosome);
                });
            }
        }

        function roundToX(num, decimals) {
            return +(Math.round(num + "e" + decimals) + "e-" + decimals);
        }

        function plotEffectChart(effectData, lodData, covar, category, chromosome) {
            logDebug('plotEffectChart() ', effectData, lodData, covar, category);

            let strainInfo = {};
            {%- for s in config.PLOT_EFFECT_STRAINS %}
            strainInfo['{{ s['key'] }}'] = {
                color: '{{ s['color'] }}',
                name: '{{ s['name'] }}',
                data: []
            };
            {%- endfor %}

            let renderTo = '';
            let isCovar = false;

            if (covar === 'additive') {
                renderTo = 'plotEffect';
                // get the data from the lod chart, but filter for just the chromosome
                lodData = g.LODChartData[covar];
            } else {
                isCovar = true;
                renderTo = 'plotEffect_' + category;
            }

            $('#allEffectPlots').append(`<div class="row"><div class="col">
                                                <div id="${renderTo}"></div>
                                            </div></div>`);

            logDebug('covar=', covar);
            logDebug('isCovar=', isCovar);
            logDebug('renderTo=', renderTo);

            let allEffectValues = [];
            let val = null;

            $.each(effectData, function(index, element) {
                //{"A":-0.1224,..."H":1.45,"chr":"4","id":"4_156339889","pos":156.3399}
                //id,chrom,pos,A-H
                let position = element[2];

            {%- for s in config.PLOT_EFFECT_STRAINS -%}
                {#- when data is expanded -#}
                {#- strainInfo['{{ s['key'] }}']['data'].push([position, element.{{ s['key'] }}]); -#}
                {#- allEffectValues.push(element.{{ s['key'] }}); -#}

                {#- when data is collapsed #}
                //val = roundToX(element[{{ loop.index0 + 3}}], 4);
                val = element[{{ loop.index0 + 3}}]
                if (!isNaN(+val))  {
                    strainInfo['{{ s['key'] }}'].data.push([position, val]);
                    allEffectValues.push(val);
                }
            {%- endfor %}
            });

            let effectMinValue = Math.floor(Math.min.apply(null, allEffectValues));
            let effectMaxValue = Math.ceil(Math.max.apply(null, allEffectValues));

            logDebug('effectMinValue=', effectMinValue);
            logDebug('effectMaxValue=', effectMaxValue);

            /*
            if ((effectMaxValue > 30.0) || (effectMinValue < -30.0)) {
                displayError('Effect Plot Error', 'Effect values appear to be invalid, ' + effectMinValue + ' to ' + effectMaxValue);
                return;
            }
            */

            logDebug('strainInfo=', strainInfo);

            let xAxisTickNum = Math.floor(g.DATASETS[g.dataSetID].chromosomes.chr[chromosome].length / 20000000);
            let xAxisTickVals = [];
            let xAxisTickText = [];
            let series = [];

            for (let i = 1; i <= xAxisTickNum; i++) {
                xAxisTickVals.push(20000000 * i);
                xAxisTickVals.push((20 * i) + 'Mb');
            }

            $.each(strainInfo, function(key, value) {
                let s = {
                    animation: 0,
                    color: value.color,
                    data: value.data,
                    lineWidth: 3,
                  //  marker: {
                 //       symbol: 'circle',
                   // },
                    name: value.name,
                    showInLegend: true,
                    stack: 0,
                    tooltip: {
                        shared: false,
                        headerFormat:'<b>{series.name}</b>',
                        pointFormatter: function() {
                            return `<br>Effect: ${this.y}<br>Position: ${this.x.toLocaleString()}`;
                        }
                    },
                    //turboThreshold: 0,
                    type: 'line',
                    yAxis:'axisEffect',
                };
                series.push(s);
            });

            let plotTitle = '';
            let currentID = '';

            if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                plotTitle = `${g.geneID} (${g.gene.gene[g.geneID].symbol})`;
                currentID = g.geneID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                plotTitle = `${g.proteinID} (${g.gene.gene[g.geneID].symbol})`;
                currentID = g.proteinID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                plotTitle = `${g.uniprotID} ${g.proteinID} (${g.gene.gene[g.geneID].symbol})`;
                currentID = g.uniprotID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                plotTitle = g.phosID;
                currentID = g.phosID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                plotTitle = g.ptmID;
                currentID = g.ptmID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                plotTitle = g.peptideID;
                currentID = g.peptideID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                plotTitle = g.phenotypeID;
                currentID = g.phenotypeID;
            } else {
                // TODO: handle error
                logError('MAJOR PROBLEM');
            }

            let newLodData = [];
            let lodMinValue = Infinity;
            let lodMaxValue = -Infinity;

            $.each(lodData, function(index, element) {
                //
                // element = {"[id]","[chr]",[position],[data]}
                //
                if ((element[1] in g.DATASETS[g.dataSetID].chromosomes.chr) && (element[1] === chromosome)) {

                    let chr = g.DATASETS[g.dataSetID].chromosomes.chr[element[1]];
                    let x = element[2];
                    let y = element[3];

                    let d = {
                        x: x,
                        y: y,
                        id: element[0],
                        //text: y,
                        //chr: chr.chromosome,
                        //chrPos: element[2],
                        //covar: covar
                    };
                    newLodData.push(d);
                    lodMinValue = Math.min(element.y, lodMinValue);
                    lodMaxValue = Math.max(element.y, lodMaxValue);
                }
            });

            series.push({
                animation: false,
                color: '#000000',
                data: newLodData,
                lineWidth: 1,
                name: 'LOD',
                showInLegend: false,
                //stack: 0,
                tooltip: {
                    shared: false,
                    headerFormat: '',
                    pointFormatter: function() {
                        if (this.id !== '') {
                            return `<b>${this.id}</b><br>LOD: ${this.y}<br>Position: ${this.x.toLocaleString()}`;
                        } else {
                            return `<br>LOD: ${this.y}<br>Position: ${this.x.toLocaleString()}`;
                        }
                    }
                },
                turboThreshold: 10000000, // TODO: what should this value be
                yAxis: 'axisLOD',
            });

            let lodTickVals = null;
            // round up to nearest 10, 11->20, 7->10, 88->90
            let lodMaxTick = Math.ceil(lodMaxValue / 10) * 10;

            // determine ticks if > 10, otherwise let the chart do it
            if (lodMaxValue > 10) {
                let lodMaxTick = lodMaxValue + 1;
                let maxNumTicks = Math.min(4, Math.ceil(lodMaxTick / 10) * 10);
                lodTickVals = [];

                let step = Math.ceil((lodMaxValue/maxNumTicks) / 10) * 10;
                for (let i = 0; i < lodMaxValue; i += step) {
                    lodTickVals.push(i);
                }
                lodTickVals.push(lodMaxTick);
            }

            let exporting = appendExportButton('Download CSV', function() {
                exportEffectData(currentID, effectData, newLodData, covar, category);
            });

            exporting.filename = currentID + '_EFFECT';

            if (isCovar) {
                exporting.filename = currentID + '_EFFECT_' + category;
                plotTitle += (' (' + category + ')');
            }

            g.chartEffect[category] = Highcharts.chart({
                boost: {
                    useGPUTranslations: true
                },
                chart: {
                    //displayErrors: true,
                    renderTo: renderTo,
                    zoomType: 'x'
                },
                exporting: exporting,
                legend: {
                    align: 'right',
                    backgroundColor: (Highcharts.theme && Highcharts.theme.legendBackgroundColor) || '#FFFFFF',
                    floating: false,
                    layout: 'vertical',
                    verticalAlign: 'top',
                    y: 100,
                },
                responsive: {
                    rules: [{
                        condition: {
                            maxWidth: 400
                        },
                        chartOptions: {
                            legend: {
                                align: 'center',
                                layout: 'horizontal',
                                verticalAlign: 'top',
                                y: 0,
                            },
                        }
                    }]
                },

                plotOptions: {
        series: {
            marker: {
                enabled: false
            }
        }
    },

                series: series,
                subtitle: {
                    text: 'Chromosome ' + chromosome,
                    verticalAlign: 'bottom',
                },
                title: {
                    text: plotTitle,
                },
                tooltip: {
                    outside: true,
                },
                xAxis: {
                    crosshair: true,
                    gridLineWidth: 1,
                    lineWidth: 0.5,
                },
                yAxis: [{
                    height: '70%',
                    id: 'axisEffect',
                    labels: {
                        formatter: function() {
                            if (this.isFirst) {
                                return '';
                            }
                            return this.value;
                        }
                    },
                    maxPadding: 0,
                    offset: 0,
                    title: {
                        text: 'EFFECT'
                    },
                }, {
                    height: '30%',
                    id: 'axisLOD',
                    labels: {
                        formatter: function() {
                            if (this.isLast) {
                                return '';
                            }
                            return this.value;
                        }
                    },
                    maxPadding: 0,
                    offset: 0,
                    plotLines: [{
                        value: lodMaxTick,
                        color: 'black',
                        zIndex: 5,
                        width: 2
                    }],
                    tickPositions: lodTickVals,
                    title: {
                        text: 'LOD'
                    },
                    top: '70%',
                }],
            });
        }

        function exportMediationData(id, data) {
            let csvContent = `"gene_id","symbol","chromosome","position","lod"\n`;

            if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                csvContent = '"protein_id",' + csvContent;
            } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                csvContent = '"phos_id","protein_id",' + csvContent;
            }

            $.each(data, function(k, v) {
                let line = `"${v.gene_id}","${v.symbol}","${v.chrom}",${v.position},${v.LOD}`;

                if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                    line = `"${v.protein_id}",` + line;
                } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                    line = `"${v.phos_id}","${v.protein_id}",` + line;
                }

                csvContent += `${line}\n`;
            });

            downloadCSV(csvContent, `${id}_MEDIATION.csv`, 'text/csv;encoding:utf-8');
        }

        /**
         * Plot mediation data
         * @param {Object} mediationData - mediation data
         */
        function plotMediation(mediationData, dsMediateAgainst) {
            logDebug('mediation');
            let newData = [];

            // build the chomosome rects
            let rects = [];
            $.each(g.DATASETS[g.dataSetID].chromosomes.contigs, function(k, v) {
                if ((k % 2) === 1) {
                    rects.push({
                        name: v.name,
                        size: v.size,
                        start: 1,
                        end: v.size
                    });
                }
            });

            if (g.DATASETS[dsMediateAgainst].datatype === 'mrna') {
                // ["ENSMUSG00000000001", "Gnai3", "3", 108.1267, 140.1666]
                $.each(mediationData, function(index, element) {
                    if (element[2] in g.DATASETS[g.dataSetID].chromosomes.chr) {
                        newData.push({
                            gene_id: element[0],
                            symbol: element[1],
                            chrom: element[2],
                            position: element[3],
                            LOD: element[4]
                        });
                    } else {
                        //logDebug(element);
                    }
                });

            } else if (g.DATASETS[dsMediateAgainst].datatype === 'protein') {
                // ["ENSMUSP00000000001", "ENSMUSG00000000001", "Gnai3", "3", 108.126713, 7.1102046955]
                $.each(mediationData, function(index, element) {
                    if (element[3] in g.DATASETS[g.dataSetID].chromosomes.chr) {
                        newData.push({
                            protein_id: element[0],
                            gene_id: element[1],
                            symbol: element[2],
                            chrom: element[3],
                            position: element[4],
                            LOD: element[5]
                        });
                    } else {
                        //logDebug(element);
                    }
                });
            } else if (g.DATASETS[dsMediateAgainst].datatype === 'protein_uniprot') {
                // ["UNIPROTID", "ENSMUSP00000000001", "ENSMUSG00000000001", "Gnai3", "3", 108.126713, 7.1102046955]
                $.each(mediationData, function(index, element) {
                    if (element[4] in g.DATASETS[g.dataSetID].chromosomes.chr) {
                        newData.push({
                            uniprot_id: element[0],
                            protein_id: element[1],
                            gene_id: element[2],
                            symbol: element[3],
                            chrom: element[4],
                            position: element[5],
                            LOD: element[6]
                        });
                    } else {
                        //logDebug(element);
                    }
                });
            } else if (g.DATASETS[dsMediateAgainst].datatype === 'phos') {
                // ["ENSMUSP00000024849_DSRTGDSMEASGF_636", "ENSMUSP00000000001", "ENSMUSG00000000001", "Gnai3", "3", 108.126713, 7.1102046955]
                $.each(mediationData, function(index, element) {
                    if (element[4] in g.DATASETS[g.dataSetID].chromosomes.chr) {
                        newData.push({
                            phos_id: element[0],
                            protein_id: element[1],
                            gene_id: element[2],
                            symbol: element[3],
                            chrom: element[4],
                            position: element[5],
                            LOD: element[6]
                        });
                    } else {
                        //logDebug(element);
                    }
                });
            } else if (g.DATASETS[dsMediateAgainst].datatype === 'peptide') {
                // ["A2ADY9", "A2ADY9_TQAQHSSPGEMASSPQGLDNPALLR_", "ENSMUSP00000099542", "ENSMUSG00000078515","Ddi2","4", 141427795, 6.9672]
                $.each(mediationData, function(index, element) {
                    if (element[5] in g.DATASETS[g.dataSetID].chromosomes.chr) {
                        newData.push({
                            uniprot_id: element[0],
                            peptide_id: element[1],
                            protein_id: element[2],
                            gene_id: element[3],
                            symbol: element[4],
                            chrom: element[5],
                            position: element[6],
                            LOD: element[7]
                        });
                    } else {
                        //logDebug(element);
                    }
                });
            } else if (g.DATASETS[dsMediateAgainst].datatype === 'ptm') {
                //"A2ADY9","A2ADY9_TQAQHSSPGEMASSPQGLDNPALLR_","A2ADY9_TQAQHSSPGEM[Oxidation_(M)]ASSPQGLDNPALLR_","ENSMUSP00000099542","ENSMUSG00000078515","Ddi2","4",141427795,6.6028
                $.each(mediationData, function(index, element) {
                    if (element[6] in g.DATASETS[g.dataSetID].chromosomes.chr) {
                        newData.push({
                            uniprot_id: element[0],
                            peptide_id: element[1],
                            ptm_id: element[2],
                            protein_id: element[3],
                            gene_id: element[4],
                            symbol: element[5],
                            chrom: element[6],
                            position: element[7],
                            LOD: element[8]
                        });
                    } else {
                        //logDebug(element);
                    }
                });
            } else {
                // TODO: handle error
                logError('MAJOR PROBLEM');
            }

            let plotTitle = '';
            let currentID = '';

            if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                plotTitle = `${g.geneID} (${g.gene.gene[g.geneID].symbol})`;
                currentID = g.geneID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                plotTitle = `${g.proteinID} (${g.gene.gene[g.geneID].symbol})`;
                currentID = g.proteinID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                plotTitle = `${g.uniprotID} ${g.proteinID} (${g.gene.gene[g.geneID].symbol})`;
                currentID = g.uniprotID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                plotTitle = g.phosID;
                currentID = g.phosID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                plotTitle = g.ptmID;
                currentID = g.ptmID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                plotTitle = g.peptideID;
                currentID = g.peptideID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                plotTitle = g.phenotypeID;
                currentID = g.phenotypeID;
            } else {
                // TODO: handle error
                logError('MAJOR PROBLEM');
            }

            let minLOD = newData.reduce((a, b) => a.LOD < b.LOD ? a : b).LOD;
            let maxLOD = newData.reduce((a, b) => a.LOD > b.LOD ? a : b).LOD;

            logDebug('minLOD=', minLOD);
            logDebug('maxLOD=', maxLOD);

            minLOD = Math.max(0, Math.floor(minLOD) - 1);
            maxLOD = Math.ceil(maxLOD) + 1;

            logDebug('minLOD=', minLOD);
            logDebug('maxLOD=', maxLOD);

            let axisTicks = calculateTicks(minLOD, maxLOD, 5);
            let axisLines = [];

            $.each(axisTicks, function(k,v) {
                axisLines.push({
                    axisLOD: v
                })
            });

            logDebug('minLOD=', minLOD);
            logDebug('maxLOD=', maxLOD);
            logDebug('axisLines=', axisLines);
            logDebug('newData=', newData);

            let trackMediation = {
                "name": "lod",
                "layer": [
                    {
                        "title": "Xbound",
                        "data": {
                            "values": rects
                        },
                        "mark": {
                            "type": "rect",
                            "tooltip": {
                                "handler": "nullHandler"
                            },
                        },
                        "encoding": {
                            "x": {
                                "chrom": "name",
                                "pos": "start",
                                "type": "locus",
                                "title": null
                            },
                            "x2": {
                                "chrom": "name",
                                "pos": "end",
                                "type": "locus",
                                "title": null
                            },
                            "color": {
                                "value": "#eeeeee"
                            }
                        },
                    },
                    {
                        "title": "Ybound",
                        "data": {
                            "values": axisLines,
                        },
                        "mark": {
                            "type": "rule",
                            "tooltip": {
                                "handler": "nullHandler"
                            },
                        },
                        "encoding": {
                            "y": {
                                "field": "axisLOD",
                                "type": "quantitative",
                                "title": null
                            },
                            "color": {
                                "value": "#e5e5e5"
                            }
                        },
                    },
                    {
                        "data": {
                            "values": newData
                        },
                        "title": "LODS",
                        "mark": {
                            "type": "point",
                            "tooltip": {
                                "handler": "mediationTooltipHandler"
                            },
                            "geometricZoomBound": 4,
                            "inwardStroke": true,
                            "stroke": "#233b5e",
                            "strokeWidth": 0.7,
                            "fillOpacity": 0.6,
                            "strokeOpacity": 0.8
                        },
                        "encoding": {
                            "x": {
                                "chrom": "chrom",
                                "pos": "position",
                                "type": "locus"
                            },
                            "y": {
                                "field": "LOD",
                                "type": "quantitative",
                                "title": "LOD",
                                "scale": { "domain": [minLOD, maxLOD] },
                                "axis": {
                                  "values": axisTicks
                                },
                            },
                            "color": {
                                "value": "#7090c0"
                            },
                            "size": {
                                "value": 100
                            },
                            "opacity": {
                                "value": 1
                            }
                        }
                    }
                ]
            };
            let tracks = [trackMediation];
            let handlers = {
                nullHandler: nullHandler,
                mediationTooltipHandler: mediationTooltipHandler
            };

            const spec = {
                "genome": {
                    "name": g.DATASETS[g.dataSetID].ensembl_species + " " + g.DATASETS[g.dataSetID].ensembl_release,
                    "contigs": g.DATASETS[g.dataSetID].chromosomes.contigs
                },
                "vconcat": tracks
            };

            logDebug("g.DATASETS[g.dataSetID].chromosomes.contigs=", g.DATASETS[g.dataSetID].chromosomes.contigs);
            logDebug('SPEC=', spec);

            $('#plotMediation').html('<div class="row"><div class="col" id="plotMediationMenu"></div></div><div class="row"><div class="col" id="plotMediationChart"></div></div>');
            $('#plotMediationChart').html('');
            $('#plotMediationMenu').html('');
            $('#plotMediationChart').height('500');

            $('#plotMediationMenu').append(`<div class="col text-center font-weight-bold">${plotTitle}
                <div class="dropdown dropleft float-right" id="mediationMenuDropdown">
                  <a class="text-secondary" href="#" role="button" id="mediationDropdownMenuLink" data-toggle="dropdown" aria-expanded="false">
                    <i class="fa-solid fa-bars"></i>
                  </a>

                  <div class="dropdown-menu" aria-labelledby="mediationDropdownMenuLink">
                    <a class="dropdown-item" id="mediationDownloadPNG" style="font-size:11px !important">Download PNG image</a>
                    <a class="dropdown-item" id="mediationDownloadCSV" style="font-size:11px !important">Download CSV</a>
                  </div>
                </div>
            </div>`);

            let n = g.DATASETS[g.dataSetID].display_name;
            $("#mediationDownloadPNG").click(function (event) {
                downloadGenomeSpy('#plotMediationChart', `${n}_MEDIATION.png`, '255,255,255');
            });

            $("#mediationDownloadCSV").click(function (event) {
                exportMediationData(currentID, newData);
            });


            g.chartMediation = genomeSpyEmbed.embed(
                document.querySelector("#plotMediationChart"),
                spec,
                {
                    tooltipHandlers: handlers
                }
            ).then(function(api) {
                api.addEventListener("click", function(event) {
                    if (event.datum != null) {
                        let datum = event.datum;
                        //logDebug('mediation clicked on', datum);
                    }
                });
            });
        }

        /**
         * Start downloading mediation data.
         *
         */
        function generateMediationPlot(id, markerID) {
            let mediateAgainst = $('#dsMediateAgainstID').val();
            let mediateURL = `{{ config.API_R_BASE }}/mediate?dataset=${g.dataSetID}&id=${id}&marker_id=${markerID}&dataset_mediate=${mediateAgainst}`;

            let submitData = {
                urls:[{
                    url_id: 'mediate',
                    url: mediateURL
                }]};

            startTask();

            // reset the chart and clear it
            g.chartMediation = null;
            $('#plotMediation').html('');

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=False) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                retries: 3,
                retryInterval: 1000,
                success: function(data, status, request) {
                    logInfo('Mediation');
                    logDebug('data = ', data);
                    updateMediationData(data.group_id, mediateAgainst);
                },
                error: function(jqXHR, textStatus, errorThrown) {
                    showErrorMessage(errorThrown, textStatus);
                }
            });
        }

        /**
         * Update the status of the downloading effect data.
         *
         * @param {string} groupID - the group identifier of the task
         */
        function updateMediationData(groupID, dsMediateAgainst) {
            // send GET request to status URL
            logDebug('updateMediationData{} ', groupID);

            if (g.runningTask) {
                let statusURL = '{{ url_for('api.api_status', group_id='', _external=False) }}' + groupID;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    retries: 3,
                    retryInterval: 1000,
                    success: function (data, status, request) {
                        logDebug('mediation data = ', data);

                        if (data.status === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data.number_tasks_errors !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.error}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data.number_tasks_errors === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if (value.status_code !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.response.error}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem performing the Mediation Analysis.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result, there will be 1 datasets to get
                                    logInfo('mediation');
                                    plotMediation(data.response_data.mediate.response.result, dsMediateAgainst);
                                    stopTask();
                                }
                            }
                        } else {
                            // rerun in 1 seconds
                            logDebug('Mediation not done, keep checking...');
                            setTimeout(function () {
                                updateMediationData(groupID, dsMediateAgainst);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                logDebug('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=False) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    logDebug(data);
                    g.chartMediation = null;
                    $('#plotMediation').html('');
                });
            }
        }

        /**
         * Start downloading effect data.
         *
         */
        function generateEffectPlot(id, chromosome, covar) {
            let effectURL = `{{ config.API_R_BASE }}/foundercoefficients?dataset=${g.dataSetID}&id=${id}&chrom=${chromosome}&intcovar=${covar}`;
            effectURL += '&blup=' + $('#performBLUP').is(':checked');

            let lodURL = `{{ config.API_R_BASE }}/lodscoresbycovar?dataset=${g.dataSetID}&id=${id}&chrom=${chromosome}&intcovar=${covar}`;

            {% if config.API_R_NUM_CORES %}
            effectURL += '&cores={{ config.API_R_NUM_CORES }}';
            lodURL += '&cores={{ config.API_R_NUM_CORES }}';
            {% endif %}

            let submitData = {
                urls:[{
                    url_id: 'effect',
                    url: effectURL,
                }]};

            if (covar !== 'additive') {
                submitData.urls.push({
                    url_id: 'lodSamples',
                    url: lodURL
                });
            }

            startTask();

            // reset the chart and clear it
            g.chartEffect = {};
            $('#allEffectPlots').html('');

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=False) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                retries: 3,
                retryInterval: 1000,
                success: function(data, status, request) {
                    updateEffectData(data.group_id, id, chromosome, covar);
                },
                error: function(jqXHR, textStatus, errorThrown) {
                    showErrorMessage(errorThrown, textStatus);
                }
            });
        }

        /**
         * Update the status of the downloading effect data.
         *
         * @param {string} groupID - the group identifier of the task
         */
        function updateEffectData(groupID, id, chromosome, covar) {
            // send GET request to status URL
            logDebug('updateEffectData() ', groupID);

            if (g.runningTask) {
                let statusURL = '{{ url_for('api.api_status', group_id='', _external=False) }}' + groupID;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    retries: 3,
                    retryInterval: 1000,
                    success: function (data, status, request) {
                        logDebug('DATA=======', data);
                        if (data.status === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data.number_tasks_errors !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.error}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data.number_tasks_errors === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if (value.status_code !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.response.error}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the Effect Plot.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result, there will be 1 datasets to get
                                    if (covar === 'additive') {
                                        plotEffectData(data.response_data.effect.response.result, null, chromosome, covar);
                                    } else {
                                        plotEffectData(data.response_data.effect.response.result,
                                                       data.response_data.lodSamples.response.result,
                                                       chromosome, covar);
                                    }
                                    stopTask();
                                }
                            }
                        } else {
                            // rerun in 1 seconds
                            logDebug('Not done, keep checking...');
                            setTimeout(function () {
                                updateEffectData(groupID, id, chromosome, covar);
                            }, 1000);
                        }
                    }
                });
            } else {
                // TODO: cleanup
                logDebug('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=False) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    logDebug(data);
                    g.chartEffect = null;
                    $('#allEffectPlots').html('');
                });
            }
        }

        function updateCorrelateDataSet(groupID) {
            // send GET request to status URL
            logDebug('updateCorrelateDataSet');
            logDebug(groupID);

            if (g.runningTask) {
                let statusURL = '{{ url_for('api.api_status', group_id='', _external=False) }}' + groupID;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    retries: 3,
                    retryInterval: 1000,
                    success: function (data, status, request) {
                        logDebug('DATA=======', data);
                        if (data.status === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data.number_tasks_errors !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.error}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data.number_tasks_errors === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if (value.status_code !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.response.error}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the Correlation Data.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    displayCorrelation(data.response_data.correlation.response.result);
                                    stopTask();
                                }
                            }
                        } else {
                            // rerun in 1 seconds
                            logDebug('Not done, keep checking...');
                            setTimeout(function () {
                                updateCorrelateDataSet(groupID);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                logDebug('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=False) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    logDebug(data);
                    g.chartCorrelation = null;
                    $('#correlationDataTable').html('');
                    $('#plotCorrelation').html('');
                });
            }

        }

        function switchCorrelationDataSet(dataSetID, interactiveCovariate) {
            $('#plotCorrelation').html('');
            $('#correlationDataTable').html('');
            g.chartCorrelation = null;
            g.correlationDatasetID = dataSetID;

            let currentID = '';

            if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                currentID = g.geneID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                currentID = g.proteinID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                currentID = g.uniprotID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                currentID = g.phosID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                currentID = g.phenotypeID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                currentID = g.ptmID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                currentID = g.peptideID;
            } else {
                // TODO: handle error
                logError('MAJOR PROBLEM');
            }


            let urlCorrelation = `{{ config.API_R_BASE }}/correlation?dataset=${g.dataSetID}&id=${currentID}&dataset_correlate=${dataSetID}`;

            if ((interactiveCovariate !== undefined) &&
                (interactiveCovariate !== null) &&
                (interactiveCovariate !== 'none')) {
                urlCorrelation += `&intcovar=${interactiveCovariate}`;
            }

            let submitData = {
                urls:[{
                    url_id: 'correlation',
                    url: urlCorrelation
                }]};

            startTask();

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=False) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                retries: 3,
                retryInterval: 1000,
                success: function(data, status, request) {
                    g.runningTask = true;
                    logDebug('data=', data);
                    logDebug('status=', status);
                    logDebug('request=', request);
                    logDebug('data.group_id=', data.group_id);
                    updateCorrelateDataSet(data.group_id);
                },
                error: function(jqXHR, textStatus, errorThrown) {
                    showErrorMessage(errorThrown, textStatus);
                }
            });
        }

        /**
         * Switch the datasets.
         *
         * @param {string} dataSetID - the dataSet identifier
         */
        function switchDataSet(dataSetID) {
            g.dataSetID = dataSetID;

            let html = `<div class="col">
                            <div class="jumbotron jumbotron-fluid" style="background: none">
                                <div class="container">
                                    <h1 class="display-3">{{ config.MAIN_TITLE|safe }}</h1>
                                    <p class="lead">Please search for a term of interest.</p>
                                </div>
                            </div>
                        </div>`;

            $('#divWelcomeMesage').html(html);

            $('#divItemInfo').addClass('invisible').removeClass('visible');
            $('#divLOD').addClass('invisible').removeClass('visible');
            $('#divSecondRow').addClass('invisible').removeClass('visible');
            $('#divProfileCorrelation').addClass('invisible').removeClass('visible');
            $('#div_current_item_information_header').html('');
            $('#div_current_item_information_body').html('');
            $('#divProfilePlotOptions').html('');

            configureCovarInfo(true);

            setDataSet(dataSetID, true);
        }

        /**
         * Display the gene information.
         * @param {Object} geneData - gene information
         */
        function setDataSet(dataSetID, force) {
            logDebug('setDataSet=', dataSetID);

            if (!(force) && (QTL.g.dataSetID === dataSetID)) {
                logDebug('Same dataset selected, do nothing');
                return;
            }

            if (!isEmpty(g.DATASETS[g.dataSetID].lodpeaks)) {
                generateLODPeaksHTML();
            }


            let ds = g.DATASETS[dataSetID];
            g.dataSetID = dataSetID;

            $('#divGenome').html(`<b>${ds.display_name}</b> is utilizing Ensembl Release <b>${ds.ensembl.release}</b>, Genome <b>${ds.ensembl.assembly_patch}</b> <div style='height: 10px'></div>`);

            // clear LOD Plot
            // TODO: clear LOD PLOT, Profile PLOT
            g.chartLOD = null;
            g.chartLODCovariateFull = null;
            g.chartLODCovariateDiff = null;
            g.chartCorrelation = null;
            g.chartEffect = null;
            $('#lodPlotChart').html('');
            $('#lodPlotChartCovariateFull').html('');
            $('#lodPlotChartCovariateDiff').html('');
            $('#correlationCovariateSelectionText').addClass('hidden').removeClass('visible');
            $('#correlationImputation').html('');

            resetSecondaryPlots();

            // make tabs invisible
            $('#navGene').hide();
            $('#navPheno').hide();
            if (isEmpty(g.DATASETS[g.dataSetID].lodpeaks)) {
                $('#navLodPeaks').hide();
            } else {
                $('#navLodPeaks').show();
            }


            // show correlation
            if (isEmpty(ds['covar_info'])) {
                $('#profilePlot a[href="#tabCorrelation"]').tab('show');
                $('#navProfile').hide();
            } else {
                $('#profilePlot a[href="#tabProfile"]').tab('show');
                $('#navProfile').show();
            }


            // show the correct tab
            if ((ds.datatype === 'mrna') || (ds.datatype === 'protein') || (ds.datatype === 'phos')) {
                $('#navGene').show();
                $('#navGene').tab('show');

                // show some secondary plots
                $('#navPlotMediation').show();

                // reset the screen
                $('#searchResultsDiv').html('');
                $('#searchResultsTableInfo').html('');

            } else if (ds.datatype === 'pheno') {
                $('#navPheno').show();
                $('#navPheno').tab('show');

                // hide some secondary plots

                // phenotype data can only mediate against mrna or protein (not itself)
                let showMed = false;
                $.each(QTL.g.DATASETS, function(k, v) {
                    if ((v.datatype === 'mrna') || (v.datatype === 'protein') || (v.datatype === 'phos')) {
                        showMed = true;
                    }
                });

                if (showMed) {
                    $('#navPlotMediation').show();
                } else {
                    $('#navPlotMediation').hide();
                }

                $('#searchPheno').html('');
                $('#searchPhenoCategoryText').html('');
                $('#searchPhenoCategory').html('');
                $('#searchPhenoResultsInfo').html('');
                $('#searchPhenoResultsDiv').html('');

                //let numResults = ds.phenotypes.length;

                // transform the phenotypes into a different structure
                let categories = new Set([]);
                let newPhenotypes = [];
                $.each(ds.phenotypes, function(index, element) {
                    if ((element['is_pheno']) && (element['is_numeric'])) {
                        categories.add(element.category);
                        element.display = element['short_name'];
                        newPhenotypes.push({phenotype_id: element['data_name'], // use for id
                                            phenotype: element['short_name'],   // display on screen
                                            category: element.category,
                                            desc: element.description});
                    }
                });

                // build the search area
                if (categories.size === 0) {
                    logError('No categories - ERROR');
                } else {
                    $('#searchPheno').html('<input type="text" id="searchTermPheno" name="searchTermPheno" value="{{ search_term }}" class="form-control" placeholder="Please enter a phenotype to filter results...">');

                    // build the phenotypes table
                    $('#searchPhenoResultsDiv').html(`<table id="phenotypesTable" class="table table-striped table-hover table-sm table-bordered">
                                                <thead>
                                                    <th style="background-color: #20a8d8" data-dynatable-sorts="Phenotype">Phenotype</th>
                                                    <th style="display: none">category</th>
                                                    <th style="display: none">desc</th>
                                                </thead>
                                                <tbody>
                                                </tbody>
                                            </table>`);

                    g.phenoDynaTable = $('#phenotypesTable')
                        .bind('dynatable:init', function(e, dynatable) {
                            dynatable.queries.functions.phenotype = function(record, queryValue) {
                                if ((record.phenotype.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1) ||
                                    (record.desc.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1)) {
                                    return true;
                                }
                            };
                        }).dynatable({
                            dataset: {
                                records: newPhenotypes,
                            },
                            features: {
                                paginate: false,
                                pushState: false,
                                recordCount: true,
                                sort: true,
                                search: false,
                            },
                            inputs: {
                                recordCountPlacement: 'before',
                            },
                            writers: {
                                _rowWriter: function(rowIndex, record, columns, cellWriter) {
                                    return `<tr><td><span class="font-weight-bold"><a href="#" phenotypeid="${record.phenotype_id}">${record.phenotype}</a></span><br><span class="font-italic">${record.category}</span><br><span class="font-weight-light">${record.desc}</span></td></tr>`;
                                }
                            },
                        }).data('dynatable');

                    // build the category select, if more than 1 category
                    categories = Array.from(categories);

                    if (categories.length > 1) {
                        $('#searchPhenoCategoryText').html('Category Filter');

                        let htmlOptions = '<select id="phenoCategorySelect" data-style="btn-secondary btn-sm" data-width="100%" class="selectpicker">';
                        htmlOptions += '<option>All</option>';
                        $.each(categories, function (i, val) {
                            htmlOptions += `<option>${val}</option>`;
                        });
                        htmlOptions += '</select>';
                        $('#searchPhenoCategory').html(htmlOptions);

                        $('#phenoCategorySelect').selectpicker();

                        $('#phenoCategorySelect').on('changed.bs.select', function (e, clickedIndex, isSelected, previousValue) {
                            let selected = $(this).find(':selected').val();
                            if (selected === 'All') {
                                g.phenoDynaTable.queries.remove('category');
                            } else {
                                g.phenoDynaTable.queries.add('category', selected);
                            }

                            g.phenoDynaTable.process();
                            $('#phenotypesTable tbody a').on('click', function (evt) {
                                evt.preventDefault();
                                let that = $(this);
                                logDebug('click from categories > 1');
                                selectPhenotype(that.attr('phenotypeid'), g.dataSetID, $('#interactiveCovarLODS').val());
                                return false;
                            });
                        });
                    }

                    $('#searchTermPheno').keyup(function (event) {
                        if ( event.which === 13 ) {
                            event.preventDefault();
                        }
                        let q = $(this).val();

                        if (q === '') {
                            q = undefined;
                        }

                        if (q) {
                            g.phenoDynaTable.queries.add('phenotype', q);
                        } else {
                            logDebug('removing');
                            g.phenoDynaTable.queries.remove('phenotype');
                        }

                        logDebug('g.phenoDynaTable.process()');
                        g.phenoDynaTable.process();
                        $('#phenotypesTable tbody a').on('click', function (evt) {
                            evt.preventDefault();
                            let that = $(this);
                            logDebug('click from keyup');
                            selectPhenotype(that.attr('phenotypeid'), g.dataSetID, $('#interactiveCovarLODS').val());
                            return false;
                        });

                    });

                    $('#phenotypesTable tbody a').on('click', function (evt) {
                        evt.preventDefault();
                        let that = $(this);
                        logDebug('click from datatype is pheno');
                        selectPhenotype(that.attr('phenotypeid'), g.dataSetID, $('#interactiveCovarLODS').val());
                        return false;
                    });
                }

            } else if ((ds.datatype === 'ptm') || (ds.datatype === 'peptide')) {
                $('#navPheno').show();
                $('#navPheno').tab('show');

                // show some secondary plots
                $('#navPlotMediation').show();
                //$('#navPlotMediation').hide();

                $('#searchPheno').html('');
                $('#searchPhenoCategoryText').html('');
                $('#searchPhenoCategory').html('');
                $('#searchPhenoResultsInfo').html('');
                $('#searchPhenoResultsDiv').html('');

                logDebug('ds.datatype=', ds.datatype);

                // transform the elements into a different structure
                let categories = new Set([]);
                let newElements = [];
                let elems = ds.ptms;
                if (ds.datatype === 'peptide') {
                    elems = ds.peptides;
                }
                logDebug('elems=', elems);
                $.each(elems, function(index, element) {
                    categories.add(element?.ptm || '');
                    //logDebug('PTM=', element?.ptm || '');
                    element.display = element?.ptm_id || element?.peptide_id;
                    newElements.push({
                        element_id: element?.ptm_id || element?.peptide_id,
                        category: element?.ptm || '',
                        gene_id: element?.gene_id || '',
                        gene_symbol: element?.gene_symbol || '',
                        protein_id: element?.protein_id || '',
                        peptide_id: element?.peptide_id || '',
                        ptm_id: element?.ptm_id || '',
                        uniprot_id: element?.uniprot_id || ''
                    });
                });
                logDebug('newElements=', newElements);
                logDebug('categories=', categories);

                // build the search area
                if (categories.size === 0) {
                    logError('No categories - ERROR');
                } else {
                    $('#searchPheno').html('<input type="text" id="searchTermPheno" value="{{ search_term }}" name="searchTermPheno" class="form-control" placeholder="Please enter text to filter results...">');

                    // build the phenotypes table
                    $('#searchPhenoResultsDiv').html(`<table id="phenotypesTable" class="table table-striped table-hover table-sm table-bordered">
                                                <thead>
                                                    <th style="background-color: #20a8d8" data-dynatable-sorts="Phenotype">Data</th>
                                                    <th style="display: none">category</th>
                                                    <th style="display: none">gene_id</th>
                                                    <th style="display: none">gene_symbol</th>
                                                    <th style="display: none">protein_id</th>
                                                    <th style="display: none">peptide_id</th>
                                                    <th style="display: none">ptm_id</th>
                                                    <th style="display: none">uniprot_id</th>
                                                </thead>
                                                <tbody>
                                                </tbody>
                                            </table>`);

                    g.phenoDynaTable = $('#phenotypesTable')
                        .bind('dynatable:init', function(e, dynatable) {
                            dynatable.queries.functions.phenotype = function(record, queryValue) {
                                if ((record.element_id.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1) ||
                                    (record.category.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1) ||
                                    (record.gene_id.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1) ||
                                    (record.gene_symbol.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1) ||
                                    (record.protein_id.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1) ||
                                    (record.peptide_id.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1) ||
                                    (record.uniprot_id.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1)) {

                                    return true;
                                }
                            };
                        }).dynatable({
                            dataset: {
                                records: newElements,
                            },
                            features: {
                                paginate: false,
                                pushState: false,
                                recordCount: true,
                                sort: true,
                                search: false,
                            },
                            inputs: {
                                recordCountPlacement: 'before',
                            },
                            writers: {
                                _rowWriter: function(rowIndex, record, columns, cellWriter) {
                                    let r = `<tr>
                                                <td>`;


                                    if (record.ptm_id !== '') {
                                           r+= `         <span class="font-weight-bold"><small><a href="#" geneid="${record.gene_id}" proteinid="${record.protein_id}" peptideid="${record.peptide_id}" ptmid="${record.element_id}">${record.element_id}</a></small></span>
                                                    <br>
                                                    `;
                                    } else {
                                           r += `         <span class="font-weight-bold"><small><a href="#" geneid="${record.gene_id}" proteinid="${record.protein_id}" peptideid="${record.element_id}">${record.element_id}</a></small></span>
                                                    <br>
                                                    `;
                                    }

                                    if (record.gene_id !== '') {
                                        r += `<span class="font-italic small">Gene ID: ${record.gene_id} ${record.gene_symbol}</span><br>`;
                                    }
                                    if (record.protein_id !== '') {
                                        r += `<span class="font-italic small">Protein ID: ${record.protein_id}</span><br>`;
                                    }

                                    r += `<span class="font-italic small">Uniprot: ${record.uniprot_id}</span>
                                          <br>
                                                    <span class="font-weight-light small">Peptide: ${record.peptide_id}</span></td></tr>
                                                    `;

                                    return r;
                                }
                            },
                        }).data('dynatable');

                    // build the category select, if more than 1 category
                    categories = Array.from(categories);
                    categories.sort();

                    if (categories.length > 1) {
                        $('#searchPhenoCategoryText').html('Category Filter');

                        let htmlOptions = '<select id="phenoCategorySelect" data-style="btn-secondary btn-sm" data-width="100%" class="selectpicker">';
                        htmlOptions += '<option>All</option>';
                        $.each(categories, function (i, val) {
                            htmlOptions += `<option>${val}</option>`;
                        });
                        htmlOptions += '</select>';
                        $('#searchPhenoCategory').html(htmlOptions);

                        $('#phenoCategorySelect').selectpicker();

                        $('#phenoCategorySelect').on('changed.bs.select', function (e, clickedIndex, isSelected, previousValue) {
                            let selected = $(this).find(':selected').val();
                            if (selected === 'All') {
                                g.phenoDynaTable.queries.remove('category');
                            } else {
                                g.phenoDynaTable.queries.add('category', selected);
                            }

                            g.phenoDynaTable.process();
                            $('#phenotypesTable tbody a').on('click', function (evt) {
                                evt.preventDefault();
                                let that = $(this);
                                console.log('click on tbody a selectPTMPeptide cat > 1', that.attr('ptmid'), that.attr('peptideid'), that.attr('proteinid'), that.attr('geneid'));
                                selectPTMPeptide(that.attr('ptmid'), that.attr('peptideid'), that.attr('proteinid'), that.attr('geneid'), g.dataSetID, $('#interactiveCovarLODS').val());
                                return false;
                            });
                        });
                    }

                    $('#searchTermPheno').keyup(function (event) {
                        logDebug('event=', event);
                        logDebug('event.which=', event.which);
                        //if ( event.which === 13 ) {
                        //    event.preventDefault();
                        //}
                        let q = $(this).val();


                        if (q === '') {
                            q = undefined;
                        }

                        if (q) {
                            g.phenoDynaTable.queries.add('phenotype', q);
                        } else {
                            logDebug('removing');
                            g.phenoDynaTable.queries.remove('phenotype');
                        }

                        logDebug('g.phenoDynaTable.process()');
                        g.phenoDynaTable.process();
                        $('#phenotypesTable tbody a').on('click', function (evt) {
                            evt.preventDefault();
                            let that = $(this);
                            console.log('click on tbody a selectPTMPeptide keyup', that.attr('ptmid'), that.attr('peptideid'), that.attr('proteinid'), that.attr('geneid'));
                            selectPTMPeptide(that.attr('ptmid'), that.attr('peptideid'), that.attr('proteinid'), that.attr('geneid'), g.dataSetID, $('#interactiveCovarLODS').val());
                            return false;
                        });

                    });

                    $('#phenotypesTable tbody a').on('click', function (evt) {
                        evt.preventDefault();
                        let that = $(this);
                        console.log('click on tbody a selectPTMPeptide all', that.attr('ptmid'), that.attr('peptideid'), that.attr('proteinid'), that.attr('geneid'));
                        selectPTMPeptide(that.attr('ptmid'), that.attr('peptideid'), that.attr('proteinid'), that.attr('geneid'), g.dataSetID, $('#interactiveCovarLODS').val());
                        return false;
                    });
                }
            } else if ((ds.datatype === 'protein_uniprot')) {
                $('#navPheno').show();
                $('#navPheno').tab('show');

                // show some secondary plots
                $('#navPlotMediation').show();
                // $('#navPlotMediation').hide();

                $('#searchPheno').html('');
                $('#searchPhenoCategoryText').html('');
                $('#searchPhenoCategory').html('');
                $('#searchPhenoResultsInfo').html('');
                $('#searchPhenoResultsDiv').html('');

                logDebug('ds.datatype=', ds.datatype);

                // transform the elements into a different structure
                let categories = new Set([]);
                let newElements = [];
                let elems = ds.uniprot_ids;
                logDebug('elems=', elems);
                $.each(elems, function(index, element) {
                    categories.add('');
                    element.display = element.uniprotID;
                    newElements.push({
                        uniprot_id: element.uniprot_id || '',
                        protein_id: element?.protein_id || '',
                        gene_id: element?.gene_id || '',
                        gene_symbol: element?.gene_symbol || '',
                    });
                });
                logDebug('newElements=', newElements);
                logDebug('categories=', categories);

                // build the search area
                if (categories.size === 0) {
                    logError('No categories - ERROR');
                } else {
                    $('#searchPheno').html('<input type="text" id="searchTermPheno" value="{{ search_term }}" name="searchTermPheno" class="form-control" placeholder="Please enter text to filter results...">');

                    // build the phenotypes table
                    $('#searchPhenoResultsDiv').html(`<table id="phenotypesTable" class="table table-striped table-hover table-sm table-bordered">
                                                <thead>
                                                    <th style="background-color: #20a8d8" data-dynatable-sorts="Phenotype">Data</th>
                                                    <th style="display: none">uniprot_id</th>
                                                    <th style="display: none">protein_id</th>
                                                    <th style="display: none">gene_id</th>
                                                    <th style="display: none">gene_symbol</th>
                                                </thead>
                                                <tbody>
                                                </tbody>
                                            </table>`);

                    g.phenoDynaTable = $('#phenotypesTable')
                        .bind('dynatable:init', function(e, dynatable) {
                            dynatable.queries.functions.phenotype = function(record, queryValue) {
                                if ((record.gene_id.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1) ||
                                    (record.gene_symbol.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1) ||
                                    (record.protein_id.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1) ||
                                    (record.uniprot_id.toLowerCase().indexOf(queryValue.toLowerCase()) !== -1)) {

                                    return true;
                                }
                            };
                        }).dynatable({
                            dataset: {
                                records: newElements,
                            },
                            features: {
                                paginate: false,
                                pushState: false,
                                recordCount: true,
                                sort: true,
                                search: false,
                            },
                            inputs: {
                                recordCountPlacement: 'before',
                            },
                            writers: {
                                _rowWriter: function(rowIndex, record, columns, cellWriter) {
                                    let r = `<tr>
                                                <td>`;

                                           r+= `         <span class="font-weight-bold"><small><a href="#" geneid="${record.gene_id}" proteinid="${record.protein_id}" uniprotid="${record.uniprot_id}">${record.uniprot_id}</a></small></span>
                                                    <br>
                                                    `;

                                    if (record.protein_id !== '') {
                                        r += `<span class="font-italic small">Protein ID: ${record.protein_id}</span><br>`;
                                    }
                                    if (record.gene_id !== '') {
                                        r += `<span class="font-italic small">Gene ID: ${record.gene_id} ${record?.gene_symbol}</span><br>`;
                                    }

                                    r += '</td></tr>';


                                    return r;
                                }
                            },
                        }).data('dynatable');

                    // build the category select, if more than 1 category
                    categories = Array.from(categories);
                    categories.sort();

                    if (categories.length > 1) {
                        $('#searchPhenoCategoryText').html('Category Filter');

                        let htmlOptions = '<select id="phenoCategorySelect" data-style="btn-secondary btn-sm" data-width="100%" class="selectpicker">';
                        htmlOptions += '<option>All</option>';
                        $.each(categories, function (i, val) {
                            htmlOptions += `<option>${val}</option>`;
                        });
                        htmlOptions += '</select>';
                        $('#searchPhenoCategory').html(htmlOptions);

                        $('#phenoCategorySelect').selectpicker();

                        $('#phenoCategorySelect').on('changed.bs.select', function (e, clickedIndex, isSelected, previousValue) {
                            let selected = $(this).find(':selected').val();
                            if (selected === 'All') {
                                g.phenoDynaTable.queries.remove('category');
                            } else {
                                g.phenoDynaTable.queries.add('category', selected);
                            }

                            g.phenoDynaTable.process();
                            $('#phenotypesTable tbody a').on('click', function (evt) {
                                evt.preventDefault();
                                let that = $(this);
                                selectUniprot(that.attr('uniprotid'), that.attr('proteinid'),that.attr('geneid'), g.dataSetID, $('#interactiveCovarLODS').val());
                                return false;
                            });
                        });
                    }

                    $('#searchTermPheno').keyup(function (event) {
                        if ( event.which === 13 ) {
                            event.preventDefault();
                        }
                        let q = $(this).val();

                        if (q === '') {
                            q = undefined;
                        }

                        if (q) {
                            g.phenoDynaTable.queries.add('phenotype', q);
                        } else {
                            logDebug('removing');
                            g.phenoDynaTable.queries.remove('phenotype');
                        }

                        logDebug('g.phenoDynaTable.process()');
                        g.phenoDynaTable.process();
                        $('#phenotypesTable tbody a').on('click', function (evt) {
                            evt.preventDefault();
                            let that = $(this);
                            selectUniprot(that.attr('uniprotid'), that.attr('proteinid'), that.attr('geneid'), g.dataSetID, $('#interactiveCovarLODS').val());
                            return false;
                        });

                    });

                    $('#phenotypesTable tbody a').on('click', function (evt) {
                        evt.preventDefault();
                        let that = $(this);
                        selectUniprot(that.attr('uniprotid'), that.attr('proteinid'), that.attr('geneid'), g.dataSetID, $('#interactiveCovarLODS').val());
                        return false;
                    });
                }
            } else {
                // TODO: handle error
                logDebug('ERROR in setDataSet:', ds);
            }


            // hide or show the mediation based upon if there are datasets that can mediate against
            let allDS = [];
            $.each(g.DATASETS, function(key, value) {
                if (value.datatype !== 'pheno') {
                    allDS.push(value);
                }
            });

            $('#divMediationSelect').html('');

            if (allDS.length === 1) {
                $('#divMediationSelect').html(`<p class="font-weight-bold">Mediating against: </p>${allDS[0]['display_name']}`);
                $('#divMediationSelect').append(`<input type="hidden" id="dsMediateAgainstID" value="${allDS[0].id}"/>`);
            } else if (allDS.length > 1) {

                let htmlMediate = `<div class="row">
                                       <div class="col-auto align-self-center font-weight-bold">
                                           Mediate Against
                                      </div>
                                      <div class="col">
                                          <select id="dsMediateAgainstID" data-style="btn-secondary btn-sm" data-width="100%" class="selectpicker">`;

                $.each(allDS, function(idx, elem) {
                    logDebug(idx, elem);
                    if (elem.id === dataSetID) {
                        htmlMediate += `<option selected value="${elem.id}">${elem['display_name']}</option>`;
                    } else {
                        htmlMediate += `<option value="${elem.id}">${elem['display_name']}</option>`;
                    }
                });
                htmlMediate += '</select></div></div>';

                $('#divMediationSelect').html(htmlMediate);
                $('#dsMediateAgainstID').selectpicker();
            }
        }

        function exportProfileData(id, data, factors) {
            // TODO: look into changing 'column.name', 'display_name'
            let csvContent = '"sample_id","expression"';
            $.each(factors, function(k, v) {
                csvContent += `,"${v['display_name']}"`;
            });
            csvContent += '\n';

            $.each(data.data, function(k, v) {
                let line = `"${v['sample_id']}",${v.expression}`;
                $.each(factors, function(fk, fv) {
                    line += `,"${v[fv['sample_column']]}"`;
                });

                csvContent += `${line}\n`;
            });

            downloadCSV(csvContent, `${id}_PROFILE.csv`, 'text/csv;encoding:utf-8');
        }


        /**
         * Plot profile data.
         * @param {string} id - the unique identifier
         */
        function plotProfile(id) {
            logDebug(`plotProfile(${id})`);

            let data = g.expressionData;

            let ds = g.DATASETS[g.dataSetID];

            // categories will be the order the user selected
            let selectedCovars = null;

            if (isEmpty(ds['covar_info'])) {
                selectedCovars = null;
            } else {
                selectedCovars = getFactorOrder();
            }

            // categorySeries will be how to color
            let seriesToColor = $('#factorSeries').find(':selected').val();

            logDebug('selectedCovars=', selectedCovars);
            logDebug('seriesToColor=', seriesToColor);

            if ((selectedCovars === null) || (selectedCovars.length === 0)) {
                // this is the case when there are no covar factors

                // do we just have 1 series of data here?
            } else {
                // just mapping column name to display name
                let cfMap = {};

                $.each(g.DATASETS[g.dataSetID]['covar_info'], function (idx, elem) {
                    cfMap[elem['sample_column']] = elem['display_name'];
                });

                logDebug('cfMap=', cfMap);

                let title = [];

                let covarPermutations = [];

                logDebug('data=', data);

                $.each(selectedCovars, function (idx, elem) {
                    covarPermutations.push(data.datatypes[elem]);
                    title.push(cfMap[elem]);
                });

                let chartTitle = title.join(' x ');
                covarPermutations = permutateArrays(covarPermutations);

                logDebug('covarPermutations=', covarPermutations);

                let covarCategories = [];
                let covarCategoriesMap = new OrderedDict();
                let categoryData = new OrderedDict();  // for box plot

                $.each(covarPermutations, function (idx, elem) {
                    logDebug(idx, elem);
                    let catName = elem.join(':');
                    logDebug(catName);
                    covarCategories.push(catName);
                    covarCategoriesMap.set(catName, idx);
                    categoryData.set(catName, []);
                });

                logDebug('covarCategories=', covarCategories);
                logDebug('covarCategoriesMap=', covarCategoriesMap);

                let seriesData = new OrderedDict();

                $.each(data.datatypes[seriesToColor], function (idx, val) {
                    console.log('seriesData', val);
                    seriesData.set(val, {values: []});
                });

                $.each(data.data, function (idx, val) {
                    if (val.expression !== undefined) {
                        let t = [];

                        $.each(selectedCovars, function (i, v) {
                            t.push(val[v]);
                        });

                        let catName = t.join(':');

                        let bb = seriesData.get(val[seriesToColor]);
                        bb.values.push({
                            x: covarCategoriesMap.get(catName),
                            y: val.expression,
                            obj: val
                        });
                        seriesData.set(val[seriesToColor], bb);

                        //seriesData[].values.push();

                        let dd = categoryData.get(catName);
                        dd.push(val.expression);
                        categoryData.set(catName, dd);
                    }
                });

                logDebug('seriesData=', seriesData);

                // convert from object, to array

                let series = [];
                let colors = ['#8B8682', '#87CEEB', '#27408B', '#FFA500', '#B22222', '#f7a35c', '#90ed7d', '#e4d354', '#2b908f', '#f45b5b'];

                let boxSeries = [];

                seriesData.forEach((value, key) => {
                    logDebug('using color', colors[series.length % colors.length]);
                    logDebug('using color', series.length % colors.length);
                    series.push({
                        data: value.values,
                        name: key,
                        marker: {
                            radius: 2
                        },
                        color: colors[series.length % colors.length],
                        jitter: {
                            x: 0.24
                        },
                        zIndex: 10
                    });
                });

/*
                $.each(seriesData, function (key, val) {
                    logDebug('using color', colors[series.length % colors.length]);
                    logDebug('using color', series.length % colors.length);
                    series.push({
                        data: val.values,
                        name: key,
                        marker: {
                            radius: 2
                        },
                        color: colors[series.length % colors.length],
                        jitter: {
                            x: 0.24
                        },
                        zIndex: 10
                    });
                });
   */
                categoryData.forEach((value, key) => {
                    if (value.length > 0) {
                        let boxValues = getBoxValues(value);

                        boxValues.x = covarCategoriesMap.get(key);
                        boxValues.name = key;
                        boxValues.enableMouseTracking = false;
                        boxSeries.push(boxValues);
                    }
                });

/*
                $.each(categoryData, function (key, val) {
                    if (val.length > 0) {
                        let boxValues = getBoxValues(val);

                        boxValues.x = covarCategoriesMap.get(key);
                        boxValues.name = key;
                        boxValues.enableMouseTracking = false;
                        boxSeries.push(boxValues);
                    }
                });
*/
                series.push({
                    type: 'boxplot',
                    data: boxSeries,
                    showInLegend: false,
                    zIndex: 1,
                });

                let exporting = appendExportButton('Download CSV', function () {
                    exportProfileData(id, data, g.DATASETS[g.dataSetID]['covar_info']);
                });
                exporting.filename = id + '_PROFILE';

                Highcharts.chart({
                    chart: {
                        type: 'scatter',
                        renderTo: 'profilePlotChart',
                    },

                    title: {
                        text: chartTitle,
                        verticalAlign: 'bottom'
                    },

                    xAxis: {
                        // min and max forces the "categories" to be shown even when no data is available
                        //min: 0,
                        //max: point[point.length-1] + jitter + jitterPad,
                        //tickPositions: point,
                        categories: covarCategories,
                        labels: {
                            rotation: -45
                        }
                    },

                    yAxis: {
                        title: {
                            text: id
                        }
                    },

                    series: series,

                    exporting: exporting,

                    plotOptions: {
                        boxplot: {
                            color: '#000000',
                            enableMouseTracking: false,
                            lineWidth: 1,
                            medianColor: '#000000',
                            medianWidth: 1,
                            stemColor: '#000000',
                            stemWidth: 1,
                            whiskerColor: '#000000',
                            whiskerLength: '50%',
                            whiskerWidth: 1
                        },
                        scatter: {
                            tooltip: {
                                headerFormat: null,
                                outside: true,
                                pointFormatter: function () {
                                    let val = this.obj;
                                    let hoverText = `<b>${val['sample_id']}</><br><p>${val.expression}</p>`;

                                    $.each(data.datatypes, function (i, e) {
                                        hoverText += `<br><p>${cfMap[i]}: ${val[i]}</p>`;
                                    });

                                    return hoverText;
                                }


                            }
                        }
                    }
                });
            }

        }

        /**
         * Perform gene search.
         */
        function findGene(searchVal) {
            //let button = $('#btnGo');
            let term = $('#searchTerm');
            term.disable(true);
            //button.button('loading');

            if (searchVal.length === 0) {
                term.focus();
                term.disable(false);
                //button.button('reset');
                return;
            }

            let options = {
                species: g.DATASETS[g.dataSetID].ensembl_species,
                release: g.DATASETS[g.dataSetID].ensembl_release,
                limit: 100,
                greedy: true
            };

            logDebug(options);

            startTask();
            g.ENSIMPL.search(searchVal, options, searchCallback);
        }

        /**
         * Populate the search results.
         */
        function searchCallback() {
            logDebug(g.ENSIMPL);

            if (g.ENSIMPL.response.result.matches === null) {
                $('#searchResultsDiv').html('');
                $('#btnGo').button('reset');
                $('#searchTerm').disable(false);
                $('#searchResultsTableInfo').html('No results found');
                stopTask();
                return;
            }

            let response = g.ENSIMPL.response;
            let currentDataSet = g.DATASETS[g.dataSetID];
            let searchResults = [];
            let searchResultsObj = {};


            $.each(response.result.matches, function(idx, match) {

                match['match'] = match.ensembl_gene_id in currentDataSet.gene_ids;
                match['datatype'] = currentDataSet.datatype;

                if ((currentDataSet.datatype === 'protein') && (match.match)) {
                    match['protein_ids'] = currentDataSet.gene_ids[match.ensembl_gene_id].protein_ids;
                } else if ((currentDataSet.datatype === 'protein_uniprot')  && (match.match)) {
                    match['protein_ids'] = currentDataSet.gene_ids[match.ensembl_gene_id].protein_ids;
                } else if ((currentDataSet.datatype === 'phos')  && (match.match)) {
                    match['protein_ids'] = currentDataSet.gene_ids[match.ensembl_gene_id].protein_ids;
                }

                searchResults.push(match);
                searchResultsObj[match.ensembl_gene_id] = match;
            });

            // build the results table
            $('#searchResultsDiv').html(`<table id="searchResultsTable" class="table table-striped table-hover table-sm table-bordered" style="font-size:0.85rem">
                                        <thead>
<th data-dynatable-no-sort></th>
                                            <th data-dynatable-sorts="ID">ID</th>
<th >Symbol</th>
<!--
                                            <th style="display: none">category</th>
                                            <th style="display: none">desc</th>
//--->
                                        </thead>
                                        <tbody>
                                        </tbody>
                                    </table>`);

            g.searchTable = $('#searchResultsTable')
                .dynatable({
                    dataset: {
                        records: searchResults,
                    },
                    features: {
                        paginate: false,
                        pushState: false,
                        recordCount: true,
                        sort: false,
                        search: false,
                    },
                    inputs: {
                        recordCountPlacement: 'before',
                    },
                    writers: {
                        _rowWriter: function(rowIndex, record, columns, cellWriter) {
                            let html = `<tr><td><i id="matchInfo" geneID="${record.ensembl_gene_id}" class="fas fa-info-circle"></i></td>`;

                            if (record.datatype === 'mrna') {
                                if (record.match) {
                                    html += `<td><span><a href="#" geneID="${record.ensembl_gene_id}">${record.ensembl_gene_id}</a></span></td>`;
                                } else {
                                    html += `<td><span>${record.ensembl_gene_id}</span></td>`;
                                }

                            } else if (record.datatype === 'protein') {
                                html += `<td style="white-space: nowrap;"><span>${record.ensembl_gene_id}</span>`;

                                if (record.match) {
                                    html += '<br/>';
                                    let proteins = [];
                                    for (let p in record.protein_ids) {
                                        proteins.push(`<span style="padding-left: 5px;"><i class="fa-solid fa-turn-up fa-rotate-90"></i> <a href="#" geneID="${record.ensembl_gene_id}" proteinID="${record.protein_ids[p]}">${record.protein_ids[p]}</a></span>`);
                                    }
                                    html += proteins.join('<br/>');
                                }

                                html += '</td>';
                            } else if (record.datatype === 'protein_uniprot') {
                                html += `<td style="white-space: nowrap;"><span>${record.ensembl_gene_id}</span>`;

                                if (record.match) {
                                    html += '<br/>';
                                    let proteins = [];
                                    for (let p in record.protein_ids) {
                                        let protein_id = record.protein_ids[p];

                                        let phos = [];
                                        for (let j in currentDataSet.protein_ids[protein_id]?.uniprot_ids) {
                                            let ph_id = currentDataSet.protein_ids[protein_id].uniprot_ids[j];
                                            phos.push(`<span style="padding-left: 25px;"><i class="fa-solid fa-turn-up fa-rotate-90"></i> <a href="#" geneID="${record.ensembl_gene_id}" proteinID="${protein_id}" uniprotID="${ph_id}">${ph_id}</a>`);
                                        }

                                        proteins.push(`<span style="padding-left: 5px;"><i class="fa-solid fa-turn-up fa-rotate-90"></i> ${protein_id}</span><br/>` + phos.join('<br/>'));
                                    }
                                    html += proteins.join('<br/>');
                                }



                                html += '</td>';
                            } else if (record.datatype === 'phos') {
                                html += `<td style="white-space: nowrap;"><span>${record.ensembl_gene_id}</span>`;

                                if (record.match) {
                                    html += '<br/>';
                                    let proteins = [];
                                    for (let p in record.protein_ids) {
                                        let protein_id = record.protein_ids[p];

                                        let phos = [];
                                        for (let j in currentDataSet.protein_ids[protein_id]?.phos_ids) {
                                            let ph = currentDataSet.protein_ids[protein_id].phos_ids[j];
                                            let ph_id = ph.split('_')[1];
                                            phos.push(`<span style="padding-left: 25px;"><i class="fa-solid fa-turn-up fa-rotate-90"></i> <a href="#" geneID="${record.ensembl_gene_id}" proteinID="${protein_id}" phosID="${ph}">${ph_id}</a>`);
                                        }

                                        proteins.push(`<span style="padding-left: 5px;"><i class="fa-solid fa-turn-up fa-rotate-90"></i> ${protein_id}</span><br/>` + phos.join('<br/>'));
                                    }
                                    html += proteins.join('<br/>');
                                }



                                html += '</td>';
                            }

                            html += `<td>${record.symbol}</td></tr>`;

                            return html;
                        }
                    },
                }).data('dynatable');


            $('#searchResultsTable a').on('click', function (evt) {
                let that = $(this);
                evt.preventDefault();
                selectGeneProtein(that.attr('geneid'), that.attr('proteinid'), that.attr('phosid'), that.attr('uniprotid'), null, null, g.dataSetID, $('#interactiveCovarLODS').val());
                return false;
            });

            setTimeout(function () {
                $('[id="matchInfo"]').each(function (idx, elem) {
                    $(this).popover({
                        placement: 'right',
                        html: true,
                        trigger: 'hover',
                        container: 'body',
                        sanitize: false,
                        content: function () {

                            let d = searchResultsObj[this.getAttribute('geneid')];


                            let h = `<table class="table table-sm">
                                        <tr><td><strong>Symbol</strong> </td><td>${d.symbol}</td></tr>
                                        <tr><td><strong>Location</strong> </td><td>${d.chromosome}:${d.position_start.toLocaleString()}-${d.position_end.toLocaleString()}</td></tr>
                                        <tr><td><strong>Name</strong> </td><td>${d.name}</td></tr>
                                        <tr><td><strong>Synonyms</strong> </td><td>`;

                            let s = d.synonyms;
                            if (s) {
                                if (s.length > 5) {
                                    let s2 = s.slice(1, 6);
                                    h += s2.join('<br>');
                                    h += '<br><i>(' + (s.length - s2.length) + ' more synonyms)</i>';
                                } else {
                                    h += s.join('<br>');
                                }
                            }
                            h += `</td></tr>
                                  <tr><td><strong>Match Reason</strong> </td><td>${d.match_reason}</td></tr>
                                  <tr><td><strong>Match Value</strong> </td><td>${d.match_value}</td></tr>
                                  </table>`;

                            return h;
                        }
                    });
                });

            });

            g.searchResults = searchResults;
            $('#btnGo').button('reset');
            $('#searchTerm').disable(false);

            stopTask();

            // simulate a click event if the search yields only 1 result
            if ($('#searchResultsTable a').length === 1) {
                $('#searchResultsTable a')[0].click();
            }

            $('#searchResultsTable').DataTable({
                "info": false,
                "filter": false,
                "scrollY": "300px",
                "false": true,
                "order": [],
                "paging": false,
            });
        }

        function generateSecondaryPlot(plot, id, markerID, chromosome, location, covar) {
            logDebug('Generating Secondary Plot');
            if (plot === 'navPlotMediation') {
                generateMediationPlot(id, markerID, covar);
            } else if (plot === 'navPlotEffect') {
                generateEffectPlot(id, chromosome, covar);
            } else if (plot === 'navPlotSNPs') {
                generateSNPAssocPlot(id, chromosome, location, covar);
            } else {
                logError('Unknown plot ', plot);
            }
        }

        // The download function takes a CSV string, the filename and mimeType as parameters
        // Scroll/look down at the bottom of this snippet to see how download is called
        function downloadCSV(content, fileName, mimeType) {
            let a = document.createElement('a');
            mimeType = mimeType || 'application/octet-stream';

            if (navigator.msSaveBlob) { // IE10
                navigator.msSaveBlob(new Blob([content], {
                    type: mimeType
                }), fileName);
            } else if (URL && 'download' in a) { //html5 A[download]
                a.href = URL.createObjectURL(new Blob([content], {
                    type: mimeType
                }));
                a.setAttribute('download', fileName);
                document.body.appendChild(a);
                a.click();
                document.body.removeChild(a);
            } else {
                location.href = 'data:application/octet-stream,' + encodeURIComponent(content); // only this mime type is supported
            }
        }

        function exportLODChartData(id, data, covar, exportName) {
            let csvContent = '"id","chromosome","position","lod"\n';
            $.each(data, function(k, v) {
                csvContent += `"${v.id}","${v.chr}",${v.chrPos},${v.y}\n`;
            });

            if (exportName !== null) {
                downloadCSV(csvContent, `${id}_LOD_${covar}_${exportName}.csv`, 'text/csv;encoding:utf-8');
            } else {
                downloadCSV(csvContent, `${id}_LOD.csv`, 'text/csv;encoding:utf-8');
            }
        }

        /**
         * Plot the LOD chart.
         *
         * @param {Array} dataLOD Array of objects containing LOD scores.
         * @param {string} covar The covariate (additive) for default.
         * @param {Array} dataDiff Array of covariate objects containing LOD score.
         */
        function plotLODChart(dataLOD, covar, dataDiff) {
            // "1_3000000", "1", 3, 1.4975
            let maxLOD = -Infinity;
            let minLOD = Infinity;
            let xAxisTitle = '';
            let xAxisSubTitle = '';
            let newLodData = [];
            let currentID = null;
            let isCovar = false;
            let renderTo = 'lodPlotChart';
            let lineColor = 'black';
            let isFull = true;
            let exportName = null;

            if (covar !== 'additive') {
                isCovar = true;
                renderTo = 'lodPlotChartCovariateFull';
                exportName = 'FULL';

                if (dataDiff !== null) {
                    lineColor = '#146582';
                    isFull = false;
                    renderTo = 'lodPlotChartCovariateDiff';
                    exportName = 'DIFF';
                }
            } else {
                //
            }

            logDebug('dataLOD=', dataLOD);
            logDebug('covar=', covar);
            logDebug('dataDiff=', dataDiff);
            logDebug('renderTo=', renderTo);

            if (dataLOD !== null) {
                maxLOD = dataLOD[0][3];
                minLOD = -2;

                if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                    currentID = g.geneID;
                    xAxisTitle = g.geneID + ' (' + g.gene.gene[g.geneID].symbol + ')';
                } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                    currentID = g.proteinID;
                    xAxisTitle = g.proteinID + ' (' + g.gene.gene[g.geneID].symbol + ')';
                } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                    currentID = g.uniprotID;
                    xAxisTitle = g.uniprotID;

                    if (g.proteinID) {
                        xAxisTitle = g.uniprotID + ' ' + g.proteinID
                    }

                    if (g.gene) {
                        xAxisTitle = g.uniprotID + ' ' + g.proteinID + ' (' + g.gene.gene[g.geneID].symbol + ')';
                    }

                } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                    currentID = g.phosID;
                    xAxisTitle = g.phosID;
                } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                    currentID = g.ptmID;
                    xAxisTitle = g.ptmID;
                } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                    currentID = g.peptideID;
                    xAxisTitle = g.peptideID;
                } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                    // pheno
                    xAxisTitle = g.phenotypeID;
                    currentID = g.phenotypeID;
                    if (currentID !== g.DATASETS[g.dataSetID].phenotypes[currentID]['short_name']) {
                        xAxisTitle += (' (' + g.DATASETS[g.dataSetID].phenotypes[currentID]['short_name'] + ')');
                    }
                } else {
                    // TODO: handle error
                    logError('MAJOR PROBLEM');
                }

                if (isCovar) {
                    xAxisSubTitle = `${covar} ${exportName}`;
                }

                $.each(dataLOD, function(index, element) {
                    //
                    // element = {"[id]","[chr]",[position],[data]}
                    //
                    if (element[1] in g.DATASETS[g.dataSetID].chromosomes.chr) {

                        let chr = g.DATASETS[g.dataSetID].chromosomes.chr[element[1]];
                        let x = chr.start + element[2];
                        let y = element[3];

                        if (!isFull) {
                            y = y - dataDiff[index][3];
                        }

                        let d = {
                            x: x,
                            y: y,
                            id: element[0],
                            text: element[3],
                            chr: chr.chromosome,
                            chrPos: element[2],
                            covar: covar
                        };

                        newLodData.push(d);
                        maxLOD = Math.max(element[3], maxLOD);
                    }
                });

                logDebug('MAX LOD = ', maxLOD);

                maxLOD = Math.ceil(maxLOD) + 1;

                logDebug('MAX LOD = ', maxLOD);
            }

            newLodData.sort(compareX);

            logDebug('newLodData=', newLodData);

            //
            // build the vertical chromosome bars
            //
            let chromosomeAxisVals = [];
            let chromosomeAxisText = {};
            let chromosomeBands = [];

            $.each(g.DATASETS[g.dataSetID].chromosomes.idx, function(index, element) {

                chromosomeAxisVals.push(element.mid);
                chromosomeAxisText[element.mid] = element.chromosome;

                // TODO: style this
                let fillColor = '#eeeeee';
                if ((index % 2) === 1) {
                    chromosomeBands.push({
                        color: fillColor,
                        from: element.start,
                        to: element.end
                    });

                    logDebug(fillColor);
                }
            });

            let plotShapes = [];
            let plotAnnotations = [];

            {% if config.PLOT_LOD_LINES %}
            //
            // build the horizontal lines
            //
            let lineVals = [
                {% for l in config.PLOT_LOD_LINES %}
                    {'val':{{l.val}}, 'color':'{{l.color}}', 'text':'{{l.text }}' },
                {% endfor %}
            ];
            let _plotLines = [
                {% for l in config.PLOT_LOD_LINES %}
                    {value:{{l.val}}, color:'{{l.color}}', zIndex:5, width:2, label:{
                        text:'{{l.text }}', align:'right'
                    } },
                {% endfor %}
                ];
            for (let l in lineVals) {
                plotShapes.push({
                    type: 'line',
                    xref: 'paper',
                    x0: 0,
                    y0: lineVals[l].val,
                    x1: 1,
                    y1: lineVals[l].val,
                    line: {
                        color: lineVals[l].color,
                        width: 1.5
                    },
                    layer: 'above',
                });

                plotAnnotations.push({
                    x: 1,
                    y: lineVals[l].val,
                    xref: 'paper',
                    yref: 'y',
                    text: lineVals[l].text,
                    xanchor: 'left',
                    yanchor: 'middle',
                    showarrow: false,
                });
            }
            {% endif %}


            let exporting = appendExportButton('Download CSV', function() {
                exportLODChartData(currentID, newLodData, covar, exportName);
            });

            exporting.filename = currentID + '_LOD';

            if (isCovar) {
                exporting.filename = `${currentID}_LOD_${covar}_${exportName}`;
            }

            g.LODAdditiveMax = maxLOD;

            let series = {
                    boostThreshold: 1,
                    color: lineColor,
                    data: newLodData,
                    name: covar,
                    lineWidth: 0.5,
                    marker: {
                       enabled: false
                    },
                    showInLegend: false,
                    turboThreshold: 0,
                    type: 'line',
                };

            let chartLOD = Highcharts.chart({
                boost: {},
                chart: {
                    alignTicks: false,
                    renderTo: renderTo,
                    zoomType: 'x',
                    events: {
                        click: function(event) {
                            if (event.target.tagName === 'tspan') {
                                // do not generate secondary plot for user clicking "Reset zoom"
                                return;
                            }

                            let plot = $('#secondaryPlotTabs').find('.active').attr('id');
                            let markerID = this.hoverPoint.id;
                            let chromosome = this.hoverPoint.chr;
                            let location = this.hoverPoint.chrPos;
                            let covar = this.hoverPoint.covar;

                            g.secondaryPlotMarkerID = markerID;
                            g.secondaryPlotChromosome = chromosome;
                            g.secondaryPlotLocation = location;
                            g.secondaryPlotCovar = covar;

                            logDebug('event', markerID, chromosome, location, this.hoverPoint.covar);

                            generateSecondaryPlot(plot, currentID, markerID, chromosome, location, covar);
                        }
                    }
                },
                exporting: exporting,
                plotOptions: {
                    series: {
                        findNearestPointBy: 'xy',
                        cursor: 'pointer',
                        events: {
                            click: function(event) {
                                logDebug('plotOptions, series', event);

                                let plot = $('#secondaryPlotTabs').find('.active').attr('id');
                                let markerID = event.point.id;
                                let chromosome = event.point.chr;
                                let location = event.point.chrPos;
                                let covar = event.point.covar;

                                g.secondaryPlotMarkerID = markerID;
                                g.secondaryPlotChromosome = chromosome;
                                g.secondaryPlotLocation = location;
                                g.secondaryPlotCovar = covar;

                                logDebug('series', markerID, chromosome, location, event.point.covar);

                                generateSecondaryPlot(plot, currentID, markerID, chromosome, location, covar);

                            }
                        }
                    }
                },

                series: [series],
                title: {
                    text: xAxisTitle
                },
                subtitle: {
                    text: xAxisSubTitle
                },
                tooltip: {
                    outside: true,
                    /*
                    positioner: function(labelWidth, labelHeight, point) {
                        let tooltipX, tooltipY;
                        let chart = chartLOD;

                        if (point.plotX + labelWidth > chart.plotWidth) {
                            tooltipX = point.plotX + chart.plotLeft - labelWidth - 20;
                        } else {
                            tooltipX = point.plotX + chart.plotLeft + 20;
                        }

                        if (point.plotY + labelHeight > chart.plotHeight) {
                            tooltipY = point.plotY + chart.plotTop - labelHeight - 20;
                        } else {
                            tooltipY = point.plotY + chart.plotTop - 20;
                        }

                        return {
                            x: tooltipX,
                            y: tooltipY
                        };
                    },
                    */
                    useHTML: true,
                    formatter: function () {
                        let p = this.point;
                        return `<b>${p.id}</b><br/>LOD: ${p.y}<br/>Position: ${p.chr}:${p.chrPos.toLocaleString()}`;
                    }
                },
                xAxis: {
                    opposite: true,
                    gridLineWidth: 0,
                    labels: {
                        formatter: function () {
                            return chromosomeAxisText[this.value];
                        }
                    },
                    tickPositions: chromosomeAxisVals,
                    plotBands: chromosomeBands,

                },
                yAxis: {
                    min: 0,
                    max: maxLOD,
                    title: {
                        text: 'LOD'
                    },
                    {% if config.PLOT_LOD_LINES %}
                    plotLines: _plotLines
                    {% endif %}
                },
            });

            if (isCovar) {
                if (isFull) {
                    g.chartLODCovariateFull = chartLOD;
                } else {
                    g.chartLODCovariateDiff = chartLOD;
                }
            } else {
                g.chartLOD = chartLOD;
            }
        }



        /**
         * Get the selected factors.
         * @returns {Array} an array of strings
         */
        function getFactorOrder() {
            let selected = [];
            $('#factor-view-select option:selected').each(function() {
                selected.push([$(this).val(), $(this).attr('order')]);
            });

            selected.sort(function(a, b) {
                return a[1] - b[1];
            });

            let ordered = [];
            for (let i = 0; i < selected.length; i++) {
                ordered.push(selected[i][0]);
            }

            logDebug('getFactorOrder() returning: ', ordered);

            return ordered;
        }


        function configureCovarInfo(resetPeaks) {
            logDebug('configureCovarInfo', resetPeaks);

            // current dataset
            let ds = g.DATASETS[g.dataSetID];

            logDebug("ds['covar_info'] = ", ds['covar_info']);

            if (isEmpty(ds['covar_info'])) {
                logDebug('NO COVAR INFO');

                // configure the lod peaks covariate selections
                if (resetPeaks) {
                    $('#divInteractiveCovariatesPeaks').html(`<input type="hidden" id="interactiveCovarPeaks" value="additive"/>`);
                }

                // configure the profile plot covariate selections
                // currently the profile plot tab is hidden when no covariates available
                let profilePlotOptionsHTML = `
                    <input type="hidden" id="factorSeries" value="additive"/>
                    <input type="hidden" id="factorSeriesCorrelation" value="additive"/>
                    <select id="factorSeries" style="hidden"><option value="" selected></option></select>;
                `;
                $('#divProfilePlotOptions').html(profilePlotOptionsHTML);

                // configure the lod plot covariate selections
                $('#divInteractiveCovariatesLOD').html(`<input type="hidden" id="interactiveCovarLODS" value="additive"/>`);

                // configure the correlation plot covariate selections
                $('#correlationCovariateSelectionText').addClass('hidden').removeClass('visible');
                $('#correlationCovariateSelection').html('<input type="hidden" id="interactiveCovarCorrelation" value="additive"/>');
                $('#correlationImputation').html('');

                // hide the correlation color select
                $("#divCorrelationCovars").addClass('hidden').removeClass('visible');

                logDebug('configureCovarInfo RETURNING');
            } else {

                // configure the profile plot covariate selections

                let profilePlotOptionsHTML = `
                    <div class="col">
                    <div class="row">
                        <div class="col font-weight-bold">
                            Select your factors
                        </div>
                    </div>

                    <div class="row">
                        <div class="col">
                            <div id="factor-view-select-wrapper">
                                <select id="factor-view-select" multiple="multiple">
                                </select>
                            </div>
                        </div>
                    </div>

                    <div class="row row-spacer"></div>

                    <div class="row">
                        <div class="col font-weight-bold">
                            Select a series to color
                        </div>
                    </div>

                    <div class="row">
                        <div class="col" id="divFactorSeries">
                        </div>
                    </div>

                    <div class="row row-spacer"></div>
                    </div>
                `;

                $('#divProfilePlotOptions').html(profilePlotOptionsHTML);

                g.orderCount = ds['covar_info'].length;
                logDebug('g.orderCount=', g.orderCount);

                $('#factor-view-select').multiselect({
                    onChange: function (option, checked) {
                        if (checked) {
                            g.orderCount++;
                            $(option).attr('order', g.orderCount);
                        } else {
                            $(option).attr('order', '');
                        }

                        if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                            plotProfile(g.geneID);
                        } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                            plotProfile(g.proteinID);
                        } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                            plotProfile(g.uniprotID);
                        } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                            plotProfile(g.phosID);
                        } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                            plotProfile(g.ptmID);
                        } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                            plotProfile(g.peptideID);
                        } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                            plotProfile(g.phenotypeID);
                        } else {
                            // TODO: major error
                            logError('MAJOR ERROR');
                        }

                    },
                    maxHeight: 400,
                    dropUp: false,
                    buttonWidth: '100%',
                    buttonClass: 'text-left btn btn-sm btn-secondary',
                    buttonText: function (options) {
                        if (options.length === 0) {
                            return 'None selected';
                        } else {
                            let selected = [];

                            options.each(function () {
                                selected.push([$(this).text(), $(this).attr('order')]);
                            });

                            selected.sort(function (a, b) {
                                return a[1] - b[1];
                            });

                            let text = '';
                            for (let i = 0; i < selected.length; i++) {
                                text += (selected[i][0] + ', ');  // without numbers
                            }

                            return text.substr(0, text.length - 2);
                        }
                    }
                });

                // show the correlation color select
                $("#divCorrelationCovars").addClass('visible').removeClass('hidden');

                let selectOptions = [];
                let html = '<select id="factorSeries" data-style="btn-secondary btn-sm" data-width="100%" class="selectpicker">';
                let htmlCorrelation = '<select id="factorSeriesCorrelation" data-style="btn-secondary btn-sm" data-width="100%" class="selectpicker">';
                let htmlInteractiveCovariatesLOD = `<div class="row">
                                                        <div class="col-auto align-self-center font-weight-bold">
                                                            <span>Plots</span>
                                                        </div>
                                                        <div class="col">
                                                            <select id="interactiveCovarLODS" data-style="btn-secondary btn-sm" data-width="100%" class="selectpicker">
                                                    `;
                let htmlInteractiveCovariatesPeaks = `<div class="row">
                                                          <div class="col-auto align-self-center">
                                                            <span>Plot type</span>
                                                          </div>
                                                          <div class="col">
                                                              <select id="interactiveCovarPeaks" data-style="btn-secondary btn-sm" data-width="100%" class="selectpicker">
                                                `;

                let htmlInteractiveCovariatesCorrelation = `<select id="interactiveCovarCorrelation" data-style="btn-secondary btn-sm" data-width="100%" class="selectpicker">`;

                g.interactiveCovariates = [];

                htmlInteractiveCovariatesLOD += `<option value="additive">Additive</option>`;
                htmlInteractiveCovariatesPeaks += `<option value="additive">Additive</option>`;
                htmlInteractiveCovariatesCorrelation += `<option value="none">None</option>`;

                {% if 'FACTOR_VIEW_SELECT_INITIAL_ALL' in config %}
                    {% set FACTOR_VIEW_SELECT_INITIAL_ALL = config.FACTOR_VIEW_SELECT_INITIAL_ALL %}
                {% else %}
                    {% set FACTOR_VIEW_SELECT_INITIAL_ALL = False %}
                {% endif %}

                $.each(ds['covar_info'], function (index, element) {
                    selectOptions.push({
                        label: element['display_name'],
                        order: index,
                    {% if FACTOR_VIEW_SELECT_INITIAL_ALL %}
                        selected: true,
                    {% else %}
                        selected: element.primary,
                    {% endif %}
                        title: element['display_name'],
                        value: element['sample_column'],
                    });

                    html += `<option value="${element['sample_column']}">${element['display_name']}</option>`;
                    htmlCorrelation += `<option value="${element['sample_column']}">${element['display_name']}</option>`;

                    // interactive covariates
                    if (element.interactive) {
                        g.interactiveCovariates.push(element);
                        htmlInteractiveCovariatesLOD += `<option value="${element['sample_column']}">${element['display_name']}</option>`;
                        htmlInteractiveCovariatesPeaks += `<option value="${element['sample_column']}">${element['display_name']}</option>`;
                        htmlInteractiveCovariatesCorrelation += `<option value="${element['sample_column']}">${element['display_name']}</option>`;
                    }
                });

                html += '</select>';
                htmlCorrelation += '</select>';
                htmlInteractiveCovariatesLOD += '</select></div></div>';
                htmlInteractiveCovariatesPeaks += '</select></div></div>';
                htmlInteractiveCovariatesCorrelation += '</select>';

                $('#factor-view-select').multiselect('dataprovider', selectOptions);

                $('#divFactorSeries').html(html);
                $('#divFactorSeriesCorrelation').html(htmlCorrelation);


                logDebug('g.interactiveCovariates=', g.interactiveCovariates);

                // display a selection if we have interactive covariates, nothing if not
                if (g.interactiveCovariates.length > 0) {
                    $('#divInteractiveCovariatesLOD').html(htmlInteractiveCovariatesLOD);

                    $('#interactiveCovarLODS').selectpicker();

                    $('#interactiveCovarLODS').on('changed.bs.select', function (e, clickedIndex, isSelected, previousValue) {
                        changeCovariate(e.target.value);
                    });

                    // add Covariate adjustment for correlation
                    $('#correlationCovariateSelectionText').addClass('visible').removeClass('hidden');
                    $('#correlationCovariateSelection').html(htmlInteractiveCovariatesCorrelation);
                    $('#interactiveCovarCorrelation').selectpicker();

                    $('#interactiveCovarCorrelation').on('changed.bs.select', function (e, clickedIndex, isSelected, previousValue) {
                        let selected = $(this).find(':selected').val();
                        let dataset = $('#correlationDatasetSelect').find(':selected').val();

                        if ((dataset == undefined) ||
                            (dataset == null) ||
                            (dataset === 'none')) {
                            dataset = g.dataSetID;
                        }

                        switchCorrelationDataSet(dataset, selected);
                    });


                    if (resetPeaks) {
                        $('#divInteractiveCovariatesPeaks').html(htmlInteractiveCovariatesPeaks);

                        $('#interactiveCovarPeaks').selectpicker();

                        $('#interactiveCovarPeaks').on('changed.bs.select', function (e, clickedIndex, isSelected, previousValue) {
                            generateLODPeaksHTML();
                        });
                    }


                } else {
                    $('#divInteractiveCovariatesLOD').html(`<input type="hidden" id="interactiveCovarLODS" value="additive"/>`);

                    // remove Covariate adjustment for correlation
                    $('#correlationCovariateSelectionText').addClass('hidden').removeClass('visible');
                    $('#correlationCovariateSelection').html('<input type="hidden" id="interactiveCovarCorrelation" value="additive"/>');

                    if (resetPeaks) {
                        $('#divInteractiveCovariatesPeaks').html(`<input type="hidden" id="interactiveCovarPeaks" value="additive"/>`);
                    }
                }

                $('#factorSeries').selectpicker();

                $('#factorSeries').on('changed.bs.select', function (e, clickedIndex, isSelected, previousValue) {
                    if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                        plotProfile(g.geneID);
                    } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                        plotProfile(g.proteinID);
                    } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                        plotProfile(g.uniprotID);
                    } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                        plotProfile(g.phosID);
                    } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                        plotProfile(g.ptmID);
                    } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                        plotProfile(g.peptideID);
                    } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                        plotProfile(g.phenotypeID);
                    } else {
                        // TODO: major error
                        logError('MAJOR ERROR');
                    }
                });

                $('#factorSeriesCorrelation').selectpicker();

                $('#factorSeriesCorrelation').on('changed.bs.select', function (e, clickedIndex, isSelected, previousValue) {
                    plotCorrelationChart();
                });

            }

            $('#btnDownloadCorrelationData').off('click');
            $('#btnDownloadCorrelationData').on('click', function(evt) {
                logDebug('Downloading correlation data');
                if (g.correlationData) {
                    let csvContent = '';
                    let id = '';

                    if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                        id = g.geneID;
                    } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                        id = g.proteinID;
                    } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                        id = g.uniprotID;
                    } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                        id = g.phosID;
                    } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                        id = g.phenotypeID;
                    } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                        id = g.ptmID;
                    } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                        id = g.peptideID;
                    } else {
                        logError('MAJOR ERROR');
                    }

                    if (g.DATASETS[g.correlationDatasetID].datatype === 'mrna') {
                        csvContent = '"gene_id","symbol","chr","start","end","correlation"\n';
                        $.each(g.correlationData.correlations, function (k, v) {
                            csvContent += `"${v.id}","${v.symbol}","${v.chr}",${v.start},${v.end},${v.cor}\n`;
                        });
                    } else if (g.DATASETS[g.correlationDatasetID].datatype === 'protein') {
                        csvContent = '"gene_id","protein_id","symbol","chr","start","end","correlation"\n';
                        $.each(g.correlationData.correlations, function (k, v) {
                            csvContent += `"${v.gene_id}","${v.id}","${v.symbol}","${v.chr}",${v.start},${v.end},${v.cor}\n`;
                        });
                    } else if (g.DATASETS[g.correlationDatasetID].datatype === 'protein_uniprot') {
                        csvContent = '"gene_id","protein_id","uniprot_id","symbol","chr","start","end","correlation"\n';
                        $.each(g.correlationData.correlations, function (k, v) {
                            csvContent += `"${v.gene_id}","${v.protein_id}","${v.id}","${v.symbol}","${v.chr}",${v.start},${v.end},${v.cor}\n`;
                        });
                    } else if (g.DATASETS[g.correlationDatasetID].datatype === 'phos') {
                        csvContent = '"gene_id","protein_id","phos_id","symbol","chr","start","end","correlation"\n';
                        $.each(g.correlationData.correlations, function (k, v) {
                            csvContent += `"${v.gene_id}","${v.protein_id}","${v.id}","${v.symbol}","${v.chr}",${v.start},${v.end},${v.cor}\n`;
                        });
                    } else if (g.DATASETS[g.correlationDatasetID].datatype === 'ptm') {
                        csvContent = '"gene_id","protein_id","ptm_id","symbol","chr","start","end","correlation"\n';
                        $.each(g.correlationData.correlations, function (k, v) {
                            csvContent += `"${v.gene_id}","${v.protein_id}","${v.id}","${v.symbol}","${v.chr}",${v.start},${v.end},${v.cor}\n`;
                        });
                    } else if (g.DATASETS[g.correlationDatasetID].datatype === 'peptide') {
                        csvContent = '"gene_id","protein_id","peptide_id","symbol","chr","start","end","correlation"\n';
                        $.each(g.correlationData.correlations, function (k, v) {
                            csvContent += `"${v.gene_id}","${v.protein_id}","${v.id}","${v.symbol}","${v.chr}",${v.start},${v.end},${v.cor}\n`;
                        });
                    } else if (g.DATASETS[g.correlationDatasetID].datatype === 'pheno') {
                        csvContent = '"id","correlation"\n';
                        $.each(g.correlationData.correlations, function (k, v) {
                            csvContent += `"${v.id}",${v.cor}\n`;
                        });
                    } else {
                        // TODO: major error
                        logError('MAJOR ERROR');
                    }

                    downloadCSV(csvContent, `${id}_correlation.csv`, 'text/csv;encoding:utf-8');
                } else {
                    logDebug('no correlation data?');
                }
            });
        }

        function formatStrToNum(val, decimals) {
            return Number.parseFloat(val).toFixed(decimals);
        }


        function displayCorrelation(correlationData) {
            $('#correlationDataTable').html('');

            {% if debug %}
            console.time('correlationTable');
            {% endif %}

            logDebug('displayCorrelation()');
            logDebug('correlationData=', correlationData);
            g.correlationData = correlationData;

            logDebug('g.DATASETS[g.correlationDatasetID].datatype=', g.DATASETS[g.correlationDatasetID].datatype);

            let htmlBody = '<table id="corTable" class="table table-striped table-bordered table-sm table-sm-text">';

            if (g.DATASETS[g.correlationDatasetID].datatype === 'mrna') {
                htmlBody += '<thead><th>Value</th><th>Symbol</th><th>ID</th></thead>';
                $.each(correlationData.correlations, function(idx, value) {
                    if (idx < 1000) {
                        htmlBody += '<tr><td data-order="' + Math.abs(formatStrToNum(value.cor, 4)) + '">' + formatStrToNum(value.cor, 4) + `</td><td>${value.symbol}</td><td><a href="#" dSymbol="${value.symbol}" dID="${value.id}">${value.id}</a></td></tr>`;
                    }
                });
            } else if (g.DATASETS[g.correlationDatasetID].datatype === 'protein') {
                htmlBody += '<thead><th>Value</th><th>Symbol</th><th>ID</th></thead>';
                $.each(correlationData.correlations, function(idx, value) {
                    if (idx < 1000) {
                        htmlBody += '<tr><td data-order="' + Math.abs(formatStrToNum(value.cor, 4)) + '">' + formatStrToNum(value.cor, 4) + `</td><td>${value.symbol}</td><td><a href="#" dSymbol="${value.symbol}" dID="${value.id}">${value.id}</a></td></tr>`;
                    }
                });
            } else if (g.DATASETS[g.correlationDatasetID].datatype === 'protein_uniprot') {
                htmlBody += '<thead><th>Value</th><th>Symbol</th><th>ID</th></thead>';
                $.each(correlationData.correlations, function(idx, value) {
                    if (idx < 1000) {
                        htmlBody += '<tr><td data-order="' + Math.abs(formatStrToNum(value.cor, 4)) + '">' + formatStrToNum(value.cor, 4) + `</td><td>${value.symbol}</td><td><a href="#" dSymbol="${value.symbol}" dID="${value.id}">${value.id}</a></td></tr>`;
                    }
                });
            } else if (g.DATASETS[g.correlationDatasetID].datatype === 'phos') {
                htmlBody += '<thead><th>Value</th><th>Gene Symbol</th><th>ID</th></thead>';
                $.each(correlationData.correlations, function(idx, value) {
                    if (idx < 1000) {
                        let ph_id = value.id.split('_')[1];
                        htmlBody += '<tr><td data-order="' + Math.abs(formatStrToNum(value.cor, 4)) + '">' + formatStrToNum(value.cor, 4) + `</td><td>${value.symbol}</td><td><a href="#" dSymbol="${value.symbol}" dID="${value.id}">${ph_id}</a></td></tr>`;
                    }
                });
            } else if (g.DATASETS[g.correlationDatasetID].datatype === 'ptm') {
                htmlBody += '<thead><th>Value</th><th>ID</th><th>Uniprot ID</th><th>Gene Symbol</th></thead>';
                $.each(correlationData.correlations, function(idx, value) {
                    if (idx < 1000) {
                        let ptm_id = value.id;
                        htmlBody += '<tr><td data-order="' + Math.abs(formatStrToNum(value.cor, 4)) + '">' + formatStrToNum(value.cor, 4) + `</td><td><a href="#" dSymbol="${value.symbol}" dID="${value.id}">${ptm_id}</a></td><td>${value.uniprot_id}</td><td>${value.symbol}</td></tr>`;
                    }
                });
            } else if (g.DATASETS[g.correlationDatasetID].datatype === 'peptide') {
                htmlBody += '<thead><th>Value</th><th>ID</th><th>Uniprot ID</th><th>Gene Symbol</th></thead>';
                $.each(correlationData.correlations, function(idx, value) {
                    if (idx < 1000) {
                        let peptide_id = value.id;
                        htmlBody += '<tr><td data-order="' + Math.abs(formatStrToNum(value.cor, 4)) + '">' + formatStrToNum(value.cor, 4) + `</td><td><a href="#" dSymbol="${value.symbol}" dID="${value.id}">${peptide_id}</a></td><td>${value.uniprot_id}</td><td>${value.symbol}</td></tr>`;
                    }
                });
            } else if (g.DATASETS[g.correlationDatasetID].datatype === 'pheno') {
                htmlBody += '<thead><th>Value</th><th>ID</th></thead>';
                $.each(correlationData.correlations, function(idx, value) {
                    if (idx < 1000) {
                        htmlBody += '<tr><td data-order="' + Math.abs(formatStrToNum(value.cor, 4)) + '">' + formatStrToNum(value.cor, 4) + `</td><td><a href="#" dSymbol="" dID="${value.id}">${value.id}</a></td></tr>`;
                    }
                });
            } else {
                // TODO: major error
                logError('MAJOR ERROR');
            }

            htmlBody += '</table>';
            $('#correlationDataTable').html(htmlBody);

            $('#corTable').DataTable({
                "info":     false,
                "language": {
                    "search": "Filter"
                },
                "scrollY":        "200px",
                "scrollCollapse": true,
                "order": [[ 0, "desc" ]],
                "paging":         false,
                "columnDefs": [{
                    "targets": 0,
                    "className": "dt-body-right"
                }]
            });

            $('#correlationImputation').html('');

            /*
            if ((correlationData.imputed_samples != null) && (correlationData.imputed_samples.length > 0)){
                $('#correlationImputation').html('<span style="color:red">* Some missing values were imputed to calculate the correlations.</span>');
            }
            */

            $('#corTable a').click(function(evt) {
                evt.preventDefault();
                let that = $(this);
                let cid = null;

                if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                    cid = g.geneID;
                } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                    cid = g.proteinID;
                } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                    cid = g.uniprotID;
                } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                    cid = g.phosID;
                } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                    cid = g.ptmID;
                } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                    cid = g.peptideID;
                } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                    cid = g.phenotypeID;
                }

                g.currentCorrelateID =  that.attr('dID');

                let interactiveCovariate = $('#interactiveCovarCorrelation').find(':selected').val();

                if (interactiveCovariate === 'none') {
                    interactiveCovariate = null;
                }

                generateCorrelationPlot(g.dataSetID, cid, $('#correlationDatasetSelect').val(), that.attr('dID'), that.attr('dSymbol'), interactiveCovariate);
                return false;
            });

            {% if debug %}
            console.timeEnd('correlationTable');
            {% endif %}

        }

        function exportCorrelationPlotData(id, idCorrelate, data) {
            let csvContent = `"sample_id","${id}","${idCorrelate}"\n`;

            $.each(data, function(k, v) {
                csvContent += `"${v['sample_id']}",${v.x},${v.y}\n`;
            });

            downloadCSV(csvContent, `${id}_${idCorrelate}_CORRELATION.csv`, 'text/csv;encoding:utf-8');
        }

        function plotCorrelationChart() {
            let data = g.correlationChartData;
            logDebug('plotCorrelationChart, data=', data);

            let categorySeries = $('#divFactorSeriesCorrelation').find(':selected').val();
            logDebug('categorySeries=', categorySeries);

            let plotTitle = '';
            let xAxis = '';
            let yAxis = g.currentCorrelateID;
            let currentID = '';

            if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                plotTitle = g.geneID;
                currentID = g.geneID;
                xAxis = plotTitle;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                plotTitle = g.proteinID;
                currentID = g.proteinID;
                xAxis = plotTitle;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                plotTitle = g.uniprotID;
                currentID = g.uniprotID;
                xAxis = plotTitle;
            } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                plotTitle = g.phosID;
                currentID = g.phosID;
                xAxis = plotTitle;
            } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                plotTitle = g.ptm;
                currentID = g.ptmID;
                xAxis = plotTitle;
                if (g.ptmID !== g.DATASETS[g.dataSetID].ptms[g.ptmID]['ptm']) {
                    plotTitle = g.DATASETS[g.dataSetID].ptms[g.ptmID]['ptm'];
                }
            } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                plotTitle = g.peptideID;
                currentID = g.peptideID;
                xAxis = plotTitle;
            } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                plotTitle = g.phenotypeID;
                currentID = g.phenotypeID;
                xAxis = plotTitle;
                if (g.phenotypeID !== g.DATASETS[g.dataSetID].phenotypes[g.phenotypeID]['short_name']) {
                    plotTitle += (' (' + g.DATASETS[g.dataSetID].phenotypes[g.phenotypeID]['short_name'] + ')');
                }
            }

            let seriesData = new Set();

            if (Array.isArray(data.datatypes[categorySeries])) {
                seriesData = new Set(data.datatypes[categorySeries]);
            } else {
                seriesData.add(data.datatypes[categorySeries]);
            }

            let minValue = Infinity;
            let maxValue = -Infinity;

            let cfMap = {};
            $.each(g.DATASETS[g.dataSetID]['covar_info'], function(idx, elem) {
                cfMap[elem['sample_column']] = elem['display_name'];
            });

            let newData = [];

            $.each(data.data, function(idx, element) {
                if ((element.x !== undefined) && (element.y !== undefined)) {
                    minValue = Math.min(minValue, element.x, element.y);
                    maxValue = Math.max(maxValue, element.x, element.y);

                    // make a copy of element
                    let d = JSON.parse(JSON.stringify(element));
                    d['map'] = cfMap;
                    d['seriesValue'] = element[categorySeries];
                    newData.push(d);
                }
            });

            minValue = Math.floor(minValue);
            maxValue = Math.ceil(maxValue);


            let axisTicks = calculateTicks(minValue, maxValue, 5);

            let axisLines = [];

            $.each(axisTicks, function(k,v) {
                axisLines.push({
                    axisLOD: v
                })
            });

            let colors = ['#8B8682', '#87CEEB', '#27408B', '#FFA500', '#B22222', '#f7a35c', '#90ed7d', '#e4d354', '#2b908f', '#f45b5b'];

            let correlationColorDomain = Array.from(seriesData);
            //let correlationColorRange = g.PLOT_COLORS.slice(0, correlationColorDomain.length);
            let correlationColorRange = colors.slice(0, correlationColorDomain.length);
            let revColors = [];

            for (let x = correlationColorRange.length - 1; x >= 0; x--) {
                revColors.push(correlationColorRange[x]);
            }

            logDebug('newData=', newData);
            logDebug('minValue=', minValue);
            logDebug('maxValue=', maxValue);
            logDebug('correlationColorDomain=', correlationColorDomain);
            logDebug('correlationColorRange=', correlationColorRange);
            logDebug('revColors=', revColors);
            logDebug('seriesData=', seriesData);
            logDebug('cfMap=', cfMap);
            logDebug('axisTicks=', axisTicks);


            let yAxisTitle = data['id.correlate'];

            /*
            if (correlateSymbol !== '') {
                yAxisTitle = correlateSymbol;
            }
            */

            /*
            let exporting = appendExportButton('Download CSV', function() {
                exportCorrelationPlotData(data.id, data['id.correlate'], data.data);
            });
            exporting.filename = data.id + '_' + data['id.correlate'] + '_CORRELATION';
            */

            let trackCorrelation = {
                "name": "correlation",
                "layer": [
                        {
                            "title": "Xbound",
                            "data": {
                                "values": axisLines,
                            },
                            "mark": {
                                "type": "rule",
                                "tooltip": {
                                    "handler": "nullHandler"
                                },
                            },
                            "encoding": {
                                "y": {
                                    "field": "axisLOD",
                                    "type": "quantitative",
                                    "title": null
                                },
                                "color": {
                                    "value": "#e5e5e5"
                                }
                            },
                        },
                        {
                            "title": "Ybound",
                            "data": {
                                "values": axisLines,
                            },
                            "mark": {
                                "type": "rule",
                                "tooltip": {
                                    "handler": "nullHandler"
                                },
                            },
                            "encoding": {
                                "x": {
                                    "field": "axisLOD",
                                    "type": "quantitative",
                                    "title": null
                                },
                                "color": {
                                    "value": "#e5e5e5"
                                }
                            },
                        },
                        {
                        "data": {
                            "values": newData
                        },

                        "transform": [
                            { "type": "formula", "expr": "datum.imputed ? 'diamond' : 'point'", "as": "myShape" }
                        ],

                        "title": "Correlation",
                        "mark": {
                            "type": "point",
                            "tooltip": {
                                "handler": "correlationTooltipHandler"
                            },
                            "geometricZoomBound": 4,
                            "inwardStroke": true,
                            "stroke": "#333333",
                            "strokeWidth": 0.7,
                            "fillOpacity": 0.6,
                            "strokeOpacity": 0.8
                        },
                        "encoding": {
                            "shape": {
                                "field": "myShape",
                                "type":"nominal"
                            },
                            "x": {
                                "title": xAxis,
                                "field": "x",
                                "type": "quantitative",
                                "scale": {
                                    "zoom": true,
                                    "domain": [minValue, maxValue]
                                },
                                "axis": {
                                    "values": axisTicks
                                },
                            },
                            "y": {
                                "title": yAxis,
                                "field": "y",
                                "type": "quantitative",
                                "scale": {
                                    "zoom": true,
                                    "domain": [minValue, maxValue]
                                },
                                "axis": {
                                    "values": axisTicks
                                },
                            },
                            "color": {
                                "type": "nominal",
                                "field": "seriesValue",
                                "scale": {
                                    "domain": correlationColorDomain,
                                    "range": revColors,
                                }
                            },
                            "size": {
                                "value": 150
                            },
                            "opacity": {
                                "value": 1
                            }
                        }
                    }
                ]
            };

            // TODO: figure out legend
            //let legend = calcLegend(correlationColorDomain, correlationColorRange, "correlationLegend");
            let tracks = [trackCorrelation]; //, legend];
            let handlers = {
                nullHandler: nullHandler,
                correlationTooltipHandler: correlationTooltipHandler
            };

            const spec = {
                "vconcat": tracks
            };

            logDebug('SPEC=', spec);
            $("#plotCorrelation").html('<div class="row"><div class="col" id="plotCorrelationMenu"></div></div><div class="row"><div class="col" id="plotCorrelationChart"></div></div>');
            $('#plotCorrelationChart').width('400');
            $('#plotCorrelationChart').height('400');

            $('#plotCorrelationMenu').append(`<div class="col text-center font-weight-bold">${plotTitle}
                <div class="dropdown dropleft float-right" id="correlationMenuDropdown">
                  <a class="text-secondary" href="#" role="button" id="correlationDropdownMenuLink" data-toggle="dropdown" aria-expanded="false">
                    <i class="fa-solid fa-bars"></i>
                  </a>

                  <div class="dropdown-menu" aria-labelledby="correlationDropdownMenuLink">
                    <a class="dropdown-item" id="correlationDownloadPNG" style="font-size:11px !important">Download PNG image</a>
                    <a class="dropdown-item" id="correlationDownloadCSV" style="font-size:11px !important">Download CSV</a>
                  </div>
                </div>
            </div>`);

            let n = currentID + '_CORRELATION';
            $("#correlationDownloadPNG").click(function (event) {
                downloadGenomeSpy('#plotCorrelationChart', `${n}.png`, '255,255,255');
            });

            $("#correlationDownloadCSV").click(function (event) {
                exportCorrelationPlotData(currentID, g.currentCorrelateID, data.data);
            });

            g.chartCorrelation = genomeSpyEmbed.embed(
                document.querySelector("#plotCorrelationChart"),
                spec,
                {
                    tooltipHandlers: handlers
                }
            ).then(function(api) {
                api.addEventListener("click", function(event) {
                    if (event.datum != null) {
                        let datum = event.datum;
                        logDebug('correlation clicked on', datum);
                    }
                });
            });
        }

        function updateCorrelationData(groupID, currentDataset, currentID, correlateDataset, correlateID, correlateSymbol) {
            // send GET request to status URL
            logDebug('updateCorrelationData');
            logDebug(currentDataset, currentID, correlateDataset, correlateID, correlateSymbol);

            if (g.runningTask) {
                let statusURL = '{{ url_for('api.api_status', group_id='', _external=False) }}' + groupID;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    retries: 3,
                    retryInterval: 1000,
                    success: function (data, status, request) {
                        logDebug('DATA=======', data);
                        if (data.status === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data.number_tasks_errors !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.error}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data.number_tasks_errors === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if (value.status_code !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.response.error}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the Correlation Plot.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    g.correlationChartData = data.response_data.correlationPlot.response.result;
                                    plotCorrelationChart();
                                    stopTask();
                                }
                            }
                        } else {
                            // rerun in 1 seconds
                            logDebug('Not done, keep checking...');
                            setTimeout(function () {
                                updateCorrelationData(groupID, currentDataset, currentID, correlateDataset, correlateID, correlateSymbol);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                logDebug('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=False) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    logDebug(data);
                    g.chartCorrelation = null;
                    $('#correlationDataTable').html('');
                    $('#plotCorrelation').html('');
                    $('#correlationImputation');
                });
            }
        }

        function generateCorrelationPlot(currentDataset, currentID, correlateDataset, correlateID, correlateSymbol, interactiveCovariate) {
            logDebug('Generating correlation plot:', currentDataset, currentID, correlateDataset, correlateID, correlateSymbol);
            let urlCorrelationPlot = `{{ config.API_R_BASE }}/correlationplot?dataset=${currentDataset}&id=${currentID}&dataset_correlate=${correlateDataset}&id_correlate=${correlateID}`;

            if ((interactiveCovariate !== undefined) &&
                (interactiveCovariate !== null) &&
                (interactiveCovariate !== 'none')) {
                urlCorrelationPlot += `&intcovar=${interactiveCovariate}`;
            }

            let submitData = {
                urls:[{
                    url_id: 'correlationPlot',
                    url: urlCorrelationPlot
                }]
            };

            startTask();

            // reset the chart and clear it
            g.correlationPlot = null;
            $('#plotCorrelation').html('');

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=False) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                retries: 3,
                retryInterval: 1000,
                success: function(data, status, request) {
                    updateCorrelationData(data.group_id, currentDataset, currentID, correlateDataset, correlateID, correlateSymbol);
                },
                error: function(jqXHR, textStatus, errorThrown) {
                    showErrorMessage(errorThrown, textStatus);
                }
            });
        }

        function resetSecondaryPlots() {
            g.chartMediation = null;
            g.chartEffect = {};
            g.chartSNPAssoc = null;
            $('#plotMediation').html('');
            $('#allEffectPlots').html('');
            $('#plotSNPAssociation').html('');
            $('#snpWindowMenuShift').html('');
            $('#snpWindowMenu').html('');
        }

        function updateGeneSelect(groupID, submitData, geneID, proteinID, phosID, uniprotID, ptmID, peptideID, dataSetID, covar) {
            // send GET request to status URL
            logDebug('update progress, calling .ajax');
            logDebug("updateGeneSelect submitData=", submitData);
            logDebug(`updateGeneSelect(${groupID}, ${submitData}, ${geneID}, ${proteinID}, ${phosID}, ${uniprotID}, ${ptmID}, ${peptideID}, ${dataSetID}, ${covar})`);


            if (g.runningTask) {
                let statusURL = `{{ url_for('api.api_status', group_id='', _external=False) }}${groupID}`;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    retries: 3,
                    retryInterval: 1000,
                    success: function (data, status, request) {
                        logDebug('DATA=======', data);

                        if (data.status === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data.number_tasks_errors !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.error}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data.number_tasks_errors === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if (value.status_code !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.response.error}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the LOD SCAN.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result, there will be 3 datasets to get

                                    // 1. gene information
                                    g.gene = data.response_data.geneData.response;
                                    displayGeneData(g.gene.gene);
                                    let geneData = null;
                                    for (let key in g.gene.gene) {
                                        geneData = g.gene.gene[key];
                                    }

                                    // 2. LOD score information
                                    if (covar === 'additive') {
                                        g.LODChartData[covar] = data.response_data.lod.response.result;
                                        plotLODChart(data.response_data.lod.response.result, covar, null);
                                    } else {
                                        g.LODChartData.additive = data.response_data.lod.response.result;
                                        g.LODChartData[covar] = data.response_data.lodCovar.response.result;

                                        plotLODChart(data.response_data.lodCovar.response.result,
                                                     covar,
                                                     null);

                                        plotLODChart(data.response_data.lodCovar.response.result,
                                                     covar,
                                                     data.response_data.lod.response.result);
                                    }

                                    $('#interactiveCovarLODS').selectpicker('val', covar);

                                    // 3. Expression information
                                    g.expressionData = data.response_data.expression.response.result;

                                    if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                                        plotProfile(g.geneID);
                                    } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                                        plotProfile(g.proteinID);
                                    } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                                        plotProfile(g.uniprotID);
                                    } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                                        plotProfile(g.phosID);
                                    } else {
                                        logDebug('ERROR');
                                    }

                                    // 4. Correlation
                                    displayCorrelation(data.response_data.correlation.response.result);

                                    // first time we need to hide jumbotron instructions
                                    // it's ok to keep hiding
                                    $('#divWelcomeMesage').html('');
                                    $('#divLOD').removeClass('invisible');
                                    $('#divItemInfo').removeClass('invisible');
                                    $('#divSecondRow').removeClass('invisible');
                                    $('#divProfileCorrelation').removeClass('invisible');
                                    stopTask();
                                }
                            }
                        }
                        else {
                            // rerun in 1 seconds
                            logDebug('Not done, keep checking...');
                            setTimeout(function () {
                                updateGeneSelect(groupID, submitData, geneID, proteinID, phosID, uniprotID, ptmID, peptideID, dataSetID, covar);
                            }, 1000);
                        }
                    }
                });
            } else {
                // TODO: cleanup
                logDebug('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=False) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    logDebug(data);
                });
            }
        }

        /**
         * Perform gene search.
         */
        function selectPhenotype(phenotypeID, dataSetID, covar) {
            let urlExpression = `{{ config.API_R_BASE }}/expression?dataset=${dataSetID}&id=${phenotypeID}`;
            let urlLOD = `{{ config.API_R_BASE }}/lodscores?dataset=${dataSetID}&id=${phenotypeID}`;
            let urlLODCovar = `{{ config.API_R_BASE }}/lodscores?dataset=${dataSetID}&id=${phenotypeID}`;
            let urlCorrelation = `{{ config.API_R_BASE }}/correlation?dataset=${dataSetID}&id=${phenotypeID}`;

            // TODO: make reset function
            g.gene = null;
            g.geneID = null;
            g.proteinID = null;
            g.phosID = null;
            g.ptmID = null;
            g.peptideID = null;
            g.phenotypeID = phenotypeID;
            g.secondaryPlotMarkerID = null;
            g.secondaryPlotChromosome = null;
            g.secondaryPlotLocation = null;
            g.secondaryPlotCovar = null;
            g.correlationData = null;
            g.correlationDatasetID = dataSetID;
            g.LODChartData = {};
            g.LODChartData.additive = null;

            // reset the covariate pickers
            configureCovarInfo(false);

            $.each(g.interactiveCovariates, function(idx, value) {
                g.LODChartData[value['column.name']] = null;
            });

            if (covar === 'additive') {
                urlLOD += '&intcovar=additive';
                urlLODCovar = null;
            } else {
                urlLOD += '&intcovar=additive';
                urlLODCovar += ('&intcovar=' + covar);
            }


            // TODO: add ncores?, regress_local? to urlLOD, or handle backend?
            //URL += '&ncores=' + _G.api_cores;
            {% if config.API_R_NUM_CORES %}
            if (urlLODCovar !== null) {
                urlLODCovar += '&cores={{ config.API_R_NUM_CORES }}';
            }
            urlLOD += '&cores={{ config.API_R_NUM_CORES }}';
            {% endif %}

            let submitData = {
                urls:[{
                    url_id: 'expression',
                    url: urlExpression
                }, {
                    url_id: 'correlation',
                    url: urlCorrelation
                }, {
                    url_id: 'lod',
                    url: urlLOD
                }]
            };

            if (urlLODCovar !== null) {
                submitData.urls.push({
                    url_id: 'lodCovar',
                    url: urlLODCovar
                });
            }

            logDebug('submitData=', submitData);

            startTask();

            $('#div_current_item_information_header').html('');
            $('#div_current_item_information_body').html('');

            g.chartLOD = null;
            g.chartLODCovariateFull = null;
            g.chartLODCovariateDiff = null;
            $('#lodPlotChart').html('');
            $('#lodPlotChartCovariateFull').html('');
            $('#lodPlotChartCovariateDiff').html('');

            g.chartCorrelation = null;
            $('#plotCorrelation').html('');
            $('#correlationDataTable').html('');
            $('#correlationImputation').html('');

            // The following line fires a change event which we do not want
            // $('#correlationDatasetSelect').selectpicker('val', g.dataSetID);

            // Do this instead
            $('#correlationDatasetSelect').val(g.dataSetID);
            $('#correlationDatasetSelect').selectpicker('render');
            $('#fact-svg-chart').html('');

            resetSecondaryPlots();

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=False) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                retries: 3,
                retryInterval: 1000,
                success: function(data, status, request) {
                    g.runningTask = true;
                    logDebug('data=', data);
                    logDebug('status=', status);
                    logDebug('request=', request);
                    //let status_url = request.getResponseHeader('Location');
                    //logDebug('status_url=', status_url);
                    logDebug('data.group_id=', data.group_id);
                    updatePhenotypeSelect(data.group_id, dataSetID, covar);
                },
                error: function(jqXHR, textStatus, errorThrown) {
                    showErrorMessage(errorThrown, textStatus);
                }
            });
        }

        function updatePTMPeptideSelect(groupID, datasetID, covar) {
            // send GET request to status URL
            logDebug('update progress, calling .ajax');
            logDebug(groupID);

            if (g.runningTask) {
                let statusURL = `{{ url_for('api.api_status', group_id='', _external=False) }}${groupID}`;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    retries: 3,
                    retryInterval: 1000,
                    success: function (data, status, request) {
                        logDebug('DATA=======', data);

                        if (data.status === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data.number_tasks_errors !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.error}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data.number_tasks_errors === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if (value.status_code !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.response.error}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the LOD SCAN.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    if ('geneData' in data.response_data) {
                                        g.gene = data.response_data.geneData.response;
                                        displayGeneData(g.gene.gene);
                                        let geneData = null;
                                        for (let key in g.gene.gene) {
                                            geneData = g.gene.gene[key];
                                        }
                                    }

                                    displayPTMPeptideData(g.phenotypeID);

                                    // 1. LOD score information
                                    //plotLODChart(data.response_data.lod.response.result);

                                    if (covar === 'additive') {
                                        g.LODChartData[covar] = data.response_data.lod.response.result;
                                        plotLODChart(data.response_data.lod.response.result, covar, null);
                                    } else {
                                        g.LODChartData.additive = data.response_data.lod.response.result;
                                        g.LODChartData[covar] = data.response_data.lodCovar.response.result;

                                        plotLODChart(data.response_data.lodCovar.response.result,
                                                     covar,
                                                     null);

                                        plotLODChart(data.response_data.lodCovar.response.result,
                                                     covar,
                                                     data.response_data.lod.response.result);
                                    }

                                    $('#interactiveCovarLODS').selectpicker('val', covar);

                                    // 2. Correlation
                                    displayCorrelation(data.response_data.correlation.response.result);

                                    // 3. Expression information
                                    g.expressionData = data.response_data.expression.response.result;
                                    if (g.ptmID !== null) {
                                        plotProfile(g.ptmID);
                                    } else {
                                        plotProfile(g.peptideID);
                                    }

                                    // first time we need to hide jumbotron instructions
                                    // it's ok to keep hiding
                                    $('#divWelcomeMesage').html('');
                                    $('#divLOD').removeClass('invisible');
                                    $('#divItemInfo').removeClass('invisible');
                                    $('#divSecondRow').removeClass('invisible');
                                    $('#divProfileCorrelation').removeClass('invisible');
                                    stopTask();
                                }
                            }
                        }
                        else {
                            // rerun in 1 seconds
                            logDebug('Not done, keep checking...');
                            setTimeout(function () {
                                updatePTMPeptideSelect(groupID, datasetID, covar);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                logDebug('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=False) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    logDebug(data);
                });
            }
        }

        function updateUniprotSelect(groupID, datasetID, covar) {
            // send GET request to status URL
            logDebug('update progress, calling .ajax');
            logDebug(groupID);

            if (g.runningTask) {
                let statusURL = `{{ url_for('api.api_status', group_id='', _external=False) }}${groupID}`;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    retries: 3,
                    retryInterval: 1000,
                    success: function (data, status, request) {
                        logDebug('DATA=======', data);

                        if (data.status === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data.number_tasks_errors !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.error}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data.number_tasks_errors === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if (value.status_code !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.response.error}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the LOD SCAN.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    if ('geneData' in data.response_data) {
                                        g.gene = data.response_data.geneData.response;
                                        displayGeneData(g.gene.gene);
                                        let geneData = null;
                                        for (let key in g.gene.gene) {
                                            geneData = g.gene.gene[key];
                                        }
                                    }

                                    if (covar === 'additive') {
                                        g.LODChartData[covar] = data.response_data.lod.response.result;
                                        plotLODChart(data.response_data.lod.response.result, covar, null);
                                    } else {
                                        g.LODChartData.additive = data.response_data.lod.response.result;
                                        g.LODChartData[covar] = data.response_data.lodCovar.response.result;

                                        plotLODChart(data.response_data.lodCovar.response.result,
                                                     covar,
                                                     null);

                                        plotLODChart(data.response_data.lodCovar.response.result,
                                                     covar,
                                                     data.response_data.lod.response.result);
                                    }

                                    $('#interactiveCovarLODS').selectpicker('val', covar);

                                    // 2. Correlation
                                    displayCorrelation(data.response_data.correlation.response.result);

                                    // 3. Expression information
                                    g.expressionData = data.response_data.expression.response.result;
                                    if (g.ptmID !== null) {
                                        plotProfile(g.ptmID);
                                    } else {
                                        plotProfile(g.peptideID);
                                    }

                                    // first time we need to hide jumbotron instructions
                                    // it's ok to keep hiding
                                    $('#divWelcomeMesage').html('');
                                    $('#divLOD').removeClass('invisible');
                                    $('#divItemInfo').removeClass('invisible');
                                    $('#divSecondRow').removeClass('invisible');
                                    $('#divProfileCorrelation').removeClass('invisible');
                                    stopTask();
                                }
                            }
                        }
                        else {
                            // rerun in 1 seconds
                            logDebug('Not done, keep checking...');
                            setTimeout(function () {
                                updateUniprotSelect(groupID, datasetID, covar);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                logDebug('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=False) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    logDebug(data);
                });
            }
        }

        function selectUniprot(uniprotID, proteinID, geneID, datasetID, covar) {
            let urlGeneData = `${g.PROTOCOL}//churchilllab.jax.org/ensimpl/api/gene/${geneID}`;
            urlGeneData += ('?release=' + g.DATASETS[datasetID].ensembl_release);
            urlGeneData += ('&species=' + g.DATASETS[datasetID].ensembl_species);

            let urlExpression = `{{ config.API_R_BASE }}/expression?dataset=${g.dataSetID}&id=${uniprotID}`;
            let urlLOD = `{{ config.API_R_BASE }}/lodscores?dataset=${g.dataSetID}&id=${uniprotID}`;
            let urlLODCovar = `{{ config.API_R_BASE }}/lodscores?dataset=${g.dataSetID}&id=${uniprotID}`;
            let urlCorrelation = `{{ config.API_R_BASE }}/correlation?dataset=${g.dataSetID}&id=${uniprotID}`;

            // TODO: make reset function
            g.gene = null;
            g.geneID = geneID;
            g.proteinID = proteinID;
            g.phosID = null;
            g.ptmID = null;
            g.uniprotID = uniprotID;
            g.ptmID = null;
            g.peptideID = null;
            g.phenotypeID = null;
            g.secondaryPlotMarkerID = null;
            g.secondaryPlotChromosome = null;
            g.secondaryPlotLocation = null;
            g.secondaryPlotCovar = null;
            g.correlationData = null;
            g.correlationDatasetID = datasetID;
            g.LODChartData = {};
            g.LODChartData.additive = null;

            // reset the covariate pickers
            configureCovarInfo(false);

            $.each(g.interactiveCovariates, function(idx, value) {
                g.LODChartData[value['column.name']] = null;
            });

            if (covar === 'additive') {
                urlLOD += '&intcovar=additive';
                urlLODCovar = null;
            } else {
                urlLOD += '&intcovar=additive';
                urlLODCovar += ('&intcovar=' + covar);
            }


            // TODO: add ncores?, regress_local? to urlLOD, or handle backend?
            //URL += '&ncores=' + _G.api_cores;
            {% if config.API_R_NUM_CORES %}
            if (urlLODCovar !== null) {
                urlLODCovar += '&cores={{ config.API_R_NUM_CORES }}';
            }
            urlLOD += '&cores={{ config.API_R_NUM_CORES }}';
            {% endif %}

            let submitData = {
                urls:[{
                    url_id: 'expression',
                    url: urlExpression
                }, {
                    url_id: 'correlation',
                    url: urlCorrelation
                }, {
                    url_id: 'lod',
                    url: urlLOD
                }]
            };

            if (geneID !== '') {
                submitData.urls.push({
                    url_id: 'geneData',
                    url_description: 'Gene Data',
                    url: urlGeneData
                });
            }

            if (urlLODCovar !== null) {
                submitData.urls.push({
                    url_id: 'lodCovar',
                    url: urlLODCovar
                });
            }

            logDebug('submitData=', submitData);

            startTask();

            $('#div_current_item_information_header').html('');
            $('#div_current_item_information_body').html('');

            g.chartLOD = null;
            g.chartLODCovariateFull = null;
            g.chartLODCovariateDiff = null;
            $('#lodPlotChart').html('');
            $('#lodPlotChartCovariateFull').html('');
            $('#lodPlotChartCovariateDiff').html('');

            g.chartCorrelation = null;
            $('#plotCorrelation').html('');
            $('#correlationDataTable').html('');
            $('#correlationImputation').html('');

            // The following line fires a change event which we do not want
            // $('#correlationDatasetSelect').selectpicker('val', g.dataSetID);

            // Do this instead
            $('#correlationDatasetSelect').val(g.dataSetID);
            $('#correlationDatasetSelect').selectpicker('render');
            $('#fact-svg-chart').html('');

            resetSecondaryPlots();

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=False) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                retries: 3,
                retryInterval: 1000,
                success: function(data, status, request) {
                    g.runningTask = true;
                    logDebug('data=', data);
                    logDebug('status=', status);
                    logDebug('request=', request);
                    //let status_url = request.getResponseHeader('Location');
                    //logDebug('status_url=', status_url);
                    logDebug('data.group_id=', data.group_id);
                    updateUniprotSelect(data.group_id, datasetID, covar);
                },
                error: function(jqXHR, textStatus, errorThrown) {
                    showErrorMessage(errorThrown, textStatus);
                }
            });
        }


        function selectPTMPeptide(ptmID, peptideID, proteinID, geneID, datasetID, covar) {
            logDebug('selectPTMPeptide=', ptmID, peptideID, proteinID, geneID, datasetID, covar);
            logDebug('ptmID=', ptmID);
            logDebug('peptideID=', peptideID);
            logDebug('proteinID=', proteinID);
            logDebug('geneID=', geneID);
            logDebug('datasetID=', datasetID);
            logDebug('covar=', covar);

            let elementID = peptideID;
            if ((ptmID !== undefined) && (ptmID !== '')) {
                elementID = ptmID;
            }
            let urlGeneData = `${g.PROTOCOL}//churchilllab.jax.org/ensimpl/api/gene/${geneID}`;
            urlGeneData += ('?release=' + g.DATASETS[datasetID].ensembl_release);
            urlGeneData += ('&species=' + g.DATASETS[datasetID].ensembl_species);


            let urlExpression = `{{ config.API_R_BASE }}/expression?dataset=${datasetID}&id=${elementID}`;
            let urlLOD = `{{ config.API_R_BASE }}/lodscores?dataset=${datasetID}&id=${elementID}`;
            let urlLODCovar = `{{ config.API_R_BASE }}/lodscores?dataset=${datasetID}&id=${elementID}`;
            let urlCorrelation = `{{ config.API_R_BASE }}/correlation?dataset=${datasetID}&id=${elementID}`;

            // TODO: make reset function
            g.gene = null;
            g.geneID = geneID;
            g.proteinID = proteinID;
            g.phosID = null;
            g.ptmID = null;
            g.uniprotID = null;
            g.peptideID = peptideID;

            if ((ptmID !== undefined) && (ptmID !== '')) {
                g.ptmID = ptmID;
            }

            g.phenotypeID = null;
            g.secondaryPlotMarkerID = null;
            g.secondaryPlotChromosome = null;
            g.secondaryPlotLocation = null;
            g.secondaryPlotCovar = null;
            g.correlationData = null;
            g.correlationDatasetID = datasetID;
            g.LODChartData = {};
            g.LODChartData.additive = null;

            // reset the covariate pickers
            configureCovarInfo(false);

            $.each(g.interactiveCovariates, function(idx, value) {
                g.LODChartData[value['column.name']] = null;
            });

            if (covar === 'additive') {
                urlLOD += '&intcovar=additive';
                urlLODCovar = null;
            } else {
                urlLOD += '&intcovar=additive';
                urlLODCovar += ('&intcovar=' + covar);
            }


            // TODO: add ncores?, regress_local? to urlLOD, or handle backend?
            //URL += '&ncores=' + _G.api_cores;
            {% if config.API_R_NUM_CORES %}
            if (urlLODCovar !== null) {
                urlLODCovar += '&cores={{ config.API_R_NUM_CORES }}';
            }
            urlLOD += '&cores={{ config.API_R_NUM_CORES }}';
            {% endif %}

            let submitData = {
                urls:[{
                    url_id: 'expression',
                    url: urlExpression
                }, {
                    url_id: 'correlation',
                    url: urlCorrelation
                }, {
                    url_id: 'lod',
                    url: urlLOD
                }]
            };

            if (geneID !== '') {
                submitData.urls.push({
                    url_id: 'geneData',
                    url_description: 'Gene Data',
                    url: urlGeneData
                });
            }

            if (urlLODCovar !== null) {
                submitData.urls.push({
                    url_id: 'lodCovar',
                    url: urlLODCovar
                });
            }

            logDebug('submitData=', submitData);

            startTask();

            $('#div_current_item_information_header').html('');
            $('#div_current_item_information_body').html('');

            g.chartLOD = null;
            g.chartLODCovariateFull = null;
            g.chartLODCovariateDiff = null;
            $('#lodPlotChart').html('');
            $('#lodPlotChartCovariateFull').html('');
            $('#lodPlotChartCovariateDiff').html('');

            g.chartCorrelation = null;
            $('#plotCorrelation').html('');
            $('#correlationDataTable').html('');
            $('#correlationImputation').html('');

            // The following line fires a change event which we do not want
            // $('#correlationDatasetSelect').selectpicker('val', g.dataSetID);

            // Do this instead
            $('#correlationDatasetSelect').val(g.dataSetID);
            $('#correlationDatasetSelect').selectpicker('render');
            $('#fact-svg-chart').html('');

            resetSecondaryPlots();

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=False) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                retries: 3,
                retryInterval: 1000,
                success: function(data, status, request) {
                    g.runningTask = true;
                    logDebug('data=', data);
                    logDebug('status=', status);
                    logDebug('request=', request);
                    //let status_url = request.getResponseHeader('Location');
                    //logDebug('status_url=', status_url);
                    logDebug('data.group_id=', data.group_id);
                    updatePTMPeptideSelect(data.group_id, datasetID, covar);
                },
                error: function(jqXHR, textStatus, errorThrown) {
                    showErrorMessage(errorThrown, textStatus);
                }
            });
        }

        function showErrorMessage(message, details) {
            let errorMessage = `<div class="row text-left"><div class="col"><span class="font-weight-bold text-danger">${message}</span></div></div>`;

            if (details !== null) {
                errorMessage += `<div class="row row-spacer"></div>
                                 <div class="row text-left">
                                     <div class="col">
                                         <div id="errorDetailAccordion" data-children=".item">
                                             <div class="item">
                                                 <a class="text-danger" data-toggle="collapse" data-parent="#errorDetailAccordion" href="#errorDetailAccordion1" role="button" aria-expanded="true" aria-controls="errorDetailAccordion1">
                                                     Details
                                                 </a>
                                                 <div id="errorDetailAccordion1" class="collapse" role="tabpanel">
                                                     <p class="mb-3">
                                                         ${details}
                                                     </p>
                                                 </div>
                                             </div>
                                         </div>
                                     </div>
                                 </div>`;
            }

            Swal.fire({
                  allowEnterKey: false,
                  allowEscapeKey: false,
                  allowOutsideClick: false,
                  buttonsStyling: false,
                  html: errorMessage,
                  showCancelButton: false,
                  showCloseButton: true,
                  showConfirmButton: false,
                  title: '<i class="fas fa-exclamation-circle"></i> Error!',
            });
        }

        function updatePhenotypeSelect(groupID, datasetID, covar) {
            // send GET request to status URL
            logDebug('update progress, calling .ajax');
            logDebug(groupID);

            if (g.runningTask) {
                let statusURL = `{{ url_for('api.api_status', group_id='', _external=False) }}${groupID}`;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    retries: 3,
                    retryInterval: 1000,
                    success: function (data, status, request) {
                        logDebug('DATA=======', data);

                        if (data.status === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data.number_tasks_errors !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.error}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data.number_tasks_errors === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if (value.status_code !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.response.error}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the LOD SCAN.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result, there will be 2 datasets to get
                                    displayPhenoData(g.phenotypeID);

                                    // 1. LOD score information
                                    //plotLODChart(data.response_data.lod.response.result);

                                    if (covar === 'additive') {
                                        g.LODChartData[covar] = data.response_data.lod.response.result;
                                        plotLODChart(data.response_data.lod.response.result, covar, null);
                                    } else {
                                        g.LODChartData.additive = data.response_data.lod.response.result;
                                        g.LODChartData[covar] = data.response_data.lodCovar.response.result;

                                        plotLODChart(data.response_data.lodCovar.response.result,
                                                     covar,
                                                     null);

                                        plotLODChart(data.response_data.lodCovar.response.result,
                                                     covar,
                                                     data.response_data.lod.response.result);
                                    }

                                    $('#interactiveCovarLODS').selectpicker('val', covar);

                                    // 2. Correlation
                                    displayCorrelation(data.response_data.correlation.response.result);

                                    // 3. Expression information
                                    g.expressionData = data.response_data.expression.response.result;
                                    plotProfile(g.phenotypeID);

                                    // first time we need to hide jumbotron instructions
                                    // it's ok to keep hiding
                                    $('#divWelcomeMesage').html('');
                                    $('#divLOD').removeClass('invisible');
                                    $('#divItemInfo').removeClass('invisible');
                                    $('#divSecondRow').removeClass('invisible');
                                    $('#divProfileCorrelation').removeClass('invisible');
                                    stopTask();
                                }
                            }
                        }
                        else {
                            // rerun in 1 seconds
                            logDebug('Not done, keep checking...');
                            setTimeout(function () {
                                updatePhenotypeSelect(groupID, datasetID, covar);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                logDebug('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=False) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    logDebug(data);
                });
            }
        }

        /**
         * Perform gene search.
         */
        function selectGeneProtein(geneID, proteinID, phosID, uniprotID, ptmID, peptideID, dataSetID, covar) {
            logDebug(`selectGeneProtein(${geneID}, ${proteinID}, ${phosID}, ${ptmID}, ${peptideID}, ${dataSetID}, ${covar})`);

            let urlGeneData = `${g.PROTOCOL}//churchilllab.jax.org/ensimpl/api/gene/${geneID}`;
            urlGeneData += ('?release=' + g.DATASETS[dataSetID].ensembl_release);
            urlGeneData += ('&species=' + g.DATASETS[dataSetID].ensembl_species);

            let urlExpression = `{{ config.API_R_BASE }}/expression?dataset=${dataSetID}&id=`;
            let urlLOD = `{{ config.API_R_BASE }}/lodscores?dataset=${dataSetID}&id=`;
            let urlLODCovar = `{{ config.API_R_BASE }}/lodscores?dataset=${dataSetID}&id=`;
            let urlCorrelation = `{{ config.API_R_BASE }}/correlation?dataset=${dataSetID}&id=`;

            g.gene = null;
            g.geneID = geneID;
            g.proteinID = proteinID;
            g.phosID = phosID;
            g.uniprotID = uniprotID;
            g.phenotypeID = null;;
            g.ptmID = ptmID;
            g.peptideID = peptideID;
            g.secondaryPlotMarkerID = null;
            g.secondaryPlotChromosome = null;
            g.secondaryPlotLocation = null;
            g.secondaryPlotCovar = null;
            g.correlationData = null;
            g.correlationDatasetID = dataSetID;
            g.LODChartData = {};
            g.LODChartData.additive = null;

            //$('#factorSeriesCorrelation').selectpicker();

            $.each(g.interactiveCovariates, function(idx, value) {
                g.LODChartData[value['column.name']] = null;
            });

            // reset the covariate pickers
            configureCovarInfo(false);

            if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                urlExpression += geneID;
                urlLOD += geneID;
                urlLODCovar += geneID;
                urlCorrelation += geneID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                urlExpression += proteinID;
                urlLOD += proteinID;
                urlLODCovar += proteinID;
                urlCorrelation += proteinID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                urlExpression += uniprotID;
                urlLOD += uniprotID;
                urlLODCovar += uniprotID;
                urlCorrelation += uniprotID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                urlExpression += phosID;
                urlLOD += phosID;
                urlLODCovar += phosID;
                urlCorrelation += phosID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                urlExpression += ptmID;
                urlLOD += ptmID;
                urlLODCovar += ptmID;
                urlCorrelation += ptmID;
            } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                urlExpression += peptideID;
                urlLOD += peptideID;
                urlLODCovar += peptideID;
                urlCorrelation += peptideID;
            }

            if (covar === 'additive') {
                urlLOD += '&intcovar=additive';
                urlLODCovar = null;
            } else {
                urlLOD += '&intcovar=additive';
                urlLODCovar += ('&intcovar=' + covar);
            }

            // TODO: add ncores?, regress_local? to urlLOD, or handle backend?
            //URL += '&ncores=' + _G.api_cores;
            {% if config.API_R_NUM_CORES %}
            if (urlLODCovar !== null) {
                urlLODCovar += '&cores={{ config.API_R_NUM_CORES }}';
            }
            urlLOD += '&cores={{ config.API_R_NUM_CORES }}';
            {% endif %}

            let submitData = {
                urls:[{
                    url_id: 'geneData',
                    url_description: 'Gene Data',
                    url: urlGeneData
                }, {
                    url_id: 'expression',
                    url_description: 'Expression Data',
                    url: urlExpression
                }, {
                    url_id: 'correlation',
                    url_description: 'Correlation Data',
                    url: urlCorrelation
                }, {
                    url_id: 'lod',
                    url_description: 'LOD Scan Data',
                    url: urlLOD
                }]};

            if (urlLODCovar !== null) {
                submitData.urls.push({
                    url_id: 'lodCovar',
                    url_description: 'Covariate LOD Scan Data',
                    url: urlLODCovar
                });
            }

            logDebug('submitData=', submitData);

            startTask();

            $('#div_current_item_information_header').html('');
            $('#div_current_item_information_body').html('');

            g.chartLOD = null;
            g.chartLODCovariateFull = null;
            g.chartLODCovariateDiff = null;
            $('#lodPlotChart').html('');
            $('#lodPlotChartCovariateFull').html('');
            $('#lodPlotChartCovariateDiff').html('');

            g.chartCorrelation = null;
            $('#plotCorrelation').html('');
            $('#correlationDataTable').html('');
            $('#correlationImputation').html('');

            // The following line fires a change event which we do not want
            // $('#correlationDatasetSelect').selectpicker('val', g.dataSetID);

            // Do this instead
            $('#correlationDatasetSelect').val(g.dataSetID);
            $('#correlationDatasetSelect').selectpicker('render');
            $('#correlationImputation').html('');

            $('#fact-svg-chart').html('');

            resetSecondaryPlots();

            $.ajax({
                type: 'POST',
                url: '{{ url_for('api.api_submit', _external=False) }}',
                contentType: 'application/json',
                data: JSON.stringify(submitData),
                retries: 3,
                retryInterval: 1000,
                success: function(data, status, request) {
                    g.runningTask = true;
                    logDebug('data=', data);
                    logDebug('status=', status);
                    logDebug('request=', request);
                    logDebug('data.group_id=', data.group_id);
                    updateGeneSelect(data.group_id, submitData, geneID, proteinID, phosID, uniprotID, ptmID, peptideID, dataSetID, covar);
                },
                error: function(jqXHR, textStatus, errorThrown) {
                    showErrorMessage(errorThrown, textStatus);
                }
            });
        }

        function updateChangeCovariate(groupID, covar) {
            // send GET request to status URL
            logDebug('update progress, calling .ajax');

            logDebug(groupID);

            if (g.runningTask) {
                let statusURL = `{{ url_for('api.api_status', group_id='', _external=False) }}${groupID}`;
                $.ajax({
                    type: 'GET',
                    url: statusURL,
                    retries: 3,
                    retryInterval: 1000,
                    success: function (data, status, request) {
                        logDebug('DATA=======', data);

                        /* A response will be like this:

                           {
                            number_tasks_completed: 3,
                            number_tasks_errors: 2,
                            number_tasks_submitted: 3,
                            response_data: {
                                lod: {
                                    error: "Connection Error"
                                    error_message: "HTTPConnectionPool(host='myapiisawesome', port=8000): Max retries exceeded with url: /lodscores?dataset=dataset.islet.rnaseq&nCores=2&id=ENSMUSG00000023224 (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f861126ce48>: Failed to establish a new connection: [Errno -5] No address associated with hostname',))"
                                    url: "http://myapiisawesome:8000/lodscores?dataset=dataset.islet.rnaseq&nCores=2&id=ENSMUSG00000023224"
                                    url_id: "lod"
                                }
                            },
                            status: "DONE",
                            task_id: "d3c0b971-9794-4ed7-80cb-046710d2f9f5",
                            error: 'UNKNOWN ERROR'   <--- ONLY WHEN AN ERROR
                           }
                        */

                        if (data.status === 'DONE') {
                            if ('error' in data) {
                                // MAJOR ERROR
                                let message = `Unfortunately, there was a problem contacting the server.  Please try again.`;
                                stopTask();
                                showErrorMessage(message, null);
                            } else if (data.number_tasks_errors !== 0) {
                                let message = `Unfortunately, we encountered an error.  Please try again.`;
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if ('error' in value) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.error}<br>`);
                                    }
                                });

                                stopTask();
                                showErrorMessage(message, errorMessages);
                            } else if (data.number_tasks_errors === 0) {
                                // check to make sure the status codes are good
                                let errorMessages = '';
                                $.each(data.response_data, function (key, value) {
                                    if (value.status_code !== 200) {
                                        errorMessages += (`<strong>${key}:</strong> ${value.response.error}<br>`);
                                    }
                                });

                                if (errorMessages !== '') {
                                    let message = `Unfortunately, there was a problem calculating the LOD SCAN.`;
                                    stopTask();
                                    showErrorMessage(message, errorMessages);
                                } else {
                                    // show result, there will be 1
                                    g.LODChartData[covar] = data.response_data.lod.response.result;

                                    plotLODChart(data.response_data.lod.response.result,
                                                 covar,
                                                 null);

                                    plotLODChart(data.response_data.lod.response.result,
                                                 covar,
                                                 g.LODChartData.additive);

                                    stopTask();
                                }
                            }
                        }
                        else {
                            // rerun in 1 seconds
                            logDebug('Not done, keep checking...');
                            setTimeout(function () {
                                updateChangeCovariate(groupID, covar);
                            }, 1000);  // TODO: change to 1000 (1 second)
                        }
                    }
                });
            } else {
                // TODO: cleanup
                logDebug('canceling');
                let cancelURL = `{{ url_for('api.api_cancel', group_id='', _external=False) }}${groupID}`;
                $.getJSON(cancelURL, function (data) {
                    logDebug(data);
                });
            }
        }

        /**
         * Change covar LOD plot.
         */
         function changeCovariate(covariate) {
            logDebug('changeCovariate(', covariate, ')');

            if (covariate === 'additive') {
                // remove all other graphs
                logDebug('additive, clear others');
                g.chartLODCovariateFull = null;
                g.chartLODCovariateDiff = null;
                $('#lodPlotChartCovariateFull').html('');
                $('#lodPlotChartCovariateDiff').html('');
            } else {
                g.chartLODChart = null;
                g.chartLODCovariateFull = null;
                g.chartLODCovariateDiff = null;
                $('#lodPlotChart').html('');
                $('#lodPlotChartCovariateFull').html('');
                $('#lodPlotChartCovariateDiff').html('');
            }


            // check to see if we have the data, if so, plot it
            // else go get it
            if (covariate in g.LODChartData) {
                logDebug('covar in g.LODChartData, so just add it');
                if (covariate === 'additive') {
                    plotLODChart(g.LODChartData[covariate], covariate, null);
                } else {
                    plotLODChart(g.LODChartData[covariate], covariate, null);
                    plotLODChart(g.LODChartData[covariate], covariate, g.LODChartData.additive);
                }
            } else {
                logDebug('covar NOT in g.LODChartData, so go get it');
                let urlLOD = `{{ config.API_R_BASE }}/lodscores?dataset=${g.dataSetID}&id=`;

                g.secondaryPlotMarkerID = null;
                g.secondaryPlotChromosome = null;
                g.secondaryPlotLocation = null;
                g.secondaryPlotCovar = null;

                if (g.DATASETS[g.dataSetID].datatype === 'mrna') {
                    urlLOD += `${g.geneID}`;
                } else if (g.DATASETS[g.dataSetID].datatype === 'protein') {
                    urlLOD += `${g.proteinID}`;
                } else if (g.DATASETS[g.dataSetID].datatype === 'protein_uniprot') {
                    urlLOD += `${g.uniprotID}`;
                } else if (g.DATASETS[g.dataSetID].datatype === 'phos') {
                    urlLOD += `${g.phosID}`;
                } else if (g.DATASETS[g.dataSetID].datatype === 'peptide') {
                    urlLOD += `${g.peptideID}`;
                } else if (g.DATASETS[g.dataSetID].datatype === 'ptm') {
                    urlLOD += `${g.ptmID}`;
                } else if (g.DATASETS[g.dataSetID].datatype === 'pheno') {
                    urlLOD += `${g.phenotypeID}`;
                } else {
                    // TODO: handle error
                    logError('MAJOR PROBLEM');
                }

                urlLOD += ('&intcovar=' + covariate);

                // TODO: add ncores?, regress_local? to urlLOD, or handle backend?
                //URL += '&ncores=' + _G.api_cores;
                {% if config.API_R_NUM_CORES %}
                urlLOD += '&cores={{ config.API_R_NUM_CORES }}';
                {% endif %}

                let submitData = {
                    urls:[{
                        url_id: 'lod',
                        url: urlLOD
                    }]};

                startTask();

                $('#fact-svg-chart').html('');

                resetSecondaryPlots();

                $.ajax({
                    type: 'POST',
                    url: '{{ url_for('api.api_submit', _external=False) }}',
                    contentType: 'application/json',
                    data: JSON.stringify(submitData),
                    retries: 3,
                    retryInterval: 1000,
                    success: function(data, status, request) {
                        g.runningTask = true;
                        logDebug('data=', data);
                        logDebug('status=', status);
                        logDebug('request=', request);
                        //let status_url = request.getResponseHeader('Location');
                        //logDebug('status_url=', status_url);
                        logDebug('data.group_id=', data.group_id);
                        updateChangeCovariate(data.group_id, covariate);
                    },
                    error: function(jqXHR, textStatus, errorThrown) {
                        showErrorMessage(errorThrown, textStatus);
                    }
                });
            }
        }

        return {
            g: g,
            init: init,
            findGene: findGene,
            startTask: startTask,
            stopTask: stopTask
        }
    }();

    $().ready(function () {
        QTL.init();

        // handle search
        $('#searchTerm').keypress(function(evt) {
            let code = evt.which ? evt.which : evt.keyCode;
            if (code === 13) {
                $(this).blur();
                $('#btnGo').focus().click();
            }
        });

        $('#btnGo').button().on('click', function(evt) {
            evt.preventDefault();
            let term = $('#searchTerm');
            QTL.findGene(term.val().trim().toUpperCase());
            return false;
        });

        // set some global options for Highcharts
        Highcharts.setOptions({
            chart: {
                style: {
                    fontFamily: $('body').css('font-family')
                },
            },

            credits: {
                enabled: false
            },

            exporting: {
                buttons: {
                    contextButton: {
                        menuItems: [
                            'downloadPNG'
                        ]
                    }
                },
            },
        });

        $('a[data-toggle="tab"]').on('shown.bs.tab', function(e){
           $($.fn.dataTable.tables(true)).DataTable().columns.adjust();
        });

        logDebug('end ready');
    });
    </script>

{% endblock %}

